char *fimo_html_string  = 
"<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\">\n"
"<html xmlns:cis=\"http://zlab.bu.edu/schema/cisml\" xmlns:fimo=\"http://noble.gs.washington.edu/schema/cisml\" xmlns:mem=\"http://noble.gs.washington.edu/meme\">\n"
"<head>\n"
"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\">\n"
"<meta charset=\"UTF-8\">\n"
"<title>FIMO Results</title>\n"
"<style type=\"text/css\">\n"
"td.left {text-align: left;}\n"
"td.right {text-align: right; padding-right: 1cm;}\n"
"</style>\n"
"</head>\n"
"<body color=\"#000000\">\n"
"<a name=\"top_buttons\"></a>\n"
"<hr>\n"
"<table summary=\"buttons\" align=\"left\" cellspacing=\"0\">\n"
"<tr>\n"
"<td bgcolor=\"#00FFFF\"><a href=\"#database_and_motifs\"><b>Database and Motifs</b></a></td>\n"
"<td bgcolor=\"#DDFFDD\"><a href=\"#sec_i\"><b>High-scoring Motif Occurences</b></a></td>\n"
"<td bgcolor=\"#DDDDFF\"><a href=\"#debugging_information\"><b>Debugging Information</b></a></td>\n"
"<td bgcolor=\"#FFDD00\"><a href=\"fimo.tsv\"\"><b>Results in TSV Format</b></a></td>\n"
"<td bgcolor=\"#FFDDFF\"><a href=\"fimo.gff\"\"><b>Results in GFF3 Format</b></a></td>\n"
"</tr>\n"
"</table>\n"
"<br/>\n"
"<br/>\n"
"<hr/>\n"
"<center><big><b>FIMO - Motif search tool</b></big></center>\n"
"<hr>\n"
"<p>\n"
"@version@\n"
"</p>\n"
"<p>\n"
"@interpreting@\n"
"</p>\n"
"<p>If you use FIMO in your research, please cite the following paper:<br>\n"
"Charles E. Grant, Timothy L. Bailey, and William Stafford Noble,\n"
"\"FIMO: Scanning for occurrences of a given motif\",\n"
"<i>Bioinformatics</i>, <b>27</b>(7):1017-1018, 2011.\n"
"<a href=\"http://bioinformatics.oxfordjournals.org/content/27/7/1017\">[full text]</a></p>\n"
"<hr>\n"
"<center><big><b><a name=\"database_and_motifs\">DATABASE AND MOTIFS</a></b></big></center>\n"
"<hr>\n"
"<div style=\"padding-left: 0.75in; line-height: 1em; font-family: monospace;\">\n"
"@database_and_motifs@\n"
"</div>\n"
"<hr>\n"
"<center><big><b><a name=\"sec_i\">SECTION I: HIGH-SCORING MOTIF OCCURENCES</a></b></big></center>\n"
"<hr>\n"
"<ul>\n"
"<li>\n"
"@counts@\n"
"The full set of motif occurences can be seen in the\n"
"TSV (tab-delimited values) output file\n"
"<a href=\"fimo.tsv\">fimo.tsv</a>, \n"
"the GFF3 file \n"
"<a href=\"fimo.gff\">fimo.gff</a> \n"
"which may be suitable for uploading to the \n"
"<a href=\"http://genome.ucsc.edu/cgi-bin/hgTables\">UCSC Genome Table Browser</a>\n"
"(assuming the FASTA input sequences included genomic coordinates in UCSC or Galaxy format),\n"
"or the XML file \n"
"<a href=\"fimo.xml\">fimo.xml</a>.\n"
"</li>\n"
"<li>\n"
"The <i>p</i>-value of a motif occurrence is defined as the\n"
"probability of a random sequence of the same length as the motif\n"
"matching that position of the sequence with as good or better a score.\n"
"</li>\n"
"<li>\n"
"The score for the match of a position in a sequence to a motif\n"
"is computed by summing the appropriate entries from each column of\n"
"the position-dependent scoring matrix that represents the motif.\n"
"</li>\n"
"<li>\n"
"The q-value of a motif occurrence is defined as the\n"
"false discovery rate if the occurrence is accepted as significant.\n"
"</li>\n"
"<li>The table is sorted by increasing <i>p</i>-value.</li>\n"
"</ul>\n"
"@match_table@\n"
"<hr>\n"
"<center><big><b><a name=\"debugging_information\">DEBUGGING INFORMATION</a></b></big></center>\n"
"<hr>\n"
"<p>\n"
"Command line:\n"
"</p>\n"
"<pre>\n"
"@command_line@\n"
"</pre>\n"
"<p>\n"
"Settings:\n"
"</p>\n"
"<pre>\n"
"<table>\n"
"@parameters@\n"
"</table>\n"
"</pre>\n"
"<p>\n"
"This information can be useful in the event you wish to report a\n"
"problem with the FIMO software.\n"
"</p>\n"
"<hr>\n"
"<span style=\"background-color: #DDDDFF\"><a href=\"#top_buttons\"><b>Go to top</b></a></span>\n"
"</body>\n"
"</html>\n"
"";

/******************
void fimo_print_version();
void fimo_print_database_and_motifs();
void fimo_print_counts();
void fimo_print_match_table();
void fimo_print_command_line();
void fimo_print_parameters();
******************/

