<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="UTF-8">
<title>GOMo Results</title>
<style type="text/css">
    
    /* START INCLUDED FILE "meme.css" */
/* The following is the content of meme.css */
body { background-color:white; font-size: 12px; font-family: Verdana, Arial, Helvetica, sans-serif;}

div.help {
  display: inline-block;
  margin: 0px;
  padding: 0px;
  width: 12px;
  height: 13px;
  cursor: pointer;
  background-image: url(data:image/gif;base64,R0lGODlhDAANAIABANR0AP///yH5BAEAAAEALAAAAAAMAA0AAAIdhI8Xy22MIFgv1DttrrJ7mlGNNo4c+aFg6SQuUAAAOw==);
}

div.help:hover {
  background-image: url(data:image/gif;base64,R0lGODlhDAANAKEAANR0AP///9R0ANR0ACH+EUNyZWF0ZWQgd2l0aCBHSU1QACH5BAEAAAIALAAAAAAMAA0AAAIdDGynCe3PgoxONntvwqz2/z2K2ImjR0KhmSIZUgAAOw==);
}

p.spaced { line-height: 1.8em;}

span.citation { font-family: "Book Antiqua", "Palatino Linotype", serif; color: #004a4d;}

p.pad { padding-left: 30px; padding-top: 5px; padding-bottom: 10px;}

td.jump { font-size: 13px; color: #ffffff; background-color: #00666a;
  font-family: Georgia, "Times New Roman", Times, serif;}

a.jump { margin: 15px 0 0; font-style: normal; font-variant: small-caps;
  font-weight: bolder; font-family: Georgia, "Times New Roman", Times, serif;}

h2.mainh {font-size: 1.5em; font-style: normal; margin: 15px 0 0;
  font-variant: small-caps; font-family: Georgia, "Times New Roman", Times, serif;}

h2.line {border-bottom: 1px solid #CCCCCC; font-size: 1.5em; font-style: normal;
  margin: 15px 0 0; padding-bottom: 3px; font-variant: small-caps;
  font-family: Georgia, "Times New Roman", Times, serif;}

h4 {border-bottom: 1px solid #CCCCCC; font-size: 1.2em; font-style: normal;
  margin: 10px 0 0; padding-bottom: 3px; font-family: Georgia, "Times New Roman", Times, serif;}

h5 {margin: 0px}

a.help { font-size: 9px; font-style: normal; text-transform: uppercase;
  font-family: Georgia, "Times New Roman", Times, serif;}

div.pad { padding-left: 30px; padding-top: 5px; padding-bottom: 10px;}

div.pad1 { margin: 10px 5px;}

div.pad2 { margin: 25px 5px 5px;}
h2.pad2 { padding: 25px 5px 5px;}

div.pad3 { padding: 5px 0px 10px 30px;}

div.box { border: 2px solid #CCCCCC; padding:10px; overflow: hidden;}

div.bar { border-left: 7px solid #00666a; padding:5px; margin-top:25px; }

div.subsection {margin:25px 0px;}

img {border:0px none;}

th.majorth {text-align:left;}
th.minorth {font-weight:normal; text-align:left; width:8em; padding: 3px 0px;}
th.actionth {font-weight:normal; text-align:left;}

.explain h5 {font-size:1em; margin-left: 1em;}

div.doc {margin-left: 2em; margin-bottom: 3em;}

th.trainingset {
  border-bottom: thin dashed black; 
  font-weight:normal; 
  padding:0px 10px;
}
div.pop_content {
  position:absolute;
  z-index:50;
  width:300px;
  padding: 5px;
  background: #E4ECEC;
  font-size: 12px;
  font-family: Arial;
  border-style: double;
  border-width: 3px;
  border-color: #AA2244;
  display:none;
}
div.pop_content_wide {
  position:absolute;
  z-index:1;
  width:700px;
  padding: 5px;
  background: #E4ECEC;
  font-size: 12px;
  font-family: Arial;
  border-style: double;
  border-width: 3px;
  border-color: #AA2244;
  display:none;
}

div.pop_content > *:first-child {
  margin-top: 0px;
}

div.pop_content h1, div.pop_content h2, div.pop_content h3, div.pop_content h4, 
div.pop_content h5, div.pop_content h6, div.pop_content p {
  margin: 0px;
}

div.pop_content p + h1, div.pop_content p + h2, div.pop_content p + h3, 
div.pop_content p + h4, div.pop_content p + h5, div.pop_content p + h6 {
  margin-top: 5px;
}

div.pop_content p + p {
  margin-top: 5px;
}

div.pop_content > *:last-child {
  margin-bottom: 0px;
}

div.pop_content div.pop_close {
  /* old definition */
  float:right;
  bottom: 0;
}

div.pop_content span.pop_close, div.pop_content span.pop_back {
  display: inline-block;
  border: 2px outset #661429;
  background-color: #CCC;
  padding-left: 1px;
  padding-right: 1px;
  padding-top: 0px;
  padding-bottom: 0px;
  cursor: pointer;
  color: #AA2244; /*#661429;*/
  font-weight: bold;
}

div.pop_content span.pop_close:active, div.pop_content span.pop_back:active {
  border-style: inset;
}

div.pop_content span.pop_close {
  float:right;
  /*border: 2px outset #AA002B;*/
  /*color: #AA2244;*/
}

div.pop_content:not(.nested) .nested_only {
  display: none;
}

div.pop_back_sec {
  margin-bottom: 5px;
}

div.pop_close_sec {
  margin-top: 5px;
}

table.hide_advanced tr.advanced {
  display: none;
}
span.show_more {
  display: none;
}
table.hide_advanced span.show_more {
  display: inline;
}
table.hide_advanced span.show_less {
  display: none;
}


/*****************************************************************************
 * Program logo styling
 ****************************************************************************/
div.prog_logo {
  border-bottom: 0.25em solid #0f5f60;
  height: 4.5em;
  width: 25em;
  display:inline-block;
}
div.prog_logo img {
  float:left;
  width: 4em;
  border-style: none;
  margin-right: 0.2em;
}
div.prog_logo h1, div.prog_logo h1:hover, div.prog_logo h1:active, div.prog_logo h1:visited {
  margin:0;
  padding:0;
  font-family: Arial, Helvetica,  sans-serif;
  font-size: 3.2em;
  line-height: 1em;
  vertical-align: top;
  display: block;
  color: #026666;
  letter-spacing: -0.06em;
  text-shadow: 0.04em 0.06em 0.05em #666;
}
div.prog_logo h2, div.prog_logo h2:hover, div.prog_logo h2:active, div.prog_logo h2:visited {
  display: block;
  margin:0;
  padding:0;
  font-family: Helvetica, sans-serif;
  font-size: 0.9em;
  line-height: 1em;
  letter-spacing: -0.06em;
  color: black;
}
div.prog_logo h3, div.prog_logo h3:hover, div.prog_logo h3:active, div.prog_logo h3:visited {
  display: block;
  margin:0;
  padding:0;
  font-family: Helvetica, sans-serif;
  font-size: 0.9em;
  line-height: 1.5em;
  letter-spacing: -0.06em;
  color: black;
}

div.big.prog_logo {
  font-size: 18px;
}

/* These are for centered columns in tables */
td.ctr {
  text-align: center;
}

/* These are for the navigation bars at the top of outputs. */
table.navigation {
  margin-top: 0px;
  border-collapse:collapse;
}
table.navigation * td
{
  padding-left: 0px;
  padding-right: 10px;
  padding-top: 0px;
  padding-bottom: 0px;
}
    /* END INCLUDED FILE "meme.css" */
    
  
        body { font-size: 12px; font-family: Verdana, Arial, Helvetica, sans-serif;}

        a.jump { margin: 15px 0 0; font-style: normal; font-variant: small-caps;
          font-weight: bolder; font-family: Georgia, "Times New Roman", Times, serif;}

        h2.mainh { font-size: 1.5em; font-style: normal; margin: 15px 0 0;
          font-variant: small-caps; font-family: Georgia, "Times New Roman", Times, serif;}

        h2.line { border-bottom: 1px solid #CCCCCC; font-size: 1.5em; font-style: normal;
          margin: 15px 0 0; padding-bottom: 3px; font-variant: small-caps;
          font-family: Georgia, "Times New Roman", Times, serif;}

        h4 { border-bottom: 1px solid #CCCCCC; font-size: 1.2em; font-style: normal;
          margin: 10px 0 0; padding-bottom: 3px; 
          font-family: Georgia, "Times New Roman", Times, serif;}

        a.help { font-size: 9px; font-style: normal; text-transform: uppercase;
          font-family: Georgia, "Times New Roman", Times, serif;}

        div.pad1 { margin: 10px 5px;}

        div.pad2 { margin: 25px 5px 5px;}

        div.pad3 { padding: 5px 0px 10px 30px;}

        div.box { border: 2px solid #CCCCCC; padding:10px;}

        div.bar { border-left: 7px solid #00666a; padding:5px; }

        img { border: none; border-width: 0px;}

        .implied a:link { color: darkslategrey;}

        .implied a:visited { color: darkslateblue;}

        tr.normal { color: black; border-bottom: 1px solid #CCCCCC;}

        tr.implied { color: grey; border-bottom: 1px solid #CCCCCC;}

        .bp { color: #770000;}

        .cc { color: #CC6600;}

        .mf { color: #007700;}

        col.em25 {width: 25em;}

        col.em7 {width: 7em;}

        col.em6 {width: 6em;}

        col.em2 {width: 2em;}

        textarea.cmd {width: 100%;}

        .explain h5 {margin-left: 1em;}

        .explain p {margin-left: 2em;}

    </style>
<script type="text/javascript">
    
    /* START INCLUDED FILE "constants.js" */
var site_url = "https://meme-suite.org/meme";
    /* END INCLUDED FILE "constants.js" */
    
  
    
    /* START INCLUDED FILE "utilities.js" */
/*
 * $
 *
 * Shorthand function for getElementById
 */
function $(el) {
  return document.getElementById(el);
}

/*
	"new" icon
*/
var new_icon_src = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABwAAAAQCAMAAAAyEe/dAAAACVBMVEX//wAAAAD////s2cV/AAAAdUlEQVQYlXVRBxLAIAhL+P+jC2HZhXcBZEWEldDsZcLIcAhHWWnK8SDcWQhMFUHdAQ1CqQ5+CWPmlHojl+nCJNRtzu4qRc3IUzmTVpXYK0nox0z0PI1stgchdK7lEv7ekhvalw8WW547Gyzedt/2/gLx8WXjXF/1AYFriNAWAAAAAElFTkSuQmCC";

/*
 * See http://stackoverflow.com/a/5450113/66387
 * Does string multiplication like the perl x operator.
 */
function string_mult(pattern, count) {
    if (count < 1) return '';
    var result = '';
    while (count > 1) {
        if (count & 1) result += pattern;
        count >>= 1, pattern += pattern;
    }
    return result + pattern;
}

/*
 * See http://stackoverflow.com/questions/814613/how-to-read-get-data-from-a-url-using-javascript
 * Slightly modified with information from
 * https://developer.mozilla.org/en/DOM/window.location
 */
function parse_params() {
  "use strict";
  var search, queryStart, queryEnd, query, params, nvPairs, i, nv, n, v;
  search = window.location.search;
  queryStart = search.indexOf("?") + 1;
  queryEnd   = search.indexOf("#") + 1 || search.length + 1;
  query      = search.slice(queryStart, queryEnd - 1);

  if (query === search || query === "") return {};

  params  = {};
  nvPairs = query.replace(/\+/g, " ").split("&");

  for (i = 0; i < nvPairs.length; i++) {
    nv = nvPairs[i].split("=");
    n  = decodeURIComponent(nv[0]);
    v  = decodeURIComponent(nv[1]);
    // allow a name to be used multiple times
    // storing each value in the array
    if (!(n in params)) {
      params[n] = [];
    }
    params[n].push(nv.length === 2 ? v : null);
  }
  return params;
}

/*
 * coords
 *
 * Calculates the x and y offset of an element.
 * From http://www.quirksmode.org/js/findpos.html
 * with alterations to take into account scrolling regions
 */
function coords(elem) {
  var myX = myY = 0;
  if (elem.getBoundingClientRect) {
    var rect;
    rect = elem.getBoundingClientRect();
    myX = rect.left + ((typeof window.pageXOffset !== "undefined") ?
        window.pageXOffset : document.body.scrollLeft);
    myY = rect.top + ((typeof window.pageYOffset !== "undefined") ?
        window.pageYOffset : document.body.scrollTop);
  } else {
    // this fall back doesn't properly handle absolutely positioned elements
    // inside a scrollable box
    var node;
    if (elem.offsetParent) {
      // subtract all scrolling
      node = elem;
      do {
        myX -= node.scrollLeft ? node.scrollLeft : 0;
        myY -= node.scrollTop ? node.scrollTop : 0;
      } while (node = node.parentNode);
      // this will include the page scrolling (which is unwanted) so add it back on
      myX += (typeof window.pageXOffset !== "undefined") ? window.pageXOffset : document.body.scrollLeft;
      myY += (typeof window.pageYOffset !== "undefined") ? window.pageYOffset : document.body.scrollTop;
      // sum up offsets
      node = elem;
      do {
        myX += node.offsetLeft;
        myY += node.offsetTop;
      } while (node = node.offsetParent);
    }
  }
  return [myX, myY];
}

/*
 * position_popup
 *
 * Positions a popup relative to an anchor element.
 *
 * The available positions are:
 * 0 - Centered below the anchor.
 */
function position_popup(anchor, popup, position) {
  "use strict";
  var a_x, a_y, a_w, a_h, p_x, p_y, p_w, p_h;
  var a_xy, spacer, margin, scrollbar, page_w;
  // define constants
  spacer = 5;
  margin = 15;
  scrollbar = 15;
  // define the positions and widths
  a_xy = coords(anchor);
  a_x = a_xy[0];
  a_y = a_xy[1];
  a_w = anchor.offsetWidth;
  a_h = anchor.offsetHeight;
  p_w = popup.offsetWidth;
  p_h = popup.offsetHeight;
  page_w = null;
  if (window.innerWidth) {
    page_w = window.innerWidth;
  } else if (document.body) {
    page_w = document.body.clientWidth;
  }
  // check the position type is defined
  if (typeof position !== "number") {
    position = 0;
  }
  // calculate the popup position
  switch (position) {
    case 1:
      p_x = a_x + a_w + spacer;
      p_y = a_y + (a_h / 2) - (p_h / 2);
      break;
    case 0:
    default:
      p_x = a_x + (a_w / 2) - (p_w / 2);
      p_y = a_y + a_h + spacer;
      break;
  }
  // constrain the popup position
  if (p_x < margin) {
    p_x = margin;
  } else if (page_w != null && (p_x + p_w) > (page_w - margin - scrollbar)) {
    p_x = page_w - margin - scrollbar - p_w;
  }
  if (p_y < margin) {
    p_y = margin;
  }
  // position the popup
  popup.style.left = p_x + "px";
  popup.style.top = p_y + "px";
}

function lookup_help_popup(popup_id) {
  var _body, pop, info;
  pop = document.getElementById(popup_id);
  if (pop == null) {
    _body = document.getElementsByTagName("body")[0];
    pop = document.createElement("div");
    pop.className = "pop_content";
    pop.id = popup_id;
    pop.style.backgroundColor = "#FFC";
    pop.style.borderColor = "black";
    info = document.createElement("p");
    info.style.fontWeight = "bold";
    info.appendChild(document.createTextNode("Error: No popup for topic \"" + popup_id + "\"."));
    pop.appendChild(info);
    // this might cause problems with the menu, but as this only happens
    // when something is already wrong I don't think that's too much of a problem
    _body.insertBefore(pop, _body.firstChild);
  }
  if (document.getElementsByTagName('body')[0].hasAttribute("data-autobtns")) {
    if (!/\bauto_buttons\b/.test(pop.className)) {
      pop.className += " auto_buttons";
      var back_btn_sec = document.createElement("div");
      back_btn_sec.className = "nested_only pop_back_sec";
      var back_btn = document.createElement("span");
      back_btn.className = "pop_back";
      back_btn.appendChild(document.createTextNode("<< back"));
      back_btn.addEventListener("click", function(e) {
        help_return();
      }, false);
      back_btn_sec.appendChild(back_btn);
      pop.insertBefore(back_btn_sec, pop.firstChild);
      var close_btn_sec = document.createElement("div");
      close_btn_sec.className = "pop_close_sec";
      var close_btn = document.createElement("span");
      close_btn.className = "pop_close";
      close_btn.appendChild(document.createTextNode("close"));
      close_btn.addEventListener("click", function(e) {
        help_popup();
      }, false);
      close_btn_sec.appendChild(close_btn);
      pop.appendChild(close_btn_sec);
    }
  }
  return pop;
}

/*
 * help_popup
 *
 * Moves around help pop-ups so they appear
 * below an activator.
 */
function help_popup(activator, popup_id) {
  "use strict";
  var pop;
  // set default values
  if (typeof help_popup.popup === "undefined") {
    help_popup.popup = [];
  }
  if (typeof help_popup.activator === "undefined") {
    help_popup.activator = null;
  }
  var last_pop = (help_popup.popup.length > 0 ? help_popup.popup[help_popup.popup.length - 1] : null);
  if (typeof(activator) == "undefined") { // no activator so hide
    if (last_pop != null) {
      last_pop.style.display = 'none';
      help_popup.popup = [];
    }
    return;
  }
  pop = lookup_help_popup(popup_id);
  if (pop == last_pop) {
    if (activator == help_popup.activator) {
      //hide popup (as we've already shown it for the current help button)
      last_pop.style.display = 'none';
      help_popup.popup = [];
      return; // toggling complete!
    }
  } else if (last_pop != null) {
    //activating different popup so hide current one
    last_pop.style.display = 'none';
  }
  help_popup.popup = [pop];
  help_popup.activator = activator;
  toggle_class(pop, "nested", false);
  //must make the popup visible to measure it or it has zero width
  pop.style.display = 'block';
  position_popup(activator, pop);
}

/*
 * help_refine
 * 
 * Intended for links within a help popup. Stores a stack of state so
 * you can go back.
 */
function help_refine(popup_id) {
  if (help_popup.popup == null || help_popup.popup.length == 0 || help_popup.activator == null) {
    //throw new Error("Cannot refine a help popup when one is not shown!");
    var pop = lookup_help_popup(popup_id);
    var act_id = popup_id + '_act';
    var activator = document.getElementById(act_id);
    help_popup(activator, popup_id);
  }
  var pop = lookup_help_popup(popup_id);
  var last_pop = help_popup.popup[help_popup.popup.length - 1];
  if (pop == last_pop) return; // slightly odd, but no real cause for alarm
  help_popup.popup.push(pop);
  toggle_class(pop, "nested", true);
  last_pop.style.display = "none";
  //must make the popup visible to measure it or it has zero width
  pop.style.display = "block";
  position_popup(help_popup.activator, pop);
}

/*
 * help_return
 * 
 * Intended for links within a help popup. Stores a stack of state so
 * you can go back.
 */
function help_return() {
  if (help_popup.popup == null || help_popup.popup.length == 0 || help_popup.activator == null) {
    throw new Error("Can not return to a earlier help popup when one is not shown!");
  }
  var last_pop = help_popup.popup.pop();
  last_pop.style.display = "none";
  var pop = (help_popup.popup.length > 0 ? help_popup.popup[help_popup.popup.length - 1] : null);
  if (pop != null) {
    toggle_class(pop, "nested", help_popup.popup.length > 1);
    pop.style.display = "block";
    position_popup(help_popup.activator, pop);
  } else {
    help_popup.activator = null;
  }
}

/*
 * update_scroll_pad
 *
 * Creates padding at the bottom of the page to allow
 * scrolling of anything into view.
 */
function update_scroll_pad() {
  var page, pad;
  page = (document.compatMode === "CSS1Compat") ? document.documentElement : document.body;
  pad = $("scrollpad");
  if (pad === null) {
    pad = document.createElement("div");
    pad.id = 'scrollpad';
    document.getElementsByTagName('body')[0].appendChild(pad);
  }
  pad.style.height = Math.abs(page.clientHeight - 100) + "px";
}

function substitute_classes(node, remove, add) {
  "use strict";
  var list, all, i, cls, classes;
  list = node.className.split(/\s+/);
  all = {};
  for (i = 0; i < list.length; i++) {
    if (list[i].length > 0) all[list[i]] = true;
  }
  for (i = 0; i < remove.length; i++) {
    if (all.hasOwnProperty(remove[i])) {
      delete all[remove[i]];
    }
  }
  for (i = 0; i < add.length; i++) {
    all[add[i]] = true;
  }
  classes = "";
  for (cls in all) {
    classes += cls + " ";
  }
  node.className = classes;
}

/*
 * toggle_class
 *
 * Adds or removes a class from the node. If the parameter 'enabled' is not 
 * passed then the existence of the class will be toggled, otherwise it will be
 * included if enabled is true.
 */
function toggle_class(node, cls, enabled) {
  var classes = node.className;
  var list = classes.replace(/^\s+/, '').replace(/\s+$/, '').split(/\s+/);
  var found = false;
  for (var i = 0; i < list.length; i++) {
    if (list[i] == cls) {
      list.splice(i, 1);
      i--;
      found = true;
    }
  }
  if (typeof enabled == "undefined") {
    if (!found) list.push(cls);
  } else {
    if (enabled) list.push(cls);
  }
  node.className = list.join(" ");
}

/*
 * find_child
 *
 * Searches child nodes in depth first order and returns the first it finds
 * with the className specified.
 * TODO replace with querySelector
 */
function find_child(node, className) {
  var pattern;
  if (node == null || typeof node !== "object") {
    return null;
  }
  if (typeof className === "string") {
    pattern = new RegExp("\\b" + className + "\\b");
  } else {
    pattern = className;
  }
  if (node.nodeType == Node.ELEMENT_NODE && 
      pattern.test(node.className)) {
    return node;
  } else {
    var result = null;
    for (var i = 0; i < node.childNodes.length; i++) {
      result = find_child(node.childNodes[i], pattern);
      if (result != null) break;
    }
    return result;
  }
}

/*
 * find_parent
 *
 * Searches parent nodes outwards from the node and returns the first it finds
 * with the className specified.
 */
function find_parent(node, className) {
  var pattern;
  pattern = new RegExp("\\b" + className + "\\b");
  do {
    if (node.nodeType == Node.ELEMENT_NODE && 
        pattern.test(node.className)) {
      return node;
    }
  } while (node = node.parentNode);
  return null;
}

/*
 * find_parent_tag
 *
 * Searches parent nodes outwards from the node and returns the first it finds
 * with the tag name specified. HTML tags should be specified in upper case.
 */
function find_parent_tag(node, tag_name) {
  do {
    if (node.nodeType == Node.ELEMENT_NODE && node.tagName == tag_name) {
      return node;
    }
  } while (node = node.parentNode);
  return null;
}

/*
 * __toggle_help
 *
 * Uses the 'topic' property of the this object to
 * toggle display of a help topic.
 *
 * This function is not intended to be called directly.
 */
function __toggle_help(e) {
  if (!e) e = window.event;
  if (e.type === "keydown") {
    if (e.keyCode !== 13 && e.keyCode !== 32) {
      return;
    }
    // stop a submit or something like that
    e.preventDefault();
  }

  help_popup(this, this.getAttribute("data-topic"));
}

function setup_help_button(button) {
  "use strict";
  var topic;
  if (button.hasAttribute("data-topic")) {
    topic = button.getAttribute("data-topic");
    if (document.getElementById(topic) != null) {
      button.tabIndex = "0"; // make keyboard selectable
      button.addEventListener("click", function() {
        help_popup(button, topic);
      }, false);
      button.addEventListener("keydown", function(e) {
        // toggle only on Enter or Spacebar, let other keys do their thing
        if (e.keyCode !== 13 && e.keyCode !== 32) return;
        // stop a submit or something like that
        e.preventDefault();
        help_popup(button, topic);
      }, false);
    } else {
      button.style.visibility = "hidden";
    }
  }
  button.className += " active";
}

/*
 * help_button
 *
 * Makes a help button for the passed topic.
 */
function help_button(topic) {
  var btn = document.createElement("div");
  btn.className = "help";
  btn.setAttribute("data-topic", topic);
  setup_help_button(btn);
  return btn;
}

/*
 * prepare_download
 *
 * Sets the attributes of a link to setup a file download using the given content.
 * If no link is provided then create one and click it.
 */
function prepare_download(content, mimetype, filename, link) {
  "use strict";
  // if no link is provided then create one and click it
  var click_link = false;
  if (!link) {
    link = document.createElement("a");
    click_link = true;
  }
  try {
    // Use a BLOB to convert the text into a data URL.
    // We could do this manually with a base 64 conversion.
    // This will only be supported on modern browsers,
    // hence the try block.
    var blob = new Blob([content], {type: mimetype});
    var reader = new FileReader();
    reader.onloadend = function() {
      // If we're lucky the browser will also support the download
      // attribute which will let us suggest a file name to save the link.
      // Otherwise it is likely that the filename will be unintelligible. 
      link.setAttribute("download", filename);
      link.href = reader.result;
      if (click_link) {
        // must add the link to click it
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
      }
    }
    reader.readAsDataURL(blob);
  } catch (error) {
    if (console && console.log) console.log(error);
    // probably an old browser
    link.href = "";
    link.visible = false;
  }
}

/*
 * add_cell
 *
 * Add a cell to the table row.
 */
function add_cell(row, node, cls, click_action) {
  var cell = row.insertCell(row.cells.length);
  if (node) cell.appendChild(node);
  if (cls && cls !== "") cell.className = cls;
  if (click_action) cell.addEventListener("click", click_action, false);
}

/*
 * add_header_cell
 *
 * Add a header cell to the table row.
 */
function add_header_cell(row, node, help_topic, cls, colspan, is_new) {
  var th = document.createElement("th");
  if (node) th.appendChild(node);
  if (help_topic && help_topic !== "") th.appendChild(help_button(help_topic));
  if (is_new && is_new !== "") {
    var br = document.createElement("span");
    br.innerHTML = "<br>";
    th.appendChild(br);
    var new_icon = document.createElement("img");
    new_icon.src = new_icon_src;
    new_icon.alt = "NEW";
    th.appendChild(new_icon);
  }
  if (cls && cls !== "") th.className = cls;
  if (typeof colspan == "number" && colspan > 1) th.colSpan = colspan;
  row.appendChild(th);
}

/*
 * add_text_cell
 *
 * Add a text cell to the table row.
 */
function add_text_cell(row, text, cls, action) {
  var node = null;
  if (typeof(text) != 'undefined') node = document.createTextNode(text);
  add_cell(row, node, cls, action);
}

/*
 * add_text_header_cell
 *
 * Add a text header cell to the table row.
 */
function add_text_header_cell(row, text, help_topic, cls, action, colspan, is_new) {
  var node = null;
  if (typeof(text) != 'undefined') {
    var nbsp = (help_topic ? "\u00A0" : "");
    var str = "" + text;
    var parts = str.split(/\n/);
    if (parts.length === 1) {
      if (action) {
        node = document.createElement("span");
        node.appendChild(document.createTextNode(str + nbsp));
      } else {
        node = document.createTextNode(str + nbsp);
      }
    } else {
      node = document.createElement("span");
      for (var i = 0; i < parts.length; i++) {
        if (i !== 0) {
          node.appendChild(document.createElement("br"));
        }
        node.appendChild(document.createTextNode(parts[i]));
      }
    }
    if (action) {
      node.addEventListener("click", action, false);
      node.style.cursor = "pointer";
    }
  }
  add_header_cell(row, node, help_topic, cls, colspan, is_new);
}

function setup_help() {
  "use strict";
  var help_buttons, i;
  help_buttons = document.querySelectorAll(".help:not(.active)");
  for (i = 0; i < help_buttons.length; i++) {
    setup_help_button(help_buttons[i]);
  }
}

function setup_scrollpad() {
  "use strict";
  if (document.getElementsByTagName('body')[0].hasAttribute("data-scrollpad") && document.getElementById("scrollpad") == null) {
    window.addEventListener("resize", update_scroll_pad, false);
    update_scroll_pad();
  }
}

// anon function to avoid polluting global scope
(function() {
  "use strict";
  window.addEventListener("load", function load(evt) {
    window.removeEventListener("load", load, false);
    setup_help();
    setup_scrollpad();
  }, false);
})();

/*
 *  make_link
 *
 *  Creates a text node and if a URL is specified it surrounds it with a link.
 *  If the URL doesn't begin with "http://" it automatically adds it, as
 *  relative links don't make much sense in this context.
 */
function make_link(text, url) {
  var textNode = null;
  var link = null;
  if (typeof text !== "undefined" && text !== null) textNode = document.createTextNode(text);
  if (typeof url === "string") {
    if (url.indexOf("//") == -1) {
      url = "http://" + url;
    }
    link = document.createElement('a');
    link.href = url;
    if (textNode) link.appendChild(textNode);
    return link;
  }
  return textNode;
}

//
// Function to create an HTML paragraph describing the 
// MEME Suite background model source.
//
function make_background_source(title, source, text) {
  var paraNode = document.createElement("P");
  var titleNode = document.createElement("B");
  var textNode1 = document.createTextNode("\u00A0\u00A0\u00A0\u00A0" + title + ": ");
  titleNode.appendChild(textNode1);
  var source_text = ((source == "--motif--") ? "the (first) motif file" : (source == "--nrdb--") ? "an old version of the NCBI non-redundant database" : (source == "--uniform--") ? "the uniform model" : (source == "--query--") ? "the query file" : (source == "--sequences--") ? "built from the (primary) sequences" : (source == "--control--") ? "built from the control (negative) sequences" : ((source == "--negatives--") ? "built from the negative (control) sequences" : "the file '" + source + "'"));
  if (text) { return source_text; }
  var textNode2 = document.createTextNode(source_text);
  paraNode.appendChild(titleNode);
  paraNode.appendChild(textNode2);
  return paraNode;
}

// Function to create a help button
function make_help_button(container, help_topic) {
  container.appendChild(help_button(help_topic));
}

// Function to toggle display.
function change_display(id) {
  var element=document.getElementById(id);
  element.style.display=(element.style.display=='none') ? element.style.display='inline' : element.style.display='none';
}
    /* END INCLUDED FILE "utilities.js" */
    
  
    
    /* START INCLUDED FILE "simple-shared-doc.js" */
//
// simple-shared-doc.js
//

//
// Function to redirect to appropriate doc file.
//
function get_doc_text(pgm, doc_type, extra, extra2) {
  switch (pgm) {
    case 'shared':
      return(get_shared_doc_text(doc_type, extra, extra2));
    case 'ame':
      return(get_ame_doc_text(doc_type, extra, extra2));
    case 'centrimo':
      return(get_centrimo_doc_text(doc_type, extra, extra2));
    case 'discovery':
      return(get_discovery_doc_text(doc_type, extra, extra2));
    case 'fimo':
      return(get_fimo_doc_text(doc_type, extra, extra2));
    case 'gomo':
      return(get_gomo_doc_text(doc_type, extra, extra2));
    case 'mcast':
      return(get_mcast_doc_text(doc_type, extra, extra2));
    case 'meme-chip':
      return(get_meme_chip_doc_text(doc_type, extra, extra2));
    case 'momo':
      return(get_momo_doc_text(doc_type, extra, extra2));
    case 'sea':
      return(get_sea_doc_text(doc_type, extra, extra2));
    case 'spamo':
      return(get_spamo_doc_text(doc_type, extra, extra2));
    case 'streme':
      return(get_streme_doc_text(doc_type, extra, extra2));
    case 'tgene':
      return(get_tgene_doc_text(doc_type, extra, extra2));
    case 'tomtom':
      return(get_tomtom_doc_text(doc_type, extra, extra2));
    case 'xstreme':
      return(get_xstreme_doc_text(doc_type, extra, extra2));
    default:
      return("<b>Unknown program type: <font color=red>" + pgm + "</font></b>");
  }
} // get_doc_text

//
// Function to replace the innerHTML of element "id" with the HTML indicated by "doc_type".
// Easier to read and update than the more flexible approach in shared-doc.js. 
//
function print_doc(id, pgm, doc_type, extra) {
  document.getElementById(id).insertAdjacentHTML('beforeend', get_doc_text(pgm, doc_type, extra));
} // print_doc

//
// Function to replace the innerHTML of element "id" with an HTML paragraph
// containing the text for 'pgm' and 'doc_type'.
// This function can be used in help pop-ups.
//
function print_doc_para(id, pgm, doc_type, extra, extra2) {
  html = "<p>" + get_doc_text(pgm, doc_type, extra, extra2) + "</p>"; 
  document.getElementById(id).insertAdjacentHTML('beforeend', html);
} // print_doc_para

//
// Function to return the Shared HTML text of a given type.
// This function can be used directly to document the output format (xx-output-format.html)
// and indirectly via print_doc_para for help pop-ups in the actual output HTML,
// to prevent duplication of documentation.
//
function get_shared_doc_text(doc_type, extra, extra2) {
  if (extra == undefined) {extra = ""};
  if (extra2 == undefined) {extra2 = ""};
  switch (doc_type) {
    case 'motif-db':
      return(`
	The name of ` + extra2 + ` a file of motifs ("motif database file") that contains ` + extra + `
      `);
    case 'motif-id':
      return(`
	The name of the ` + extra + ` motif, which is unique ` + extra2 + ` in the motif database file.
      `);
    case 'motif-alt-id':
      return(`
	An alternate name for the ` + extra + ` motif that may be provided ` + extra2 + ` in the motif database file.
      `);
    case 'motif-width':
      return(`
	The width of the motif. No gaps are allowed in motifs supplied to ` + extra + `
        as it only works for motifs of a fixed width.
      `);
    case 'motif-cons':
      return(`
	A consensus sequence computed from the ` + extra + ` motif (as described <a href="#consensus_doc">below</a>).
      `);
    case 'motif-match-score':
     return(`
	` + extra2 + ` The motif match score of a position in a sequence is
	computed by summing the appropriate entry from each column of the
	position-dependent scoring matrix that represents the motif. ` + extra + `
     `);
    case 'motif-match-p-value':
      return(`
	The <i>p</i>-value of a motif match is the probability of a single random
	subsequence of the length of the motif
	<a href="javascript:help_refine('pop_motif_match_score')">scoring</a>
	at least as well as the observed match.
      `);
    case 'bh-q-value':
      if (extra2 == "") extra2 = "match";
      return(`
	The q-value is the minimum False Discovery Rate (FDR) required to consider this
        ` + extra2 + ` significant.</br>` +
        get_shared_doc_text('bh-q-value-method', extra, extra2) + `
      `);
    case 'bh-q-value-method':
      return(`
        <br>` + extra + ` estimates q-values from all the ` + extra2 + ` <i>p</i>-values 
	using the method proposed by Benjamini & Hochberg (<i>Journal of the Royal Statistical Society B</i>, 57:289-300, 1995).
	See also Storey JD, Tibshirani R. Statistical significance for
	genome-wide studies, <i>Proc. Natl. Acad. Sci. USA</i> (2003) <b>100</b>:9440&ndash;9445.
      `);
    case 'sdb-name':
      return(`
	The name of the (FASTA) sequence database file.
      `);
    case 'sdb-psp':
      return(`
	The name of the position specific priors (PSP) file.
      `);
    case 'sdb-dist':
      return(`
	The name of the binned distribution of priors file.
      `);
    case 'sdb-count':
      return(`
	The number of sequences in the database.
      `);
    case 'sdb-letters':
      return(`
	The number of letters in the sequence database.
      `);
    case 'lastmod':
      return(`
	The date of the last modification to the ` + extra + ` database.
      `);
    case 'sequence-id':
      return(`
        The identifier of the sequence (from the FASTA sequence header line)` + extra + `
      `);
    case 'sequence-name':
      return(`
	` + extra + `name of the sequence extracted from the sequence identifier (in the FASTA sequence header line).<br>
	When you use the <code>--parse-genomic--coord</code> option, the sequence name ends at the
	first colon ':' (if any) present in the sequence\'s FASTA identifier.  Typically this is the
	chromosome or contig name.  With the <code>--parse-genomic--coord</code> option,
	the start and stop positions are in 0-based coordinates relative to the sequence start given 
	in the FASTA sequence identifier (just after the sequence name).</td> </tr>
      `);
    case 'sequence-desc':
      return(`
        The description appearing after the identifier of the sequence in the FASTA header line.
      `);
    case 'sequence-name':
    case 'alph-name':
      return(`
	The name of the alphabet symbol.
      `);
    case 'alph-bg':
      return(`
	The frequency of the alphabet symbol as defined by the background model.
      `);
    case 'match-start-seq':
      return(`
	The start position of the ` + extra + `; 1-based sequence coordinates.
      `);
    case 'match-stop-seq':
      return(`
	The end position of the ` + extra + `; 1-based sequence coordinates.
      `);
    case 'match-start-genomic':
      return(`
	The start position of the ` + extra + `; genomic coordinates.
      `);
    case 'match-stop-genomic':
      return(`
	The end position of the ` + extra + `; genomic coordinates.
      `);
    case 'parse-genomic-coord':
      return(`
	` + extra + ` was run with the <code>--parse-genomic-coord</code> option
	and has split the sequence identifier into sequence name, sequence start and sequence end 
	in genomic coordinates.
      `);
    case 'motif-consensus':
      return(`
        <p id="consensus_doc">
           A <b>consensus sequence</b> is constructed from each column in a
           motif's frequency matrix using the <b>"50% rule"</b>
           as follows:
        </p>
        <ol>
          <li>The letter frequencies in the column are sorted in decreasing order.</li>
          <li>Letters with frequency less 50% of the maximum are discarded.</li>
          <li>The letter used in this position in the consensus sequence is determined
          by the first rule below that applies:</li>
          <ul>
            <li>If there is only one letter left, or if the remaining letters exactly match
            an ambiguous symbol in the alphabet, the <b>letter</b> or <b>ambiguous symbol</b>,
            respectively, is used.</li>
            <li>Otherwise, if the remaining set contains at least 50% of the core
            symbols in the alphabet, the alphabet's <b>wildcard</b>
            (e.g., "N" for DNA or RNA, and "X" for protein) is used.</li>
            <li>Otherwise, the letter with the <b>maximum frequency</b> is used.</li>
          </ul>
        </ol>
      `);
    default:
      return("Error--Unrecognized shared doc_type: " + doc_type);
  }
} // get_shared_doc_text
    /* END INCLUDED FILE "simple-shared-doc.js" */
    
  
    
    /* START INCLUDED FILE "gomo_doc.js" */
//
// gomo_doc.js
//

//
// Function to return the HTML text of a given type.
// This function can be used directly to document the output format (xx-output-format.html)
// and indirectly via print_doc_para for help pop-ups in the actual output HTML,
// to prevent duplication of documentation.
//
function get_gomo_doc_text(doc_type, extra) {
  var html;
  if (extra == undefined) {extra = ""};

  switch (doc_type) {
    case 'gomo-go-term':
      return(`
        The Gene Ontology Consortium term for a specific role or locality.
        Used for annotating genes with their functions.
      `);
    case 'gomo-score':
      return( `
        A score generated as the <a href="https://en.wikipedia.org/wiki/Geometric_mean">
        geometric mean</a> of <a href="https://en.wikipedia.org/wiki/Mann-Whitney_U_test">rank-sum test(s)</a>
        for the particular Gene Ontology term. The two groups compared by the rank-sum test are scores of genes annotated
        with the GO term and scores of genes not annotated with the GO term.</td> </tr>
      `);
    case 'gomo-p-value':
      return(`
        An empirically generated <i>p</i>-value for the enrichment of the GO term.<br>
        The null hypothesis is that by shuffling the labels on gene scores,
        any possible association between the set of genes that a GO term annotates is destroyed.
        A large number of scores are generated using the null hypothesis and the number of null
        hypothesis scores that are better than each of the real scores is summed and then divided
        by the total null hypothesis scores generated to calculate a <i>p</i>-value.</td> </tr>
      `);
    case 'gomo-results-tsv':
      return(`
        <p>
          GOMo outputs a tab-separated values (TSV) file ('gomo.tsv') that contains one line for each
          motif-GO-term pair found to be significantly enriched.
          The lines are grouped by motif and sorted in order of decreasing statistical significance.
          The first line contains the (tab-separated) names of the fields.
          Your command line is given at the end of the file in a comment line starting with the character '#'.
          The names and meanings of each of the fields are described in the table below.
        </p>
        <table class="dark" style="width:100%" border=1>
          <tr> <th>field</th> <th>name</th> <th>contents</th> </tr>
          <tr> <td>1</td> <td>Motif_Identifier</td> <td> ` + get_doc_text('shared', 'motif-id') + ` </td> </tr>
          <tr> <td>2</td> <td>GO_Term_Identifier</td> <td> ` + get_gomo_doc_text('gomo-go-term') + ` </td> </tr>
          <tr> <td>3</td> <td>GOMo_Score</td> <td> ` + get_gomo_doc_text('gomo-score') + ` </td> </tr>
          <tr> <td>4</td> <td>p-value</td> <td> ` + get_gomo_doc_text('gomo-p-value') + ` </td> </tr>
          <tr> <td>5</td> <td>q-value</td> <td> ` + get_doc_text('shared', 'bh-q-value', 'GOMo') + ` </td> </tr>
        </table>
      `);
    case 'gomo-results-xml':
      return(`
        <p>
        GOMo outputs an XML file ('gomo.xml') with the following format.
        </p>
        <table class="bordertable" border="1">
          <tr>
            <th>Tag</th><th>Child of</th><th>Description</th>
          </tr>
          <tr>
            <td >&lt;gomo&gt;</td><td >Nothing</td>
            <td>
              Information about this run of GOMo.
              <ul>
                <li>version - The version of GOMo that generated the XML file.</li>
                <li>release - The release date of the version that generated the XML.</li>
              </ul>
            </td>
          </tr>
          <tr>
            <td >&lt;program&gt;</td>
            <td >&lt;gomo&gt;</td>
            <td>
              Information about the state of the program when it ran.<br />
              <ul>
                <li>name - name of the program.</li>
                <li>cmd - the command line passed to the program.</li>
                <li>gene_url - the url used to lookup further information on the gene ids.
                The url has ampersands (&amp;) converted into <b>&amp;amp;</b> and the place where
                  the gene ID should be replaced by <b>!!GENEID!!</b> .</li>
                <li>outdir - the output directory that the program wrote to.</li>
                <li>clobber - true if GOMo was allowed to overwrite the output directory.</li>
                <li>text_only - true if GOMo wrote to stdout, in which case this file would
                  not exist so it must be false.</li>
                <li>use_e_values - true if GOMo used <i>E</i>-values (converted from <i>p</i>-values) as
                  input scores, false if GOMo used gene scores.</li>
                <li>score_e_thresh - if GOMo used <i>E</i>-values then this is the threshold that
                  GOMo assumed the worst <i>E</i>-value (<i>p</i>-value = 1.0) for the gene to smooth out noise.</li>
                <li>min_gene_count - the minimum number of genes that a GO term was annotated
                  with before GOMo would calculate a score for it.</li>
                <li>motifs - if present then a space delimited list of the motifs that GOMo
                  calculated a score for, otherwise GOMo scored all motifs.</li>
                <li>shuffle_scores - the number of times GOMo generated a shuffled mapping of
                  gene id to gene id to be used to generate scores from the null model.</li>
                <li>q_threshold - GOMo filtered the results to only show those with a better
                (smaller) q-value.</li>
              </ul>
            </td>
          </tr>
          <tr>
            <td>&lt;gomapfile&gt;</td>
            <td>&lt;program&gt;</td>
            <td>
              Information about the GO mapping file.<br />
              <ul>
                <li>path - the path to the mapping file.</li>
              </ul>
            </td>
          </tr>
          <tr>
            <td>&lt;seqscorefile&gt;</td>
            <td>&lt;program&gt;</td>
            <td>
              Information about the sequence scoring file.<br />
              <ul>
                <li>path - the path to the sequence scoring file.</li>
              </ul>
            </td>
          </tr>
          <tr>
            <td>&lt;motif&gt;</td>
            <td>&lt;gomo&gt;</td>
            <td>
              Information about the motif.<br />
              <ul>
                <li>id - the motif identifier.</li>
                <li>genecount - the number of scored sequences that were used to compute the result.</li>
              </ul>
            </td>
          </tr>
          <tr>
            <td>&lt;goterm&gt;</td>
            <td>&lt;motif&gt;</td>
            <td>
              Information about the GO term.<br />
              <ul>
                <li>id - the GO identifier.</li>
                <li>score - the geometric mean across all species of the rank-sum test <i>p</i>-value.</li>
                <li>pvalue - the empirically calculated <i>p</i>-value.</li>
                <li>qvalue - the empirically calculated q-value.</li>
                <li>annotated - the number of genes annotated with the go term.</li>
                <li>group - the subgroup that the term belongs to. For the Gene Ontology
                    b = biological process, c = cellular component and m = molecular function.</li>
                <li>nabove - the number of more general terms that link to this one.</li>
                <li>nbelow - the number of more specific terms that link from this one.</li>
                <li>implied - is the go term implied by other significant go terms?
                  Allows values 'y', 'n' or 'u' (default) for yes, no or unknown.</li>
                <li>description - the GO term description.</li>
              </ul>
            </td>
          </tr>
          <tr>
            <td>&lt;gene&gt;</td>
            <td>&lt;goterm&gt;</td>
            <td>
              Information about the GO term's annotated genes for the primary species.<br />
              <ul>
                <li>id - the gene identifier.</li>
                <li>rank - the rank of the scored gene.</li>
              </ul>
            </td>
          </tr>
        </table>
      `);
    default:
      return("Error--Unrecognized doc_type: " + doc_type);
  }
} // get_gomo_doc_text
    /* END INCLUDED FILE "gomo_doc.js" */
    
  
    
    /* START INCLUDED FILE "citation.js" */
//
// citation.js
//
function get_citation_text(doc_type, extra) {
  var html;

  switch (doc_type) {
    case 'AMA':
      return(get_citation_text("GOMo", extra));
    case 'AME':
      return(extra + `
        <span class="citation">
          Robert C. McLeay and Timothy L. Bailey,
          &quot;Motif Enrichment Analysis: a unified framework and an evaluation on ChIP data&quot;,
          <i>BMC Bioinformatics</i>, <b>11</b>:165, 2010.
          <a href="http://www.biomedcentral.com/1471-2105/11/165">[full text]</a>
        </span>
      `);
    case 'CentriMo':
      return(extra + `
        <span class="citation">
          Timothy L. Bailey and Philip Machanick,
          &quot;Inferring direct DNA binding from ChIP-seq&quot;,
          <i>Nucleic Acids Research</i>, <b>40</b>:e128, 2012.
          <a href="http://nar.oxfordjournals.org/content/40/17/e128">[Full Text]</a>
        </span>
      `);
    case 'DREME':
      return(extra + `
        <span class="citation">
          Timothy L. Bailey,
          &quot;DREME: Motif discovery in transcription factor ChIP-seq data&quot;,
          <i>Bioinformatics</i>, <b>27</b>(12):1653-1659, 2011.
          <a href="http://bioinformatics.oxfordjournals.org/content/27/12/1653">[full text]</a>
        </span>
      `);
    case 'FIMO':
      return(extra + `
        <span class="citation">
          Charles E. Grant, Timothy L. Bailey and William Stafford Noble,
          &quot;FIMO: Scanning for occurrences of a given motif&quot;,
          <i>Bioinformatics</i> <b>27</b>(7):1017-1018, 2011.
          <a href="http://bioinformatics.oxfordjournals.org/content/27/7/1017">[full text]</a>
        </span>
      `);
    case 'GLAM2':
    case 'GLAM2SCAN':
      return(extra + `
        <span class="citation">
          Martin C. Frith, Neil F. W. Saunders, Bostjan Kobe and Timothy L. Bailey,
          &quot;Discovering sequence motifs with arbitrary insertions and deletions&quot;,
          <i>PLoS Computational Biology</i>, <b>4</b>(5):e1000071, 2008.
          <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000071">[full text]</a>
        </span>
      `);
    case 'GOMo':
      return(extra + `
        <span class="citation">
          Fabian A. Buske, Mikael Bod&eacute;n, Denis C. Bauer and Timothy L. Bailey,
          &quot;Assigning roles to DNA regulatory motifs using comparative genomics&quot;,
          <i>Bioinformatics</i>, <b>26</b>(7), 860-866, 2010.
          <a href="http://bioinformatics.oxfordjournals.org/cgi/content/full/26/7/860">[full text]</a>
        </span>
      `);
    case 'MAST':
      return(extra + `
        <span class="citation">
          Timothy L. Bailey and Michael Gribskov,
          &quot;Combining evidence using p-values: application to sequence homology searches&quot;,
          <i>Bioinformatics</i>, <b>14</b>(1):48-54, 1998.
          <a href="http://bioinformatics.oxfordjournals.org/content/14/1/48">[full text]</a>
        </span>
      `);
    case 'MCAST':
      return(extra + `
        <span class="citation">
          Timothy Bailey and William Stafford Noble,
          &quot;Searching for statistically significant regulatory modules&quot;,
          <i>Bioinformatics (Proceedings of the European Conference on Computational Biology)</i>,
          <b>19</b>(Suppl. 2):ii16-ii25, 2003.
          <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/suppl_2/ii16">[full text]</a>
        </span>
      `);
    case 'Meta-MEME':
      return(extra + `
        <span class="citation">
	  William N. Grundy, Timothy L. Bailey, Charles P. Elkan and Michael E. Baker.
	  &quot;Meta-MEME: Motif-based Hidden Markov Models of Protein Families&quot;
	  <i>Computer Applications in the Biological Sciences (CABIOS)</i>,
	  <b>13</b>(4):397-406, 1997.
	  <a href="http://bioinformatics.oxfordjournals.org/content/13/4/397">[full text]</a>
        </span>
      `);
    case 'MEME':
      return(extra + `
        <span class="citation">
          Timothy L. Bailey and Charles Elkan,
          &quot;Fitting a mixture model by expectation maximization to
          discover motifs in biopolymers&quot;,
          <em>Proceedings of the Second International Conference on Intelligent Systems
          for Molecular Biology</em>, pp. 28-36, AAAI Press, Menlo Park, California, 1994.
          <a href="http://www.aaai.org/Papers/ISMB/1994/ISMB94-004.pdf">[full text]</a>
        </span>
      `);
    case 'MEME-ChIP':
      return(extra + `
        <span class="citation">
          Philip Machanick and Timothy L. Bailey,
          &quot;MEME-ChIP: motif analysis of large DNA datasets&quot;,
          <i>Bioinformatics</i> <b>27</b>(12):1696-1697, 2011.
        <a href="http://bioinformatics.oxfordjournals.org/content/27/12/1696.full">[full text]</a>
        </span>
      `);
    case 'MEME_SUITE':
      return(extra + `
        <span class="citation">
          Timothy L. Bailey, James Johnson, Charles E. Grant, William S. Noble,
          &quot;The MEME Suite&quot;,
          <i>Nucleic Acids Research</i>, <b>43</b>(W1):W39-W49, 2015.
          <a href="https://academic.oup.com/nar/article/43/W1/W39/2467905">[full text]</a>
        </span>
      `);
    case 'MoMo':
      return(extra + `
        <span class="citation">
          Alice Cheng, Charles Grant, Timothy L. Bailey and William Noble,
          &quot;MoMo: Discovery of statistically significant post-translational modification motifs&quot;, 
          <i>Bioinformatics</i>, <b>35</b>(16):2774-2782, 2018.
          <a href="https://doi.org/10.1093/bioinformatics/bty1058">[full text]</a>
        </span>
      `);
    case 'PSPs':
      return(extra + `
        <span class="citation">
          Timothy L. Bailey, Mikael Bod&eacute;n, Tom Whitington and Philip Machanick,
          &quot;The value of position-specific priors in motif discovery using MEME&quot;,
          <i>BMC Bioinformatics</i>, <b>11</b>(1):179, 2010.
          <a href="http://www.biomedcentral.com/1471-2105/11/179">[full text]</a>
        </span>
      `);
    case 'SEA':
      return(extra + `
        <span class="citation">
          Timothy L. Bailey, &quot;SEA: Simple Enrichment Analysis of motifs&quot;,
          <i>BioRxiv</i>, 2021.
          <a href="FIXME">[full text]</a>
        </span>
      `);
    case 'SpaMo':
      return(extra + `
        <span class="citation">
          Tom Whitington, Martin C. Frith, James Johnson and Timothy L. Bailey
          &quot;Inferring transcription factor complexes from ChIP-seq data&quot;,
          <i>Nucleic Acids Res.</i> <b>39</b>(15):e98, 2011.
          <a href="http://nar.oxfordjournals.org/content/39/15/e98">[full text]</a>
        </span>
      `);
    case 'STREME':
      return(extra + `
        <span class="citation">
          Timothy L. Bailey,
          &quot;STREME: accurate and versatile sequence motif discovery&quot;,
          <i>Bioinformatics</i>, Mar. 24, 2021.
          <a href="https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btab203/6184861" >[full text]</a>
        </span>
      `);
    case 'Tomtom':
      return(extra + `
        <span class="citation">
          Shobhit Gupta, JA Stamatoyannopolous, Timothy Bailey and William Stafford Noble,
          &quot;Quantifying similarity between motifs&quot;,
          <i>Genome Biology</i>, <b>8</b>(2):R24, 2007.
          <a href="http://genomebiology.com/2007/8/2/R24">[full text]</a>
        </span>
      `);
    case 'T-Gene':
      return(extra + `
        <span class="citation">
          Timothy O'Connor, Charles E. Grant, Mikael Bod&eacute;n, Timothy L. Bailey,
          &quot;T-Gene: Improved target gene prediction&quot;,
            <i>Bioinformatics</i>, <b>36</b>(12):3902-3904, 2020.
          <a href="https://academic.oup.com/bioinformatics/article/36/12/3902/5815978?guestAccessKey=aa625a49-a2aa-4d7a-858e-8bc82867a534">[Full Text]</a>
        </span>
      `);
    case 'XSTREME':
      return(extra + `
        <span class="citation">
          Charles E. Grant and Timothy L. Bailey, &quot;XSTREME: comprehensive motif analysis of biological sequence datasets&quot;,
          <i>BioRxiv</i>, 2021.
          <a href="FIXME">[full text]</a>
        </span>
      `);
    default:
      return("Unknown program: " + doc_type);
  }
} // get_citation_text

//
// Function to replace the innerHTML of element "id" with an HTML paragraph
// containing the text for 'program', which is known to function get_citation_text.
// If "id" is either "citation" or "reference" some extra text is printed.
//
function print_citation(id, program) {
  var extra;
  switch (id) {
    case 'citation':
      extra = "If you use " + program + " in your research, please cite the following paper:<br>";
      break;
    case 'reference':
      extra = "<h5>Reference</h5>";
      break;
    default:
      extra = "";
      break;
  };
  var html = get_citation_text(program, extra);
  document.getElementById(id).insertAdjacentHTML('beforeend', html);
} // print_citation

// 
// Function to convert a citation for a program to a C #define statement.
//
function print_citation_define(lang, pgm) {
  var citation = get_citation_text(pgm, '');
  citation = citation.replace(/<[^>]*>/g, '');
  citation = citation.replace(/\[.*\]/g, '');
  citation = citation.replace(/\n\s*/g, '\\n');
  citation = citation.replace(/&quot;/g, '\\"');
  citation = citation.replace(/&eacute;/g, 'e');
  citation = citation.replace(/^\\n/, '');
  pgm = pgm.replace(/-/, '');
  citation = "If you use this program in your research, please cite:\\n\\n" + citation;
  if (lang == "C") {
    citation = "#define " + pgm + '_CITE "' + citation + '"';
  } else if (lang == "perl") {
    citation = '"' + pgm + '" => "' + citation + '",';
  }
  return(citation);
} // print_citation_define

//
// Main program (for use with nodejs "node" javascript engine)
// to create citation.js.h and citation.pm from citation.js.
// The command line:
//   node citation.js C > citation.js.h
// will output the C #define statements for each of the
// programs listed below, defining macros <program>_CITE.
// The command line:
//   node citation.js perl > citation.js.pm
// will output perl hash <program> => text
//
//
if (typeof process !== 'undefined') {
  var lang = process.argv[2];
  var programs = ['AMA', 'AME', 'CentriMo', 'DREME', 'FIMO', 'GLAM2', 
    'GLAM2SCAN', 'GOMo', 'MAST', 'MCAST', 'Meta-MEME', 'MEME',
    'MEME-ChIP', 'MEME_SUITE', 'MoMo', 'PSPs', 'SEA', 'SpaMo',
    'STREME', 'Tomtom', 'T-Gene', 'XSTREME'];

  if (lang == "C") {
    console.log("// Do not edit this file.  It is created from etc/citation.js.");
    console.log("#ifndef citation_js_h\n#define citation_js_h\n");
    for (var i=0; i<programs.length; i++) {
      console.log(print_citation_define(lang, programs[i]));
    }
    console.log("\n#endif");
  } else if (lang == "perl") {
    console.log("# Do not edit this file.  It is created from etc/citation.js.");
    console.log("package Citation;");
    console.log("sub cite {\n  my ($pgm) = @_;\n  return $citation{$pgm};\n}");
    console.log("%citation = (");
    for (var i=0; i<programs.length; i++) {
      console.log(print_citation_define(lang, programs[i]));
    }
    console.log(");");
  }
}
    /* END INCLUDED FILE "citation.js" */
    
  </script>
</head>
<body bgcolor="#ffffff">
<div class="pop_content" id="pop_results_tsv">
<script>print_doc("pop_results_tsv", "gomo", "gomo-results-tsv");</script><div style="float:right; bottom:0px;">[
        <a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_results_xml">
<script>print_doc("pop_results_xml", "gomo", "gomo-results-xml");</script><div style="float:right; bottom:0px;">[
        <a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_motif">
<script type="text/javascript">print_doc_para('pop_motif', 'shared', 'motif-id');</script><div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_logo">
<p>The visual representation of the position weight matrix otherwise known as the motif logo.</p>
<div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_predictions">
<p>The number of term predictions.</p>
<div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_bestpred">
<p>If specificity data is avaliable then this displays the top 5 specific predictions, otherwise it just
        displays the top 5 predictions.</p>
<div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_goterm">
<script type="text/javascript">print_doc_para('pop_goterm', 'gomo', 'gomo-go-term');</script><div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_gscore">
<script type="text/javascript">print_doc_para('pop_gscore', 'gomo', 'gomo-score');</script><div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_epvalue">
<script type="text/javascript">print_doc_para('pop_epvalue', 'gomo', 'gomo-p-value');</script><div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_eqvalue">
<script type="text/javascript">print_doc_para('pop_eqvalue', 'shared', 'bh-q-value', 'GOMo');</script><div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_specif">
<p>A measure of the relative position in the GO hierarchy showing how specific a term is. A value of 100% implies that
        the GO term is the most specific form whereas a value of 0% implies that the GO term is one of the three roots of
        the Gene Ontology with only terms below it. A tilda (~) infront of the percentage implies that the value shown has 
        been rounded and is not exact.</p>
<div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_goname">
<p>The Gene Ontology hierarchy and Gene Ontology name of the term. The hierarchy is specified by the abbreviations 
        <span class="bp">BP</span> for biological process,
        <span class="cc">CC</span> for cellular component and
        <span class="mf">MF</span> for molecular function.</p>
<div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pop_content" id="pop_generank">
<p>Gene name and rank of (up to) top 10 genes annotated with the GO term for the primary species.</p>
<div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
</div>
<div class="pad1">
<div class="prog_logo big">
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEgAAABICAMAAABiM0N1AAAAAXNSR0IArs4c6QAAAwBQTFRFVWlpjY2Ng5eYcXJymNPh2dnZ3d3dvePsKWlr8PDwfHx8QJyqq8HC+Pj4ra2tgICAxebuoKCgzc3NaL/SlrS0dsXYPJilOHd4VXJzhamp9fX1MDAwuLi4laipxsbGVYeIjLzGc5qcWXd6ycnJO2NjeKGmdYWGNVlZkJCQ1OToU7XKR2ho8/Pz9PT0b56pR3d6KoKJqKio+fz97u7uRKOy0uzya4yMsLCw6vb5KFhZ6urq4vP3YGBgBkRF4ODgsMTEbGxsJXyC7Ozsvb6+HXJ2bJWV0NDQy+nwtbW14errZqy75eXlICAgSqq74uLiR4aSBzg56Ojo1tbWMYuUlZWVIHZ7wMDApaWlmpqa5ubmbsLWC1tbAAAAobq7nZ2dic3dYXZ2SYCCRlxd9fv8HWNjDFRV3fD11NTUN5Keg8rb/f7+qb6+s72+G1papNnlKkdHqtvnt5O2F0ZIh4eHtN/qW1xeUbLGqbOz2O70FWlrn7KyQEBAp8jQEVlatMnJdKm00tLSHlVV5vT4us3NGW5xbX5+zNHRw8PDutDSUFBRdnh4Ta7AEGNl8/r8FVVVgomJ7ff6KmBi0t3e8Pn7VpSXWKKzWYyQw8rLoKurzODla3JyUGFhDGBhjaCjn9fjE11eX319////RW9wkZqbU1paVrfOEmZo6fHzpL/CM05OLoeQy9fZBllZHGtsr93oYGtrwtHRk9HgQqCuB1VVjrO4JE9RXbLGu8fKl66ydo+QOmpuGzw9oKWluru7BlJSYb3SbmNu297ex9HRAVNTF05Q5uvsT09PCU1OwNXV/v7+/f39/Pz8+vr6+/v7/v//EBAQ/f7/9/z9+/3+xdTU3+XlzKPITI+fsLW1rri6R6a2Sai4hsPRX46VCB4g8PX1BVdYyc7OcmGAUkxUDRgaDFdYmp2eA1VWqqurf8DPj8/flcHKoc7ZLW5uNi41ZWdncrG/xdTYOFFTj5WV7u/vt8XIwMXFvsPDpba3mXqWgZGRZnh5TmxvisTSP1JTT0JOts7UsG2xTQAAAAF0Uk5TAEDm2GYAAAABYktHRACIBR1IAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAB3RJTUUH3AscATI7mY54oAAACGdJREFUWMPNmH9Qk/cdx3VFQ8hoQ7gEa53lilFMFR9iMiJpGI88TUAGRGKiQBRdjvQh04AWs2RIqblQOkK4QIplZwjGmhTCD49UFKVAvcLYA/I0PHd1Tdod161rZd3mlO06xtz2Tagr2AZ5+Gvfu1zyPPk+r3t/Pt/P9/P9fJ41axaGc7VjzSPj/x30G9Ectvg6rY+xKhDWiIioi29cmO5bFahF7nG/tVjQZ4ailtWAGlGWvHGxIAQy9a0C1AL7vVBJ2v+u/14lQ+GhNPKgPSxdAHV/Y8s2iMeTLbF1ZSAFkuMys5Ct+NfXW+U8nRk1XcDIgjagOk4Oimx5GAptqNmloyEXySpKa/Oac7wyw/zCJXUAopk5Lh7cRxY0B/FoqAwp2fZbacPYqS2XiyDUr3MFIBFZ0B6PDEY+a9zVwOJ52k7E5YkMHh7H5feQBYmr5Ab72bt3I3O87hvVM6XMvBIZLWCmwWRBifaS6bP/ckR60RPsnvJyR0SjwWSCZZCpkRzIt3XPLlF8+RVENvJTiYROobDfHxrZdbaovmiOHOh4M/PYRQr9IjzyN6FaredLdpW8qO0tzRsRKUiBsGOFM8+e4f+yoe1Nq9UqMArV99rpFdqIUubwuSX5wKmJaw89EAZ0/Oqx+TN8YevQz1Sq0dNPnO4UqvmSEKlweC4U6q+w2xdC3m6Iy8Kd498Nwtg3Ep3pQusP5wmCOB3jxKOMwDw6pcfRy4y9HZSEZ7mRBS2IDO5uck6GUZT0e4XzqTdVHTc1vpvKFKcz3SgMgUKSUkJT3kBReV4W+KbxPN1N0WFAjcpnnvpgkNuC474M5TMpUZWCoG0UYFsxc7g/OAP/EarzQ92vOPP8nICnezQMKE+pVH5AaHAnjmeAn8rT3wI5szw6s8w03ST3czgBKC6Mjz4pUyqjCB9wJy4Ogo5bgZNCILBuPwjNGZXTvHB9V5wJ5ek4rO5wPrqz9S3uzcrKoKQXlMo/WGzASUFQb2lhsziYDP745Rseminy2fZjJllOICyotVhvU3HTE3HcOTvLuMlV2QQhkGOmNDYPKN5RTbdyNfv+5CN8ml/D8upjkWFA4li11aL61Qs+fLZy1qd5COpxfO0itkRFYHj801NTKsvr9iu2tw+HAU0k6TvTKyt/8eP3/pzmxH3ElMVqDILA6t8G2b+z3AY42NvxFRHanvI4Cp1yJdxe+4piFPzl009PfvgecNPujFFVCKQF8XgdrEC8ekqDa6yOfezvd25/vT3+zp3msLu/lS/cfunkhyfTfbtrpCOiB1kAVOEoLmwGcT0bbyUwzNJbLX64NQuuhE9s8Xrhtd/949/rfefrIVRmuhcClcb2g0Oks8dG4BrKPurDAwaLzg0Lwpri+WrjR1lchkjOyjGjY9f09J4IZlAQfphu0eCq4sPfTI7OXuY4Erfy9ULrVIYIAedQDlIDQDOFTwdPtS/1Kh+mL8xdBNq/3EmLf9FaYbSMSushP8cMXVfTy2digxsWT9JzfRiledHUyb3L10fUj9i2qfsiO2rO8WxR03uKQynEmaTmYkT5x4smTsw+tmJrUXEZD9w0r6cvBArd6zcSGNfBXjRrfMdjQOKWOQvXlyaFWVDVV5KK3uaQX/oFBG6JSFpckkUv5+wNZ0+YTD8XTBGao/UwbBjop0QwW4P//Aesvm0JyHkrPIh6CuIFvHEVQguBzXZ0FTW8eIHiYJ4Lho7CpsGtS0ET4UHvIn6Xzp9XrrZxfSC9ZYDQppcXXw0Gs0KlwQSOJSBsmSIC8fpZbXd7+Fawr3AN1yJQBwOpdQGEGx1sfGVFhKIEgtyRMxV8gYoAqZsAIOBtJlgrfDvQ+Iii5QqtoyV26e0ICt8InOT0ESrrX+c2PLcv+Hg7iMeVK8IwsWKO6QAg4CQnTgxuk3Z1fS49DzLDYa5PY9Qmkaj8z5VqKXyhjQucVCOqR0yG+q4BsPn5XEwjjGCvHPRacwgEvO2smUYgmQdGDCLxzTPAaYR+pn/lIMVwqaNCoraqiIwxk4fmZckgub3OaOFqMIuk+BoZEDOih643WqaO2mGWWRdgoRBSZ5siMI0VpLWVg6i0wmJtBV8osGxxe3g6EJ80FK4D5xSh0vdeIdMdXR9mzpRT+GrBNoMMgFy6AM3zT71aqJb0VlPJ9GvUT2KZM9oKieQ5N0TL4bg45gA679BqI5iPch7XQYr7rjJLex3apCEY9ZpdgITe+5iddKaJdCtaI4q8HVtYWCg1QCxQrwPS9zTjoCAg39PWjCGRcTcUMSLEEzTOpSvbv3vHLYw8SDFgh1GktmCtSC4LGueCz79aQJ0kB1Kk1HSI7AiE0uYVg3UDVaGVMyNl4r23VgyiJj6QSsEutRsQ2IOiG96NOXLkiUsdKDCONraJOr5SEHVg7J13iuz1BoMJcbft+cmBA4mJ4HOK5nL5TWXfwQkHihn6vKvI7ja4z16uizqQuBaMxAOHymQcl9fUkYqtvPJ/TTQN9LgbLifHHAphACdxAAbFYv2DVFK9yAWgaOz5g4ADjAJWHYqJGoA5OnTkPrl2Hdt5v6YmedOTyQeBk4/EHPzi4PkSGeDcE5N8EzE5m5vP2JyxKTMFjMxNg2sH3CyXjlU2qiEJGt+ZmrbuJTEjISGBkbC5NkqKeF0679iTaanRGLl+bWIjNXV/dm5udu76lxJeHpPzAAe+XMtYV7CRbHOMjQfHrdR1jIyBEAc6FZOpYKxLnVzNi6jxHWliRaYUNQdY0Fj6wZTBzfn7o1cDmqCuZyhSygwQJK/qiEkOKto7sRoQtjE7P2EwuUwkkl5KTglZFr26d2zYzux8Rm1GzMuZmYO1Cfm5uydX+7IOi96dvT5fzGC8mr++gHoLe1wc/RezjaFHpyLZNAAAAABJRU5ErkJggg==" alt="GOMo Logo"><h1>GOMo</h1>
<h2>Gene Ontology for Motifs</h2>
</div>
<p>
        For further information on how to interpret these results please access
        <a href="https://meme-suite.org/meme/doc/gomo-output-format.html">https://meme-suite.org/meme/doc/gomo-output-format.html</a>.<br>
 	To get a copy of the MEME software please access 
        <a href="https://meme-suite.org">https://meme-suite.org</a>.
      </p>
<p id="citation"><script>print_citation("citation", "GOMo");</script></p>
</div>
<a name="top_buttons"></a><div class="pad2">
<a class="jump" href="#overview">Investigated Motifs</a>
        |  
      <a class="jump" href="#version">Program information</a>
        |  
      <a class="jump" href="gomo.tsv">Results in TSV Format</a> <span id="results_tsv_help"></span>
        |  
      <a class="jump" href="gomo.xml">Results in XML Format</a> <span id="results_xml_help"></span><script>
        make_help_button($("results_tsv_help"), "pop_results_tsv");
        make_help_button($("results_xml_help"), "pop_results_xml");
      </script>
</div>
<a name="overview"></a><h2 class="mainh pad2">Investigated Motifs</h2>
<div class="box">
<h4>Overview</h4>
<div class="pad3"><table border="0" cellspacing="5" cellpadding="0" bgcolor="#ffffff">
<thead><tr>
<th>Motif <div class="help" onclick="help_popup(this,'pop_motif')"></div>
</th>
<th>Logo <div class="help" onclick="help_popup(this,'pop_logo')"></div>
</th>
<th>Predictions <div class="help" onclick="help_popup(this,'pop_predictions')"></div>
</th>
<th>Top 5 specific predictions <div class="help" onclick="help_popup(this,'pop_bestpred')"></div>
</th>
</tr></thead>
<tbody>
<tr bgcolor="#cccccc" height="1"><td colspan="4"></td></tr>
<tr>
<td><a href="#motif_crp">crp</a></td>
<td><a href="#motif_crp"><img src="res/logo1.png" height="80" alt="Missing Motif crp Logo" title="Motif crp Logo" border="0"></a></td>
<td align="center"><a href="#motif_crp">13</a></td>
<td>
<span class="bp">BP </span>carbohydrate transport<br><span class="mf">MF </span>sugar transmembrane transporter activity<br><span class="bp">BP </span>cellular carbohydrate catabolic process<br><span class="mf">MF </span>intramolecular oxidoreductase activity, interconverting aldoses and ketoses<br><span class="bp">BP </span>cellular macromolecule catabolic process<br>
</td>
</tr>
<tr bgcolor="#cccccc" height="1"><td colspan="4"></td></tr>
<tr>
<td><a href="#motif_nagC">nagC</a></td>
<td><a href="#motif_nagC"><img src="res/logo2.png" height="80" alt="Missing Motif nagC Logo" title="Motif nagC Logo" border="0"></a></td>
<td align="center"><a href="#motif_nagC">9</a></td>
<td>
<span class="bp">BP </span>biofilm formation<br><span class="bp">BP </span>peptidoglycan biosynthetic process<br><span class="bp">BP </span>peptidoglycan-based cell wall organization<br><span class="mf">MF </span>hydrolase activity, hydrolyzing O-glycosyl compounds<br>
</td>
</tr>
<tr bgcolor="#cccccc" height="1"><td colspan="4"></td></tr>
<tr>
<td><a href="#motif_narP">narP</a></td>
<td><a href="#motif_narP"><img src="res/logo3.png" height="80" alt="Missing Motif narP Logo" title="Motif narP Logo" border="0"></a></td>
<td align="center"><a href="#motif_narP">14</a></td>
<td>
<span class="bp">BP </span>anaerobic respiration<br><span class="mf">MF </span>molybdenum ion binding<br><span class="bp">BP </span>cytochrome complex assembly<br><span class="bp">BP </span>chemical homeostasis<br><span class="bp">BP </span>cellular homeostasis<br>
</td>
</tr>
</tbody>
</table></div>
</div>
<br><br><a name="motif_crp"></a><table width="100%" border="0" cellspacing="1" cellpadding="4" bgcolor="#FFFFFF"><tr>
<td><h2 class="mainh">Motif crp</h2></td>
<td align="right" valign="bottom">
<a href="#motif_nagC">Next</a> <a href="#top_buttons">Top</a>
</td>
</tr></table>
<div class="box">
          GO terms are shown in grey if a more specific GO term was also significantly associated with this motif. 
          The most specific GO terms are shown in black.<br><span class="bp">BP</span> stands for biological process, 
          <span class="cc">CC</span> stands for cellular component and 
          <span class="mf">MF</span> stands for molecular function.<br><br><table width="100%" cellpadding="0" cellspacing="2">
<col class="em6">
<col class="em7">
<col class="em6">
<col class="em6">
<col class="em7">
<col class="em2">
<col class="em25">
<col>
<thead><tr bgcolor="#ffffff" valign="top">
<th>GO term <div class="help" onclick="help_popup(this,'pop_goterm')"></div>
</th>
<th>score <div class="help" onclick="help_popup(this,'pop_gscore')"></div>
</th>
<th>
<i>p</i>-value <div class="help" onclick="help_popup(this,'pop_epvalue')"></div>
</th>
<th>
<i>q</i>-value <div class="help" onclick="help_popup(this,'pop_eqvalue')"></div>
</th>
<th nowrap>Specificity <div class="help" onclick="help_popup(this,'pop_specif')"></div>
</th>
<th colspan="2">GO name <div class="help" onclick="help_popup(this,'pop_goname')"></div>
</th>
<th>Gene ID / Rank (2518 genes in total) 
                <div class="help" onclick="help_popup(this,'pop_generank')"></div>
</th>
</tr></thead>
<tbody>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0044248" target="go_query">GO:0044248</a></td>
<td align="center">1.341e-06</td>
<td align="center">5.615e-07</td>
<td align="center">4.550e-04</td>
<td align="center">~1%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>cellular catabolic process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=deoC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">deoC (25)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucA (52)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=uxaC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">uxaC (64)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=exuT&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">exuT (65)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=tdcA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">tdcA (71)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (147)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (148)</a>, ...42 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0009056" target="go_query">GO:0009056</a></td>
<td align="center">8.869e-07</td>
<td align="center">5.615e-07</td>
<td align="center">4.550e-04</td>
<td align="center">~0%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>catabolic process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=deoC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">deoC (25)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=iraP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">iraP (37)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucA (52)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=uxaC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">uxaC (64)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=exuT&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">exuT (65)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=tdcA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">tdcA (71)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (147)</a>, ...49 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0008643" target="go_query">GO:0008643</a></td>
<td align="center">1.075e-04</td>
<td align="center">7.299e-06</td>
<td align="center">3.943e-03</td>
<td align="center">~4%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>carbohydrate transport</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=malK&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">malK (138)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=malE&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">malE (139)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ulaA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ulaA (174)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylF&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylF (179)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=idnD&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">idnD (198)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yjjL&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yjjL (824)</a>
</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0051119" target="go_query">GO:0051119</a></td>
<td align="center">2.285e-04</td>
<td align="center">2.639e-05</td>
<td align="center">8.038e-03</td>
<td align="center">~5%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>sugar transmembrane transporter activity</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=glpT&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">glpT (99)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=malK&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">malK (138)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=malE&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">malE (139)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylF&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylF (179)</a>
</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0044275" target="go_query">GO:0044275</a></td>
<td align="center">2.652e-04</td>
<td align="center">2.976e-05</td>
<td align="center">8.038e-03</td>
<td align="center">~7%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>cellular carbohydrate catabolic process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucA (52)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (147)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (148)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nagB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nagB (153)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylA (178)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=idnD&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">idnD (198)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nanA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nanA (201)</a>, ...10 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0015144" target="go_query">GO:0015144</a></td>
<td align="center">2.665e-04</td>
<td align="center">2.976e-05</td>
<td align="center">8.038e-03</td>
<td align="center">~3%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>carbohydrate transmembrane transporter activity</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=glpT&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">glpT (99)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=malK&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">malK (138)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=malE&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">malE (139)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ulaA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ulaA (174)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylF&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylF (179)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ycaI&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ycaI (1391)</a>
</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0016861" target="go_query">GO:0016861</a></td>
<td align="center">3.954e-04</td>
<td align="center">4.604e-05</td>
<td align="center">1.066e-02</td>
<td align="center">~14%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>intramolecular oxidoreductase activity, interconverting aldoses and ketoses</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=uxaC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">uxaC (64)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=srlA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">srlA (111)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nagB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nagB (153)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylA (178)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=hisL&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">hisL (641)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yihU&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yihU (858)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rpiB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rpiB (897)</a>, ...3 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0016860" target="go_query">GO:0016860</a></td>
<td align="center">4.737e-04</td>
<td align="center">6.569e-05</td>
<td align="center">1.331e-02</td>
<td align="center">~4%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>intramolecular oxidoreductase activity</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=uxaC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">uxaC (64)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=srlA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">srlA (111)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nagB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nagB (153)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylA (178)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=hisL&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">hisL (641)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yihU&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yihU (858)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rpiB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rpiB (897)</a>, ...4 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0016052" target="go_query">GO:0016052</a></td>
<td align="center">5.421e-04</td>
<td align="center">8.085e-05</td>
<td align="center">1.456e-02</td>
<td align="center">~3%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>carbohydrate catabolic process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucA (52)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (147)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (148)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nagB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nagB (153)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ybjQ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ybjQ (158)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylA (178)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=idnD&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">idnD (198)</a>, ...14 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0009057" target="go_query">GO:0009057</a></td>
<td align="center">5.850e-04</td>
<td align="center">9.096e-05</td>
<td align="center">1.474e-02</td>
<td align="center">~2%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>macromolecule catabolic process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=iraP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">iraP (37)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucA (52)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (147)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (148)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nagB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nagB (153)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ybjQ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ybjQ (158)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylA (178)</a>, ...20 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0016853" target="go_query">GO:0016853</a></td>
<td align="center">7.322e-04</td>
<td align="center">1.190e-04</td>
<td align="center">1.754e-02</td>
<td align="center">~1%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>isomerase activity</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=uxaC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">uxaC (64)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yqcC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yqcC (77)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=srlA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">srlA (111)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nagB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nagB (153)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylA (178)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nanA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nanA (201)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nanC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nanC (239)</a>, ...27 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0044265" target="go_query">GO:0044265</a></td>
<td align="center">8.893e-04</td>
<td align="center">1.404e-04</td>
<td align="center">1.896e-02</td>
<td align="center">~6%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>cellular macromolecule catabolic process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucA (52)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (147)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fruB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fruB (148)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nagB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nagB (153)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylA (178)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=idnD&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">idnD (198)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nanA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nanA (201)</a>, ...13 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0019321" target="go_query">GO:0019321</a></td>
<td align="center">1.272e-03</td>
<td align="center">1.976e-04</td>
<td align="center">2.464e-02</td>
<td align="center">~16%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>pentose metabolic process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fucP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucP (51)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucA (52)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rhaB (83)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=araB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">araB (144)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=xylA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">xylA (178)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=rbsD&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">rbsD (364)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=aldA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">aldA (1071)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yiaK&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yiaK (1938)</a>
</td>
</tr>
</tbody>
</table>
</div>
<br><a name="motif_nagC"></a><table width="100%" border="0" cellspacing="1" cellpadding="4" bgcolor="#FFFFFF"><tr>
<td><h2 class="mainh">Motif nagC</h2></td>
<td align="right" valign="bottom">
<a href="#motif_crp">Previous</a> <a href="#motif_narP">Next</a> <a href="#top_buttons">Top</a>
</td>
</tr></table>
<div class="box">
          GO terms are shown in grey if a more specific GO term was also significantly associated with this motif. 
          The most specific GO terms are shown in black.<br><span class="bp">BP</span> stands for biological process, 
          <span class="cc">CC</span> stands for cellular component and 
          <span class="mf">MF</span> stands for molecular function.<br><br><table width="100%" cellpadding="0" cellspacing="2">
<col class="em6">
<col class="em7">
<col class="em6">
<col class="em6">
<col class="em7">
<col class="em2">
<col class="em25">
<col>
<thead><tr bgcolor="#ffffff" valign="top">
<th>GO term <div class="help" onclick="help_popup(this,'pop_goterm')"></div>
</th>
<th>score <div class="help" onclick="help_popup(this,'pop_gscore')"></div>
</th>
<th>
<i>p</i>-value <div class="help" onclick="help_popup(this,'pop_epvalue')"></div>
</th>
<th>
<i>q</i>-value <div class="help" onclick="help_popup(this,'pop_eqvalue')"></div>
</th>
<th nowrap>Specificity <div class="help" onclick="help_popup(this,'pop_specif')"></div>
</th>
<th colspan="2">GO name <div class="help" onclick="help_popup(this,'pop_goname')"></div>
</th>
<th>Gene ID / Rank (2518 genes in total) 
                <div class="help" onclick="help_popup(this,'pop_generank')"></div>
</th>
</tr></thead>
<tbody>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0051704" target="go_query">GO:0051704</a></td>
<td align="center">1.801e-04</td>
<td align="center">2.302e-05</td>
<td align="center">2.681e-02</td>
<td align="center">~0%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>multi-organism process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=srlA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">srlA (79)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=solA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">solA (181)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yjbE&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yjbE (305)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=bhsA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">bhsA (376)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ymgC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ymgC (387)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=dcuB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">dcuB (442)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yihR&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yihR (478)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=hlyE&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">hlyE (536)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ttdA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ttdA (588)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=zwf&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">zwf (720)</a>, ...3 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0042710" target="go_query">GO:0042710</a></td>
<td align="center">4.020e-04</td>
<td align="center">5.503e-05</td>
<td align="center">2.681e-02</td>
<td align="center">~15%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>biofilm formation</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=srlA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">srlA (79)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=solA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">solA (181)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yjbE&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yjbE (305)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=bhsA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">bhsA (376)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ymgC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ymgC (387)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=dcuB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">dcuB (442)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yihR&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yihR (478)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ttdA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ttdA (588)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ymgA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ymgA (821)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yjiP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yjiP (889)</a>, ...1 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0009252" target="go_query">GO:0009252</a></td>
<td align="center">8.308e-04</td>
<td align="center">1.202e-04</td>
<td align="center">2.681e-02</td>
<td align="center">~89%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>peptidoglycan biosynthetic process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=mltB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltB (78)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mltA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltA (98)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mipA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mipA (128)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mraZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mraZ (166)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=zwf&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">zwf (720)</a>
</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0007047" target="go_query">GO:0007047</a></td>
<td align="center">8.308e-04</td>
<td align="center">1.202e-04</td>
<td align="center">2.681e-02</td>
<td align="center">~37%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>cellular cell wall organization</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=mltB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltB (78)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mltA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltA (98)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mipA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mipA (128)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mraZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mraZ (166)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=zwf&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">zwf (720)</a>
</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0031504" target="go_query">GO:0031504</a></td>
<td align="center">8.308e-04</td>
<td align="center">1.202e-04</td>
<td align="center">2.681e-02</td>
<td align="center">100%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>peptidoglycan-based cell wall organization</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=mltB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltB (78)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mltA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltA (98)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mipA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mipA (128)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mraZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mraZ (166)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=zwf&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">zwf (720)</a>
</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0009273" target="go_query">GO:0009273</a></td>
<td align="center">8.308e-04</td>
<td align="center">1.202e-04</td>
<td align="center">2.681e-02</td>
<td align="center">~29%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>peptidoglycan-based cell wall biogenesis</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=mltB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltB (78)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mltA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltA (98)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mipA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mipA (128)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mraZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mraZ (166)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=zwf&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">zwf (720)</a>
</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0042546" target="go_query">GO:0042546</a></td>
<td align="center">8.308e-04</td>
<td align="center">1.202e-04</td>
<td align="center">2.681e-02</td>
<td align="center">~8%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>cell wall biogenesis</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=mltB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltB (78)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mltA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltA (98)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mipA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mipA (128)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mraZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mraZ (166)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=zwf&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">zwf (720)</a>
</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0016798" target="go_query">GO:0016798</a></td>
<td align="center">1.101e-03</td>
<td align="center">1.595e-04</td>
<td align="center">2.767e-02</td>
<td align="center">~1%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>hydrolase activity, acting on glycosyl bonds</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=chiA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">chiA (11)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=uidA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">uidA (95)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mltA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltA (98)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=essD&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">essD (386)</a>
</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0004553" target="go_query">GO:0004553</a></td>
<td align="center">1.101e-03</td>
<td align="center">1.595e-04</td>
<td align="center">2.767e-02</td>
<td align="center">~2%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>hydrolase activity, hydrolyzing O-glycosyl compounds</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=chiA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">chiA (11)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=uidA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">uidA (95)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mltA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mltA (98)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=essD&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">essD (386)</a>
</td>
</tr>
</tbody>
</table>
</div>
<br><a name="motif_narP"></a><table width="100%" border="0" cellspacing="1" cellpadding="4" bgcolor="#FFFFFF"><tr>
<td><h2 class="mainh">Motif narP</h2></td>
<td align="right" valign="bottom">
<a href="#motif_nagC">Previous</a> <a href="#top_buttons">Top</a>
</td>
</tr></table>
<div class="box">
          GO terms are shown in grey if a more specific GO term was also significantly associated with this motif. 
          The most specific GO terms are shown in black.<br><span class="bp">BP</span> stands for biological process, 
          <span class="cc">CC</span> stands for cellular component and 
          <span class="mf">MF</span> stands for molecular function.<br><br><table width="100%" cellpadding="0" cellspacing="2">
<col class="em6">
<col class="em7">
<col class="em6">
<col class="em6">
<col class="em7">
<col class="em2">
<col class="em25">
<col>
<thead><tr bgcolor="#ffffff" valign="top">
<th>GO term <div class="help" onclick="help_popup(this,'pop_goterm')"></div>
</th>
<th>score <div class="help" onclick="help_popup(this,'pop_gscore')"></div>
</th>
<th>
<i>p</i>-value <div class="help" onclick="help_popup(this,'pop_epvalue')"></div>
</th>
<th>
<i>q</i>-value <div class="help" onclick="help_popup(this,'pop_eqvalue')"></div>
</th>
<th nowrap>Specificity <div class="help" onclick="help_popup(this,'pop_specif')"></div>
</th>
<th colspan="2">GO name <div class="help" onclick="help_popup(this,'pop_goname')"></div>
</th>
<th>Gene ID / Rank (2518 genes in total) 
                <div class="help" onclick="help_popup(this,'pop_generank')"></div>
</th>
</tr></thead>
<tbody>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0009061" target="go_query">GO:0009061</a></td>
<td align="center">3.737e-04</td>
<td align="center">5.109e-05</td>
<td align="center">4.327e-02</td>
<td align="center">~54%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>anaerobic respiration</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fdnG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fdnG (3)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nirB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nirB (9)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narG (21)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=frdA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">frdA (25)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=dmsA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">dmsA (32)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=torC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">torC (228)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=lldP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">lldP (715)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narZ (798)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=zwf&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">zwf (2156)</a>, ...1 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0045333" target="go_query">GO:0045333</a></td>
<td align="center">5.182e-04</td>
<td align="center">7.692e-05</td>
<td align="center">4.327e-02</td>
<td align="center">15%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>cellular respiration</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fdnG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fdnG (3)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nirB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nirB (9)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narG (21)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=frdA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">frdA (25)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=dmsA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">dmsA (32)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=torC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">torC (228)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=sdhC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">sdhC (672)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=lldP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">lldP (715)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narZ (798)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=cyoA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">cyoA (984)</a>, ...5 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0030151" target="go_query">GO:0030151</a></td>
<td align="center">6.633e-04</td>
<td align="center">1.028e-04</td>
<td align="center">4.327e-02</td>
<td align="center">100%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>molybdenum ion binding</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fdnG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fdnG (3)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narG (21)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=dmsA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">dmsA (32)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=torC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">torC (228)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narZ (798)</a>
</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0015980" target="go_query">GO:0015980</a></td>
<td align="center">9.717e-04</td>
<td align="center">1.718e-04</td>
<td align="center">4.327e-02</td>
<td align="center">~6%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>energy derivation by oxidation of organic compounds</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fdnG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fdnG (3)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nirB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nirB (9)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narG (21)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=frdA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">frdA (25)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=dmsA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">dmsA (32)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=torC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">torC (228)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=sdhC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">sdhC (672)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=lldP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">lldP (715)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narZ (798)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=cyoA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">cyoA (984)</a>, ...6 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0006091" target="go_query">GO:0006091</a></td>
<td align="center">9.717e-04</td>
<td align="center">1.718e-04</td>
<td align="center">4.327e-02</td>
<td align="center">~4%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>generation of precursor metabolites and energy</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fdnG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fdnG (3)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nirB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nirB (9)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narG (21)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=frdA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">frdA (25)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=dmsA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">dmsA (32)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=torC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">torC (228)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=sdhC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">sdhC (672)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=lldP&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">lldP (715)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narZ (798)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=cyoA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">cyoA (984)</a>, ...6 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0017004" target="go_query">GO:0017004</a></td>
<td align="center">1.063e-03</td>
<td align="center">1.903e-04</td>
<td align="center">4.327e-02</td>
<td align="center">~46%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>cytochrome complex assembly</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=napF&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">napF (1)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fdnG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fdnG (3)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narG (21)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=cydA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">cydA (609)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=sdhC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">sdhC (672)</a>
</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0043623" target="go_query">GO:0043623</a></td>
<td align="center">1.063e-03</td>
<td align="center">1.903e-04</td>
<td align="center">4.327e-02</td>
<td align="center">~9%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>cellular protein complex assembly</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=napF&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">napF (1)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fdnG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fdnG (3)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narG (21)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=cydA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">cydA (609)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=sdhC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">sdhC (672)</a>
</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0042592" target="go_query">GO:0042592</a></td>
<td align="center">1.382e-03</td>
<td align="center">2.482e-04</td>
<td align="center">4.327e-02</td>
<td align="center">~1%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>homeostatic process</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=aqpZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">aqpZ (28)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nhaB (30)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=adiC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">adiC (60)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=gadB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">gadB (80)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mscS&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mscS (155)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=gadA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">gadA (255)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=adiA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">adiA (370)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nhaA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nhaA (532)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=kefA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">kefA (565)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mscL&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mscL (821)</a>, ...3 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0048878" target="go_query">GO:0048878</a></td>
<td align="center">1.382e-03</td>
<td align="center">2.482e-04</td>
<td align="center">4.327e-02</td>
<td align="center">~2%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>chemical homeostasis</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=aqpZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">aqpZ (28)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nhaB (30)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=adiC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">adiC (60)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=gadB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">gadB (80)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mscS&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mscS (155)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=gadA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">gadA (255)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=adiA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">adiA (370)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nhaA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nhaA (532)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=kefA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">kefA (565)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mscL&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mscL (821)</a>, ...3 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0019725" target="go_query">GO:0019725</a></td>
<td align="center">1.382e-03</td>
<td align="center">2.482e-04</td>
<td align="center">4.327e-02</td>
<td align="center">~4%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>cellular homeostasis</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=aqpZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">aqpZ (28)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nhaB (30)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=adiC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">adiC (60)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=gadB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">gadB (80)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mscS&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mscS (155)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=gadA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">gadA (255)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=adiA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">adiA (370)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nhaA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nhaA (532)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=kefA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">kefA (565)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mscL&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mscL (821)</a>, ...3 more...</td>
</tr>
<tr valign="top" class="normal">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0009055" target="go_query">GO:0009055</a></td>
<td align="center">1.539e-03</td>
<td align="center">2.802e-04</td>
<td align="center">4.441e-02</td>
<td align="center">~13%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>electron carrier activity</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fdnG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fdnG (3)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narG (21)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=frdA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">frdA (25)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=dmsA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">dmsA (32)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=torC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">torC (228)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=cydA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">cydA (609)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=sdhC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">sdhC (672)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narZ (798)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=cyoA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">cyoA (984)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=ndh&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">ndh (1013)</a>, ...4 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0065007" target="go_query">GO:0065007</a></td>
<td align="center">1.695e-03</td>
<td align="center">3.094e-04</td>
<td align="center">4.495e-02</td>
<td align="center">~0%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>biological regulation</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=napF&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">napF (1)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=aqpZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">aqpZ (28)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nhaB (30)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=adiC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">adiC (60)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=gadB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">gadB (80)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=acnA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">acnA (106)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=lptA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">lptA (147)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mscS&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mscS (155)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=sohA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">sohA (224)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=alaS&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">alaS (236)</a>, ...36 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0065008" target="go_query">GO:0065008</a></td>
<td align="center">1.839e-03</td>
<td align="center">3.408e-04</td>
<td align="center">4.571e-02</td>
<td align="center">~0%
      </td>
<td align="left"><span class="bp">BP </span></td>
<td>regulation of biological quality</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=aqpZ&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">aqpZ (28)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nhaB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nhaB (30)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=adiC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">adiC (60)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=gadB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">gadB (80)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=mscS&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">mscS (155)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=sohA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">sohA (224)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=gadA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">gadA (255)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=adiA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">adiA (370)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nhaA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nhaA (532)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=kefA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">kefA (565)</a>, ...5 more...</td>
</tr>
<tr valign="top" class="implied">
<td align="left"><a href="http://amigo.geneontology.org/amigo/term/GO:0043169" target="go_query">GO:0043169</a></td>
<td align="center">2.049e-03</td>
<td align="center">3.953e-04</td>
<td align="center">4.923e-02</td>
<td align="center">~9%
      </td>
<td align="left"><span class="mf">MF </span></td>
<td>cation binding</td>
<td>
<a href="http://regulondb.ccg.unam.mx/search?term=fdnG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fdnG (3)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nirB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nirB (9)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=narG&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">narG (21)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=dmsA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">dmsA (32)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=uxaC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">uxaC (172)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=torC&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">torC (228)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=alaS&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">alaS (236)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=nagB&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">nagB (246)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=yfdR&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">yfdR (378)</a>, <a href="http://regulondb.ccg.unam.mx/search?term=fucA&amp;amp;type=gene&amp;amp;organism=ECK12" target="gene_query">fucA (510)</a>, ...16 more...</td>
</tr>
</tbody>
</table>
</div>
<br><div class="bar">
<a name="version"></a><b>GOMo version:</b><br>5.4.0 (Release date: Tue Mar 9 17:38:20 2021 -0800)<br><br><div class="subsection" id="reference"><script>print_citation("reference", "GOMo");</script></div>
<br><a name="sequences"></a><b>Input Data: </b><br>
      go-term-sequence mapping: bacteria_escherichia_coli_k12_1000_199.na.csv<br>
        scored sequence file: ama.xml<br><br><a name="command"></a><b>Command line summary:</b><br>
      This information can also be useful in the event you wish to report
      a problem with the GOMo software.<br><br>
      Command:<br><textarea rows="5" class="cmd" readonly>gomo --oc gomo_example_output_files --dag go.dag --motifs dpinteract_subset.meme bacteria_escherichia_coli_k12_1000_199.na.csv ama.xml</textarea><br><br>
      Seed: 3779869387<br><br>
      Significance Threshold: 0.05<br>
</div>
</body>
</html>
