<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Citing - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/citation.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2>Citing MEME Suite Programs</h2>
    <h3>To cite the full MEME Suite</h3>
    <ul>
      <li> <div class="citation" id="meme_suite"> <script>print_citation("meme_suite", "MEME_SUITE");</script> </div> </li>
    </ul>
    <h3>To cite individual tools</h3>
    <ul>
      <li> <b>AMA</b> <div class="citation" id="ama"> <script>print_citation("ama", "AMA");</script> </div> </li>
      <li> <b>AME</b> <div class="citation" id="ame"> <script>print_citation("ame", "AME");</script> </div> </li>
      <li> <b>CentriMo</b> <div class="citation" id="centrimo"> <script>print_citation("centrimo", "CentriMo");</script> </div> </li>
      <li> <b>DREME</b> <div class="citation" id="dreme"> <script>print_citation("dreme", "DREME");</script> </div> </li>
      <li> <b>FIMO</b> <div class="citation" id="fimo"> <script>print_citation("fimo", "FIMO");</script> </div> </li>
      <li> <b>GLAM2 and GLAM2SCAN</b> <div class="citation" id="glam2"> <script>print_citation("glam2", "GLAM2");</script> </div> </li>
      <li> <b>GOMo</b> <div class="citation" id="gomo"> <script>print_citation("gomo", "GOMo");</script> </div> </li>
      <li> <b>MAST</b> <div class="citation" id="mast"> <script>print_citation("mast", "MAST");</script> </div> </li>
      <li> <b>MCAST</b> <div class="citation" id="mcast"> <script>print_citation("mcast", "MCAST");</script> </div> </li>
      <li> <b>MEME</b> <div class="citation" id="meme"> <script>print_citation("meme", "MEME");</script> </div> </li>
      <li> <b>MEME-ChIP</b> <div class="citation" id="meme-chip"> <script>print_citation("meme-chip", "MEME-ChIP");</script> </div> </li>
      <li> <b>MoMo</b> <div class="citation" id="momo"> <script>print_citation("momo", "MoMo");</script> </div> </li>
      <li> <b>PSPs</b> <div class="citation" id="psps"> <script>print_citation("psps", "PSPs");</script> </div> </li>
      <li> <b>SEA</b> <div class="citation" id="sea"> <script>print_citation("sea", "SEA");</script> </div> </li>
      <li> <b>SpaMo</b> <div class="citation" id="spamo"> <script>print_citation("spamo", "SpaMo");</script> </div> </li>
      <li> <b>STREME</b> <div class="citation" id="streme"> <script>print_citation("streme", "STREME");</script> </div> </li>
      <li> <b>Tomtom</b> <div class="citation" id="tomtom"> <script>print_citation("tomtom", "Tomtom");</script> </div> </li>
      <li> <b>T-Gene</b> <div class="citation" id="t-gene"> <script>print_citation("t-gene", "T-Gene");</script> </div> </li>
      <li> <b>XSTREME</b> <div class="citation" id="xstreme"> <script>print_citation("xstreme", "XSTREME");</script> </div> </li>
    </ul>

    <h3>Related papers</h3>
    <h4>GOMO related</h4>
    <ul>
      <li class="citation">
        Mikael Bod&eacute;n and Timothy L. Bailey,
        &quot;Associating transcription factor binding site motifs with target Go terms and target genes&quot;,
        <em>Nucleic Acids Res</em>, <b>36</b>(12):4108-4117, 2008.
        <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2475605/">[full text]</a>
      </li>
    </ul>
    <h4>MAST related</h4>
    <ul>
      <li class="citation">
        Timothy L. Bailey and Michael Gribskov,
        &quot;Score distributions for simultaneous matching to multiple motifs&quot;
        <em>Journal of Computational Biology</em>,
        <b>4</b>:45-59, 1997.
        <a href="http://tlbailey.bitbucket.io/papers/sumofgum.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey and Michael Gribskov, 
        &quot;Methods and statistics for combining motif match scores&quot;
        <em>Journal of Computational Biology</em>,
        <b>5</b>:211-221, 1998.
        <a href="http://tlbailey.bitbucket.io/papers/stats.pdf">[pdf]</a>
      </li>
    </ul>
    <h4>MEME related</h4>
    <ul>
      <li class="citation">
        Timothy L. Bailey and Charles Elkan,
        &quot;Fitting a mixture model by expectation maximization to discover motifs in biopolymers&quot;,
        UCSD Technical Report, CS94-351, March, 1994, University of California at San Diego.
        <a href="http://tlbailey.bitbucket.io/papers/memeplus.tech.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey,
        &quot;Discovering motifs in DNA and protein sequences: The approximate common substring problem&quot;,
        Ph.D. dissertation, University of California at San Diego, 1995.
        <a href="http://tlbailey.bitbucket.io/papers/thesis.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey and Charles Elkan,
        &quot;Unsupervised Learning of Multiple Motifs in Biopolymers using EM&quot;
        <em>Machine Learning</em>,
        <b>21</b>(1-2):51-80, October, 1995.
        <a href="https://link.springer.com/content/pdf/10.1007/BF00993379.pdf">[pdf]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey and Charles Elkan,
        &quot;The value of prior knowledge in discovering motifs with MEME&quot;,
        <em>Proceedings of the Third International Conference on Intelligent Systems for Molecular Biology</em>,
        pp. 21-29, AAAI Press, Menlo Park, California, 1995.
        <a href="http://www.aaai.org/Papers/ISMB/1995/ISMB95-003.pdf">[full text]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey and Michael Gribskov,
        &quot;The megaprior heuristic for discovering protein sequence patterns&quot;",
        <em>Proceedings of the Fourth International Conference on Intelligent Systems for Molecular Biology</em>,
        pp. 15-24 AAAI Press, Menlo Park, California, 1996.
        <a href="http://www.aaai.org/Papers/ISMB/1996/ISMB96-002.pdf">[full text]</a>
      </li>
      <li class="citation">
        William N. Grundy, Timothy L. Bailey and Charles P. Elkan,
        &quot;ParaMEME: A Parallel Implementation and a Web Interface
        for a DNA and Protein Motif Discovery Tool&quot;,
        <em>Computer Applications in the Biological Sciences (CABIOS)</em>,
        <b>12</b>(4):303-310, 1996.
        <a href="http://bioinformatics.oxfordjournals.org/content/12/4/303">[full text]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey, Michael E. Baker and Charles P. Elkan,
        &quot;An artificial intelligence approach to motif discovery in protein sequences:
        application to steroid dehydrogenases&quot;,
        <em>Journal of steroid biochemistry and molecular biology</em>,
        <b>62</b>(1):22-44, 1997.
        <a href="https://www.sciencedirect.com/science/article/abs/pii/S0960076097000137">[full text]</a>
      </li>
      <li class="citation">
        Martin Tompa, Nan Li, Timothy L. Bailey, George M. Church, Bart De Moor,
        Eleazar Eskin, Alexander V. Favorov, Martin C. Frith, Yutao Fu, W. James Kent,
        Vsevolod J. Makeev, Andrei A. Mironov,, William S. Noble,
        Giulio Pavesi, Graziano Pesole, Mireille Regnier, Nicolas Simonis, Saurabh Sinha, 
        Gert Thijs, Jacques van Helden, Mathias Vandenbogaert, Zhiping Weng, Christopher Workman,
        Chun Ye and Zhou Zhu.
        &quot;Assessing computational tools for the discovery of
        transcription factor binding sites.&quot;  
        <em>Nature Biotechnology</em>,
        <b>23</b>(1):137-144, 2005.
        <a href="http://www.nature.com/nbt/journal/v23/n1/full/nbt1053.html">[full text]</a>
      </li>
      <li class="citation">
        Timothy L. Bailey, Nadya Williams, Chris Misleh and Wilfred W. Li,
        &quot;MEME: discovering and analyzing DNA and protein sequence motifs&quot;
        <em>Nucleic Acids Research</em>,
        <b>34</b>:W369-W373, 2006.
        <a href="http://nar.oxfordjournals.org/cgi/content/full/34/suppl_2/W369">[full text]</a>
      </li>
      <li class="citation">
        Wenxiu Ma, William Stafford Noble and Timothy L. Bailey,
        &quot;Motif-based analysis of large nucleotide datasets using
        MEME-ChIP&quot;
        <em>Nature Protocols</em>,
        <b>9</b>(6):1428-1450, 2014.
        <a
        href="http://www.nature.com/nprot/journal/v9/n6/full/nprot.2014.083.html">[full text]</a>
      </li>
    </ul>
    <h4>Tomtom related</h4>
    <ul>
      <li class="citation">
        Emi Tanaka, Timothy L. Bailey, Charles E. Grant, William S. Noble and Uri Keich,
        &quot;Improved similarity scores for comparing motifs&quot;
        <em>Bioinformatics</em>
        <b>27</b>(12): 1603-1609, 2011.
        <a href="http://bioinformatics.oxfordjournals.org/content/27/12/1603.full">[full text]</a>
      </li>
    </ul>
    <h4>Other</h4>
    <ul>
      <li class="citation">
        William N. Grundy, Timothy L. Bailey, Charles P. Elkan and Michael E. Baker.
        &quot;Hidden Markov Model Analysis of Motifs in Steroid Dehydrogenases and their Homologs&quot;
        <em>Biochemical and Biophysical Research Communications</em>,
        <b>231</b>:760-766, 1997.
        <a href="http://tlbailey.bitbucket.io/papers/dehydrog-bbrc97.pdf">[full text]</a>
      </li>
      <li class="citation" id="meta-meme"> <script>print_citation("meta-meme", "Meta-MEME");</script></li>
      <li class="citation">
        Michael E. Baker, William N. Grundy and Charles P. Elkan,
        &quot;Spinach CSP41, an mRNA-binding protein and ribonuclease, is homologous to
        nucleotide-sugar epimerases and hydroxysteroid dehydrogenases&quot;,
        <em>Biochemical and Biophysical Research Communications</em>
        <b>248</b>(2):250-254, 1998.
        <a href="https://www.sciencedirect.com/science/article/abs/pii/S0006291X98989513?via%3Dihub">[full text]</a>
      </li>
      <li class="citation">
        Michael E. Baker, William N. Grundy and Charles P. Elkan,
        &quot;A common ancestor for a subunit in the mitochondrial proton-translocating NADH:ubiquinone
        oxidoreductase (complex I) and short-chain dehydrogenases/reductases&quot;,
        <em>Cellular and Molecular Life Sciences</em>,
        <b>55</b>(3):450-455, 1999.
        <a href="https://link.springer.com/article/10.1007/s000180050301">[full text]</a>
      </li>
      <li class="citation">
        John Hawkins, Charles Grant, William S. Noble and Timothy L. Bailey,
        &quot;Assessing phylogenetic motif models for predicting transcription factor binding sites&quot;
        <em>Bioinformatics
        (Proceedings of the Intelligent Systems for Molecular Biology Conference)</em>,
        <b>25</b>(12):i339-347, 2009.
        <a href="http://bioinformatics.oxfordjournals.org/content/25/12/i339">[full text]</a>
      </li>
      <li class="citation">
        Gabriel Cuellar-Partida, Fabian A. Buske, Robert C. McLeay, Tom Whitington,
        William Stafford Noble and Timothy L. Bailey,
        &quot;Epigenetic priors for identifying active transcription 
        factor binding sites&quot;,
        <em>Bioinformatics</em> <b>28</b>(1):56-62, 2012
        <a href="http://bioinformatics.oxfordjournals.org/content/28/1/56.full">
        [full text]
        </a>
      </li>
      <li>
        <div class="citation">
          Timothy L. Bailey, Mikael Bod&eacute;n, Fabian A. Buske, Martin Frith, Charles E. Grant, Luca Clementi,
          Jingyuan Ren, Wilfred W. Li, William S. Noble,
          &quot;MEME SUITE: tools for motif discovery and searching&quot;,
          <i>Nucleic Acids Research</i>, <b>37</b>:W202-W208, 2009.
          <a
          href="http://nar.oxfordjournals.org/content/37/suppl_2/W202.full">[full text]</a>
        </div>
      </li>
    </ul>
    <h3>Google Scholar Page</h3>
    <div class="indent">
      <a href="http://scholar.google.com/citations?user=4PFFWg0AAAAJ">MEME Suite on Google Scholar</a>
    </div>
  </body>
</html>

