<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>fasta-grep - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2>fasta-grep</h2>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>fasta-grep &lt;re&gt; (-dna | -prot) [options]</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p><code>fasta-grep</code> displays the non-overlapping occurrences of a 
      PERL regular expression in FASTA sequences. 
      <code>fasta-grep</code> supports the <a href="alphabets.html">IUPAC
      alphabets</a> for amino acids and nucleotides.
    </div>

    <h3>Input</h3>
    <div class="indent">
      <h4><span class='popt'>&lt;re&gt;</span></h4>
      <p>A PERL regular expression.</p>
      <h4><span class='popt'>-dna | -prot</span></h4>
      <p>The sequences are DNA or protein, respectively.</p>
      <p>Reads FASTA formatted sequences from standard input.
    </div>

    <h3>Output</h3>
    <div class="indent">
      <p>Writes matches to the regular expression to standard output or,
      optionally, a file. The output is in the form
      <pre>
        FASTA sequence ID line followed by number of matches
        [line 1 of sequence]
        [match line 1]
        [line 2 of sequence]
        [match line 2]
        ...
        [last line of sequence]
        [last match line]
      </pre>
      For proteins, occurrences are marked on the match lines as:
      <pre>
        &gt;    start of occurrence
        &lt;    end of occurrence
      </pre>
      For DNA, occurrences are marked on the match line as:
      <pre>
        &gt;    start of occurrence
        &lt;    end of occurrence
        *    start and end of two occurrences
      </pre>
    </div>

    <h3 id="options">Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behavior</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">-dna</td><td></td>
            <td>Sequence is DNA.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-prot</td><td></td>
            <td>Sequence is protein.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-s</td><td></td>
            <td>Print whole matching sequences only.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-p</td><td></td>
            <td>Print positions only, not sequence; 1-based.
            (relative to input strand if DNA; see below)
            </td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-m</td><td></td>
            <td>Print IDs of matching sequences only.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-x</td><td></td>
            <td>Print IDs of non-matching sequences only.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-o</td><td></td>
            <td>Print occurrences only in "raw" format.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-f</td><td></td>
            <td>Print occurrences only in FASTA format.
            1-based positions (relative to the strand of the match if DNA)
            are appended to the sequence ID.
            </td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-a</td><td></td>
            <td>Print all occurrences (even overlapping ones);
            ignored unless -o or -f given.
            </td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-norc</td><td></td>
            <td>Only print matches to given strand.
            </td>
            <td>print matches for both DNA strands.</td>
          </tr>
          <tr>
            <td class="popt">-prosite</td><td></td>
            <td>&lt;re&gt; is in PROSITE format.
            </td>
            <td>print matches for both DNA strands.</td>
          </tr>
          <tr>
            <td class="popt">-erase</td><td></td>
            <td>Replace occurrences with 'N's (-dna) or 'X's (-prot).
            </td>
            <td></td>
          </tr>
	  <script>
            wropt("all-help", {"short": true});
	  </script>
        </tbody>
      </table>
    </div>


    <h3>Examples</h3>
    <div class="indent">
    <pre>
    fasta-grep WGATAAN -dna &lt; ~/crp0.s
    fasta-grep A[AT]G -dna &lt; ~/crp0.fasta 
    </pre>
    </div>
  </body>
</html>

