<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>gendb - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2>gendb</h2>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>gendb [options] &lt;sequence count&gt;</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>gendb generates the specified number of random sequences using a Markov
      model. The sequence lengths are selected uniformly at random within the
      range specified by <span class="popt">--minseq</span> and
      <span class="popt">--maxseq</span>.</p>
    </div>

    <h3>Input</h3>
    <div class="indent">
      <h4><span class='popt'>&lt;sequence count&gt;</span></h4>
      <p>The number of random sequences to generate.</p>
    </div>

    <h3>Output</h3>
    <div class="indent">
      <p>Writes the sequences in <a href="fasta-format.html">FASTA format</a>
      to standard output.</p>
    </div>
    
    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behavior</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">--alph</td><td><span class="pdat">file</span></td>
            <td>Generate random sequences using the alphabet defined in 
		file <span class=pdat>file</span>, an
		<a href="alphabet-format.html">alphabet definition</a> file.
              Note that this overrides the <span class="popt">--type</span> option.</td>
            <td>Protein sequences are generated unless overridden using
              the <span class="popt">--type</span> option.</td>
          </tr>
          <tr>
            <td class="popt">--ambig</td><td><span class="pdat">ambig fraction</span></td>
            <td>Sets the fraction of symbols that will be ambiguous (overrides
              <span class="popt">--type</span>)</td>
            <td>The default depends on the <span class="popt">--type</span>
              option.</td>
          </tr>
          <tr>
            <td class="popt">--bfile</td><td><span class="pdat">file</span></td>
            <td>Sets the background model used to generate the sequences from a
              <span class="pdat">file</span> in <a href="bfile-format.html"
                >background model format</a>.</td>
            <td>For the standard DNA and Protein alphabets a built-in 0-order
            background is used. If a non-standard alphabet is provided without
            a background then a uniform frequency distribution is used.</td>
          </tr>
          <tr>
            <td class="popt">--order</td><td><span class="pdat">n</span></td>
            <td>Load the background model up to order <span class="pdat">n</span>.</td>
            <td>Load the background model completely.</td>
          </tr>
          <tr>
            <td class="popt">--type</td><td>0|1|2|3|4</td>
            <td>Allowed types are:
            <ul>
              <li>0 = Protein with 1% ambiguous symbols (default)</li>
              <li>1 = DNA with 1% ambiguous symbols</li>
              <li>2 = codons (ignores <span class="popt">-bfile</span>)</li>
              <li>3 = DNA without ambiguous symbols</li>
              <li>4 = Protein without ambiguous symbols</li>
            </ul>
            </td>
            <td>If an alphabet is not specified with the
              <span class="popt">--alph</span> option then protein sequences
              are generated.</td>
          </tr>
          <tr>
            <td class="popt">--minseq</td><td><span class="pdat">min</span></td>
            <td>Minimum sequence length.</td>
            <td>The minimum sequence length is 50.</td>
          </tr>
          <tr>
            <td class="popt">--maxseq</td><td><span class="pdat">max</span></td>
            <td>Maximum sequence length.</td>
            <td>The maximum sequence length is 2,000.</td>
          </tr>
          <tr>
            <td class="popt">--dummy</td><td>&nbsp;</td>
            <td>Print a "dummy" sequence record before the generated sequences.
            The "dummy" sequence record is a a FASTA header line listing the 
            gendb parameters but not followed by any sequence lines.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">--seed</td><td><span class="pdat">seed</span></td>
            <td>Seed for random number generator.</td>
            <td></td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>
</html>

