<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>glam2mask - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2>glam2mask</h2>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>glam2mask [options] &lt;glam2 motif&gt; &lt;sequence file&gt;</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>This program masks the occurrences of a GLAM2 motif so that weaker
      motifs can be found. Masking replaces residues aligned to key positions
      with the symbol 'x'. By alternately applying glam2 and glam2mask several
      times, it is possible to find the strongest, second-strongest,
      third-strongest, etc. motifs in a set of sequences.</p>
    </div>

    <h3>Input</h3>
    <div class="indent">
      <h4><span class='popt'>&lt;glam2 motif&gt;</span></h4>
      <p>The name of a GLAM2 output file.</p>
      <h4><span class='popt'>&lt;sequence file&gt;</span></h4>
      <p>The name of a file containing <a href="fasta-format.html">FASTA formatted</a> sequences.</p>
    </div>

    <h3>Output</h3>
    <div class="indent">
      <p>The input sequences are output with the GLAM2 motif sites masked out
      with 'x's. The sequences are normally sent to standard output but this
      can be changed by specifying the <span class="popt">-o</span> option.</p>
    </div>

    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behavior</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">-o</td><td><span class="pdat">output file</span></td>
            <td>The masked sequences are written to <span class="pdat">output file</span>.</td>
            <td>The masked sequences are written to standard output.</td>
          </tr>
          <tr>
            <td class="popt">-x</td><td><span class="pdat">mask character</span></td>
            <td>The instances of the glam2 motif are masked out with the 
            <span class="pdat">mask character</span>.</td>
            <td>The instances of the glam2 motif are masked out with 'x'.</td>
          </tr>
        </tbody>
      </table>
    </div>

    <h3>Examples</h3>
    <div class="indent">
      <ul>
        <li>Mask the top motif in my_motif.glam2 out of my_seqs.fa, and
        store the result in mask1.fa:<br>
        <code>glam2mask -o mask1.fa my_motif.glam2 my_seqs.fa</code>
        </li>
        <li>Mask the top motif in my_motif.glam2 out of my_seqs.fa, using
        'n' as the mask symbol:<br>
        <code>glam2mask -x n my_motif.glam2 my_seqs.fa</code>
        </li>
      </ul>
    </div>

  </body>
</html>

