<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>MAST - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
    <script src="js/citation.js"></script>
    <style>
      td.high {
        background-color:#FFFACD;
      }
      span.L_A {
        color:red;
      }
      span.L_C {
        color:blue;
      }
      span.L_G {
        color:orange;
      }
      span.L_T {
        color:green;
      }
    </style>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/mast_icon.png" alt="MAST Logo">
      <h1>MAST</h1>
      <h2>Motif Alignment &amp; Search Tool</h2>
    </div>
   
    <div style='display:inline' id='man_web_button'>
      <p> <button onclick="toggle_man_display()">View Web Version</button> </p>
    </div>

    <div style='display:none' id='man_cmd_button'>
      <p> <button onclick="toggle_man_display()">View Command-Line Version</button> </p>
    </div>

    <div style='display:inline' id='man_usage'>
      <div class="usage">
        <h3>Usage:</h3>
        <code>mast [options] &lt;motif file&gt; &lt;sequence file&gt;</code>
      </div>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <span class="man blurb" data-id="mast"></span>
    </div>
 
    <!-- Command Line Version Doc -->
    <div style='display:inline' id='man_cmd'>
      <h3>Input</h3>
      <div class="indent">
      <h4><span class='popt'>&lt;motif file&gt;</h4>
        <div class="indent">
          <p>The name of a file containing <a href="meme-format.html">MEME formatted motifs</a>.
          Outputs from MEME and DREME are supported, as well as Minimal MEME
          Format. You can convert many other motif formats to MEME format
          using <a href="overview.html#motif_conversion_utilities">conversion scripts</a>
          available with the MEME Suite.
          MAST previously required a log-odds matrix in the motif format
          but, while the log-odds section is used preferentially, it is no longer
          required.</p>
        </div>

      <h4><span class='popt'>&lt;sequence file&gt;</h4>
        <div class="indent">
          <p>The name of a file containing <a href="fasta-format.html">FASTA formatted sequences</a>
          that are suspected to contain motif sites. See the
          <span class="popt">-dblist</span> option if you wish to specify
          multiple sequence files.</p>
        </div>
      </div>

      <h3>Output</h3>
      <div class="indent">
        <p>
          MAST writes its output to files in a directory named
          <code>mast_out</code>, which it creates if necessary. You can change the
          output directory using the <span class='popt'>-o</span> or <span class='popt'>-oc</span> options.
          The directory will contain:
        </p>
        <ul>
          <li><a href=examples/mast_example_output_files/mast.html><code>mast.html</code></a> -
            an HTML file that provides the results in a human-readable format</li>
          <li><a href=examples/mast_example_output_files/mast.txt><code>mast.txt</code></a> -
            a plain text file of the results for backwards compatibility with earlier versions of MAST</li>
          <li><a href=examples/mast_example_output_files/mast.xml><code>mast.xml</code></a> -
            an XML file that provides the results in a format designed for machine processing</li>
        </ul>

        <p>MAST works by calculating match scores for each sequence in the
        sequence file compared with each of the motifs in the group of motifs you
        provide. For each sequence, the match scores are converted into various
        types of <i>p</i>-values and these are used to determine the overall
        match of the sequence to the group of motifs and the probable order and
        spacing of occurrences of the motifs in the sequence.</p>

        <p>The MAST HTML output contains:</p>
        <ol>
          <li>A summary of the MOTIFS in the input, including their sequence
	  logos, and a correlation matrix indicating how similar each pair of motifs is
          to help avoid including redundant motifs that would bias the query.</li>
          <li>The SEARCH RESULTS showing the sequences with significant
          overall matches to the motifs in the query.  The sequences are
	  sorted by their match <i>E</i>-values, and each sequence is accompanied by
          a motif block diagram showing the order and spacing of the matches
          to the query motifs.  Further details about each of the motif matches
          is available by clicking near the desired sequence.</li>
          <li>The INPUTS & SETTINGS used in the query to MAST.  These
	  include a description of the sequences alphabet, a description of
	  the sequence file, a description of the query motif file and 
          other settings that affect how MAST runs.</li>
        </ol>
        
        <p>In order to avoid biased scores when multiple motif scores are combined, MAST 
        computes the pairwise correlations between each pair of motifs and displays
        it in the MOTIFS section of its HTML output. The
        correlation between two motifs is the maximum sum of Pearson's
        correlation coefficients for aligned columns, divided by the width of the
        shorter motif. The maximum is found by trying all alignments of the two
        motifs. Motifs with correlations below 0.60 have little effect on the
        accuracy of the combined scores. Pairs of motifs with higher
        correlations should be removed from the query.  This is done by
        default on the MAST web server, and can be requested via an option
        when MAST is run from the command line.</p>

      </div>
      <h3>Options</h3>
      <div class="indent">
        <table class="options">
          <thead>
            <tr>
              <th>Option</th>
              <th>Parameter</th>
              <th>Description</th>
              <th>Default Behavior</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th colspan="4">Input Options</th>
            </tr>
            <script>
              wropt("all-bg", {"mast": true, "bfile": true});
            </script>
            <tr>
              <td class="popt">-dblist</td><td></td>
              <td>The sequence file contains a
                <a href="#dblist">list of file names of FASTA-formatted files</a>.</td>
              <td>The sequence file contains FASTA-formatted sequences.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Output Options</th>
            </tr>
            <script>
              wropt("all-o", {"dir": "mast_out"});
              wropt("all-oc", {"dir": "mast_out"});
            </script>
            <tr>
              <td class="popt">-hit_list</td><td></td>
              <td>Write a machine-readable (plain text) list of all non-overlapping
              motif matches (or just the single best hit for each motif,
              see <span class="popt">-best</span>, below)
              to standard output.  No other output is created.
              See the section titled <b>Hit List</b> below for details of the
	      output format.</td>
              <td>MAST outputs HTML, XML and plain text, and does not output a "hit list".</td>
            </tr>
            <tr>
              <td class="popt">-best</td><td></td>
              <td>Include only the best motif hits in the list of motif sites
              generated by <span class="popt">-hit_list</span>.
              This option has no effect unless <span class="popt">-hit_list</span>
              is specified.</td>
              <td>All non-overlapping motif sites are listed in the
	      <span class="popt">-hit_list</span> text output.</td>
            </tr>
            <tr>
              <td class="popt">-nostatus</td><td></td>
              <td>Do not print progress updates to standard error.</td>
              <td>Progress updates are printed to standard error.</td>
            </tr>
            <tr>
              <td class="popt">-notext</td><td></td>
              <td>Do not create plain text output.</td>
              <td>MAST creates HTML, XML and plain text output.</td>
            </tr>
            <tr>
              <td class="popt">-nohtml</td><td></td>
              <td>Do not create HTML output.</td>
              <td>MAST creates HTML, XML and plain text output.</td>
            </tr>
	    <script>
	      wropt("all-version");
	    </script>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Which Motifs To Use</th>
            </tr>
            <tr>
              <td class="popt">-remcorr</td><td></td>
              <td>Remove highly correlated motifs from query.</td>
              <td>No motifs are removed from the query.</td>
            </tr>
            <tr>
              <td class="popt">-m</td><td><span class="pdat">n</span></td>
              <td>Use only motifs appearing at the <span class="pdat">n</span>th
              position in the file. This option may be repeated.</td>
              <td>Use all the motifs.</td>
            </tr>
            <tr>
              <td class="popt">-c</td><td><span class="pdat">count</span></td>
              <td>Only use the first <span class="pdat">count</span> motifs.</td>
              <td>Use all the motifs.</td>
            </tr>
            <tr>
              <td class="popt">-mev</td><td><span class="pdat">thresh</span></td>
              <td>Use only motifs with <i>E</i>-values (or <i>p</i>-values) &le;
              <span class="pdat">thresh</span>.</td>
              <td>Use all the motifs.</td>
            </tr>
            <tr>
              <td class="popt">-diag</td><td><span class="pdat">diagram</span></td>
              <td>The nominal order an spacing of motifs is specified by
              <span class="pdat">diagram</span>, which is a block diagram. MAST
              uses the preferred order and spacing to compute the "spacing <i>p</i>-value"
              for any observed motif spacing in a sequence.  The spacing <i>p</i>-value
              is treated as though it were an additional motif in computing the
              sequence <i>E</i>-value.
              In the diagram, motifs should be referred to by their <b>position</b> in the
              motif file, not by their name. For example, if motifs number 1 and 2 typically occur separated
              gap of 5, with motif 2 site preceding the motif 1 site, 
              the diagram would be <code>[2]-5-[1]</code>.  Each input motif may be specified
	      at most once in the diagram.  Any leading and trailing gaps are ignored.</td>
              <td>Sequence <i>E</i>-values ignore motif order and spacing.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Options for Alphabets with Complements (e.g., DNA)</th>
            </tr>
            <tr>
              <td class="popt">-norc</td><td></td>
              <td>Do not score the reverse complement strand. This option is
              not compatible with the <span class="popt">-sep</span> or
              <span class="popt">-dna</span> options.</td>
              <td>The <i>p</i>-value of a motif site is the minimum of
              its <i>p</i>-values on the two strands (for complementable alphabets).</td>
            </tr>
            <tr>
              <td class="popt">-sep</td><td></td>
              <td>Score the reverse complement strand as a separate
              sequence. This option is not compatible with the
              <span class="popt">-norc</span> or
              <span class="popt">-dna</span> options.</td>
              <td>The <i>p</i>-value of a motif site is the minimum of
              its <i>p</i>-values on the two strands (for complementable alphabets).</td>
            </tr>
            <tr>
              <td class="popt">-dna</td><td></td>
              <td>(DNA sequences only) Translate the DNA sequences to protein so protein motifs may
              be scanned. The motifs must be protein and the sequences must be
              DNA. This option is not compatible with 
              <span class="popt">-norc</span> or
              <span class="popt">-sep</span>.</td>
              <td>DNA sequences are not translated to protein and only DNA motifs may be used to scan them.</td>
            </tr>
            <tr>
              <td class="popt">-comp</td><td></td>
              <td>Adjust the <i>p</i>-values and <i>E</i>-values for sequence
              composition.</td>
              <td><i>P</i>-values are based on the overall background mode (see <span class="popt">-bfile</span>, above).</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Which Results To Print</th>
            </tr>
            <tr>
              <td class="popt">-ev</td><td><span class="pdat">evalue</span></td>
              <td>Output results for sequences with <i>E</i>-values &lt;
              <span class="pdat">evalue</span>.</td>
              <td>Output results for sequences with <i>E</i>-values &lt; 10.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Appearance of Block Diagrams</th>
            </tr>
            <tr>
              <td class="popt">-mt</td><td><span class="pdat">mt</span></td>
              <td>Show motif matches with <i>p</i>-value &lt;
              <span class="pdat">mt</span>.</td>
              <td>Show motif matches with a <i>p</i>-value &lt; 0.0001.</td>
            </tr>
            <tr>
              <td class="popt">-w</td><td></td>
              <td>show weak matches (<span class="pdat">mt</span> &lt;
              <i>p</i>-value &lt; <span class="pdat">mt</span> * 10) in angle
              brackets in the hit list or when the XML is converted to plain text.</td>
              <td>Only strong matches (see <span class="popt">-mt</span>) are indicated
		in the plain text output.</td>
            </tr>
            <tr>
              <td class="popt">-seqp</td><td></td>
              <td>Use SEQUENCE <i>p</i>-values for motif thresholds.</td>
              <td>use POSITION <i>p</i>-values for motif thresholds.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Miscellaneous</th>
            </tr>
            <tr>
              <td class="popt">-mf</td><td><span class="pdat">mf</span></td>
              <td>In results use <span class="pdat">mf</span> as motif file name.</td>
              <td>The actual name of the motif file is used.</td>
            </tr>
            <tr>
              <td class="popt">-df</td><td><span class="pdat">df</span></td>
              <td>In results use <span class="pdat">df</span> as the 'database' name 
              for the sequence file.
              This option is ignored when <span class="popt">-dblist</span> is
              specified.</td>
              <td>The actual name of the sequence file is used.</td>
            </tr>
            <tr>
              <td class="popt">-dl</td><td><span class="pdat">dl</span></td>
              <td>If there is on-line annotation for the sequences in your
              sequence file that can be accessed via a link of the form
              <tt>"http://anything?anything=anythingSEQUENCEIDanything"</tt>, 
              you can have MAST link each sequence ID in its results to its annotation. 
              The actual FASTA sequence ID of the sequence will be used
	      to replace the token <tt>SEQUENCEID</tt> in the
              pattern <span class="pdat">dl</span> that you specify.
              This is option is ignored when <span class="popt">-dblist</span> 
              is specified.</td>
              <td>Sequence IDs in the results are not linked to anything.</td>
            </tr>
            <tr>
              <td class="popt">-minseqs</td><td><span class="pdat">ms</span></td>
              <td>The lower bound on the number of sequences in the sequence file. 
              This will reduce the amount of memory required by MAST.</td>
              <td>MAST uses more memory.</td>
            </tr>
          </tbody>
        </table>
      </div>



      <h3>Match Scores</h3>
      <p>
      The match score of a motif to a position in a sequence is the sum of the
      score from each column of the position-dependent scoring matrix
      corresponding to the letter at that position in the sequence. For example, if
      the sequence is 
      </p>
      <pre>
       <span style="color:gray;">TAA</span><span class="L_T">T</span><span class="L_G">G</span><span class="L_T">T</span><span class="L_T">T</span><span class="L_G">G</span><span class="L_G">G</span><span class="L_T">T</span><span class="L_G">G</span><span style="color:gray;">CTGGTTTTTGTGGCATCGGGCGAGAATAGCGC</span>
          ========
      </pre>
      <p>
      and the motif is represented by the position-dependent scoring matrix (where
      each row of the matrix corresponds to a position in the motif) 
      </p>
      <table>
        <tr>
          <th>Position</th><th><span class="L_A">A</span></td><th><span class="L_C">C</span></th>
          <th><span class="L_G">G</span></th><th><span class="L_T">T</span></th>
        </tr>
        <tr><td>1</td><td>1.447</td><td>0.188</td><td>-4.025</td><td class="high">-4.095</td></tr>
        <tr><td>2</td><td>0.739</td><td>1.339</td><td class="high">-3.945</td><td>-2.325</td></tr>
        <tr><td>3</td><td>1.764</td><td>-3.562</td><td>-4.197</td><td class="high">-3.895</td></tr>
        <tr><td>4</td><td>1.574</td><td>-3.784</td><td>-1.594</td><td class="high">-1.994</td></tr>
        <tr><td>5</td><td>1.602</td><td>-3.935</td><td class="high">-4.054</td><td>-1.370</td></tr>
        <tr><td>6</td><td>0.797</td><td>-3.647</td><td class="high">-0.814</td><td>0.215</td></tr>
        <tr><td>7</td><td>-1.280</td><td>1.873</td><td>-0.607</td><td class="high">-1.993</td></tr>
        <tr><td>8</td><td>-3.076</td><td>1.035</td><td class="high">1.414</td><td>-3.913</td></tr>
      </table>
      <p>
      then the match score of the fourth position in the sequence (underlined)
      would be found by summing the score for T in position 1, G in position 2 and
      so on until G in position 8. So the match score would be 
      </p>
      <pre>
         score = -4.095 + -3.945 + -3.895 + -1.994
                 + -4.054 + -0.814 + -1.933 + 1.414 
               = -19.316
      </pre>
      <p>
      The match scores for other positions in the sequence are calculated in the
      same way. Match scores are only calculated if the match completely fits within
      the sequence. Match scores are not calculated if the motif would overhang
      either end of the sequence. 
      </p>
      <h3><i>p</i>-values</h3>
      <p>
      MAST reports all matches of a sequence to a motif or group of motifs in terms
      of the <i>p</i>-value of the match. MAST considers the <i>p</i>-values of four types of
      events: 
      </p>
      <ul>
      <li>position <i>p</i>-value: the match of a single position within a sequence to
      a given motif,</li>
      <li>sequence <i>p</i>-value: the best match of any position within a sequence
      to a given motif,</li>
      <li>combined <i>p</i>-value: the combined best matches of a sequence to a
      group of motifs, and</li>
      <li><i>E</i>-value: observing a combined <i>p</i>-value at least as small in a random
      sequence file of the same size.</li>
      </ul>
      <p>
      All <i>p</i>-values are based on a random sequence model that assumes each
      position in a random sequence is generated according to the average letter
      frequencies of all sequences in the appropriate (peptide or nucleotide)
      non-redundant database (ftp://ncbi.nlm.nih.gov/blast/db/) on September 22,
      1996. This can be overridden by specifying the <span class="popt">-bfile</span>
      or <span class="popt">-comp</span> options (see below).  For DNA sequences, 
      unless <span class="popt">-norc</span> is given, the positive and reverse 
      complement strand frequencies are averaged together.
      </p>
      <ol>
        <li><span class="popt">-bfile</span> <span class="pdat">bfile</span>
        The random model uses the letter frequencies given in 
        <span class="pdat">bfile</span> instead of the non-redundant database frequencies.
        The <span class="pdat">bfile</span> is in <a href="bfile-format.html"
        >Markov Background Model</a> format.
        You can create files in the appropriate format based on the base/residue
        composition of your own FASTA sequence files using the 
        command <a href="fasta-get-markov.html">fasta-get-markov</a> included in
        the MEME distribution.</li>

        <li><code>-comp</code>
        The random model uses the letter frequencies in the current target
        sequence instead of the non-redundant database frequencies.  This
        causes <i>p</i>-values and <i>E</i>-values to be compensated individually 
        for the actual composition of each sequence in the sequence file.  This option
        can increase search time substantially due to the need to compute
        a different score distribution for each high-scoring sequence.
        With this option and sequences on an alphabet with complements (e.g., DNA), 
        the positive and reverse complement strand frequencies are <i>not</i> averaged together.
        </li>
      </ol>

      <h4>Position <i>p</i>-value</h4>
      <p>The <i>p</i>-value of a match of a given position within a sequence to
      a motif is defined as the probability of a randomly selected position in a
      randomly generated sequence having a match score at least as large
      as that of the given position. 
      <b>Note:</b> If MAST is combining reverse complement strands,
      the position <i>p</i>-value is <b>not</b> corrected for multiple tests.</p>

      <h4>Sequence <i>p</i>-value</h4>
      <p>The <i>p</i>-value of a match of a sequence to a motif is defined as
      the probability of a randomly generated sequence of the same length
      having a match score at least as large as the largest match score of
      any position in the sequence. 
      </p>

      <h4>Combined <i>p</i>-value</h4>
      <p>The <i>p</i>-value of a match of a sequence to a group of motifs is
      defined as the probability of a randomly generated sequence of the same
      length having sequence <i>p</i>-values whose product is at least as small
      as the product of the sequence <i>p</i>-values of the matches of the
      motifs to the given sequence.</p>

      <h4><i>E</i>-value</h4>
      <p>The <i>E</i>-value of the match of a sequence in a sequence file to a group
      of motifs is defined as the expected number of sequences in a random
      sequence file of the same size that would match the motifs as well as the
      sequence does and is equal to the combined <i>p</i>-value of the sequence
      times the number of sequences in the sequence file.</p>

      <h3>High-scoring Sequences</h3>
      <p>MAST lists the names and part of the descriptive text of all sequences
      whose <i>E</i>-value is less than <i>E</i>. Sequences shorter than one or
      more of the motifs are skipped. The sequences are sorted by increasing
      <i>E</i>-value. The value of <i>E</i> is set to 10 for the WEB server but
      is user-selectable in the downloadable version of MAST.</p>

      <h3>Motif Diagrams (Plain text output and <span class="popt">-diag</span> input option only)</h3>
      <p>Motif diagrams show the order and spacing of non-overlapping matches to
      the motifs in each high-scoring sequence. Motif occurrences are determined
      based on the position <i>p</i>-value of matches to the motif. Strong
      matches (<i>p</i>-value &lt; <span class="pdat">mt</span>) are shown in
      square brackets (`[ ]'), weak matches (<span class="pdat">mt</span> &lt;
      <i>p</i>-value &lt; <span class="pdat">mt</span> * 10) are shown in angle
      brackets (`&lt; &gt;') and the length of non-motif sequence ("spacer") is
      shown between hyphens (`-'). 
      The value of <span class="pdat">mt</span> is 0.0001 for the WEB server but is
      user-selectable in the downloadable version of the MEME Suite.
      For example,</p>
      <pre>
             27-[+3]-44-&lt;4&gt;-99-[-1]-7
      </pre>
      <p>shows an initial spacer of length 27, followed by a strong match to
      motif 3, a spacer of length 44, a weak match to motif 4 on the positive
      strand, a spacer of length 99, a strong match to motif 1 on the negative
      strand and a final non-motif sequence of length 7.
      <p>Note that when scanning
      DNA sequences with protein motifs (<span class="popt">-dna</span> command
      line option), the frame of the match is indicated by one of the letters 
      "a", "b" or "c" following the motif number.  For example
      <pre>
             27-[+3a]-44-&lt;4c&gt;-99-[-1b]-7
      </pre>
      indicates that the matches are in frames "a", "c" and "b", respectively.
      </p>

      <h3>Annotated Sequences</h3>
      <p>MAST annotates each high-scoring sequence by printing the sequence
      along with the position and strength of all the non-overlapping motif
      occurrences. The four lines above each motif occurrence contain,
      respectively,</p>
      <ul>
        <li>the motif number of the occurrence,</li>
        <li>the position <i>p</i>-value of the occurrence,</li>
        <li>the best possible match to the motif, and</li>
        <li>a plus sign (`+') above each letter in the occurrence that has a
        positive match score to the motif.</li>
      </ul>
      <p>The best possible match to a motif is the sequence of letters that
      would achieve the highest match score.</p>

      <h3>Hit List</h3>
      <p>
      If you specify the <span class="popt">-hit_list</span> switch to MAST,
      MAST outputs ONLY a list of "hits" in easily machine-readable format.
      Each line corresponds to one motif occurrence in one sequence.
      The format of the hit lines is</p>
      <div class="indent">
        [<span class="pdat">sequence_name</span>
        [<span class="pdat">strand</span>]<span class="pdat">motif</span>
        <span class="pdat">id</span> <span class="pdat">alt_id</span>
        <span class="pdat">start</span> <span class="pdat">end</span>
        <span class="pdat">score</span> <span class="pdat"><i>p</i>-value</span>]+
      </div>
      <p>where</p>
      <div class="indent">
        <table>
          <tr><td><span class="pdat">sequence_name</span></td><td>is the name of the sequence containing the hit</td></tr>
          <tr><td><span class="pdat">strand</span></td><td>is the strand (+ or - for DNA, blank for protein),</td></tr>
          <tr><td><span class="pdat">motif</span></td><td>is the motif number,</td></tr>
          <tr><td><span class="pdat">id</span></td><td>is the motif ID,</td></tr>
          <tr><td><span class="pdat">alt_id</span></td><td>is the motif alternate ID,</td></tr>
          <tr><td><span class="pdat">start</span></td><td>is the starting position of the hit,</td></tr>
          <tr><td><span class="pdat">end</span></td><td>is the ending position of the hit, and</td></tr>
          <tr><td><span class="pdat">score</span></td><td>is the score the hit,</td></tr>
          <tr><td><span class="pdat"><i>p</i>-value</span></td><td>is the position <i>p</i>-value of the hit.</td></tr>
        </table>
      </div>
      <p>
      Two comment lines (starting with "#") are written above the list of hits,
      and the MAST command line is printed as a comment line after the list.
      An example of the output using the <code>-hit_list</code> switch to MAST is:
      </p>
      <textarea rows="11" style="width:100%" readonly>
# All non-overlapping hits in all sequences from "tests/common/crp0.s".
# sequence_name (strand+/-)motif id alt_id hit_start hit_end score hit_p-value
ce1cg +1 TKTGANCNABNTCACAHWT MEME-1 64 82  1504.00 4.71e-06
ara +1 TKTGANCNABNTCACAHWT MEME-1 58 76  1951.44 8.17e-08
bglr1 -1 TKTGANCNABNTCACAHWT MEME-1 76 94  1490.84 5.17e-06
ilv +2 CGGYGGGG MEME-2 5 12  1780.00 6.62e-06
male +1 TKTGANCNABNTCACAHWT MEME-1 17 35  1622.44 1.93e-06
male +2 CGGYGGGG MEME-2 41 48  1720.00 1.47e-05
trn9cat -1 TKTGANCNABNTCACAHWT MEME-1 84 102  1385.56 1.04e-05

# mast -hit_list tests/meme/meme.crp0.zoops tests/common/crp0.s
      </textarea>
      <h3 id="dblist">Loading Multiple Sequence Databases</h3>
      <p>Multiple sequence files can be loaded by MAST by putting the file names 
      into a file and specifying that file instead of the sequence file with the 
      option -dblist.</p>
      <p>The file list has one file name on each line with the optional name and link
      as follows:</p>
      <pre>
      &lt;file&gt; [&lt;name&gt; &lt;link&gt;]
      ...
      ...
      </pre>
      <p>If it is specified then the name will be used instead of the file name in 
      the output. If the link is specified then all sequences for that sequence file in
      the HTML output will have a hyperlink to the URL specified with the text 
      SEQUENCEID replaced with the FASTA sequence id.</p>

      <h3>Examples</h3>
      <p>
      The following examples assume that file "meme.results" is the
      output of a MEME run containing at least 3 motifs that was
      created on the trainingset "training.fasta" and file
      SwissProt is a copy of the Swiss-Prot database on your local disk.
      DNA_DB is a copy of a DNA database on your local disk.
      </p> 
      <ol>
      <li>Annotate the training set:
      <pre>
        mast meme.results training.fasta
      </pre>  
      </li>
      <li>Find sequences matching the motif and annotate them in
      the SwissProt database:
      <pre> 
        mast meme.results SwissProt
      </pre> 
      </li>
      <li>Show sequences with weaker combined matches to motifs.
      <pre>
        mast meme.results SwissProt -ev 200
      </pre>
      </li>
      <li>Include a nominal order and spacing of the first three motifs
      in the calculation of the sequence <i>p</i>-values to increase the
      sensitivity of the search for matching sequences.  Note that the
      leading and trailing gaps are ignored ("9-" and "-91" in the example):
      <pre>
        mast meme.results SwissProt -diag "9-[2]-61-[1]-62-[3]-91"
      </pre>
      </li>
      <li>Use only the first and third motifs in the search:
      <pre> 
        mast meme.results SwissProt -m 1 -m 3
      </pre> 
      </li>
      <li>Use only the first two motifs in the search:
      <pre>
        mast meme.results SwissProt -c 2
      </pre>
      </li>
      <li>Search DNA sequences using protein motifs, adjusting <i>p</i>-values and <i>E</i>-values 
      for each sequence by that sequence's composition:
      <pre>
        mast meme.results DNA_DB -dna -comp
      </pre>
      </li>
      </ol>
    </div>

    <h3>Citing</h3>
    <div class="subsection" id="citation"> <script>print_citation("citation", "MAST");</script></div>

    <script src="js/manual.js"></script>

  </body>
</html>

