<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>MEME-ChIP - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
    <script src="js/citation.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/memechip_icon.png" alt="MEME-ChIP Logo">
      <h1>MEME-ChIP</h1>
      <h2>Motif Analysis of Large Nucleotide Datasets</h2>
    </div>

    <div style='display:inline' id='man_web_button'>
      <p> <button onclick="toggle_man_display()">View Web Version</button> </p>
    </div>

    <div style='display:none' id='man_cmd_button'>
      <p> <button onclick="toggle_man_display()">View Command-Line Version</button> </p>
    </div>

    <div style='display:inline' id='man_usage'>
      <div class="usage">
        <h3>Usage:</h3>
        <code>meme-chip [options] [-db &lt;motif file&gt;]* &lt;primary sequence file&gt;</code>
      </div>
    </div>

    <h3>Description</h3>
    <div class="indent">
       <span class="man blurb" data-id="memechip"></span>
    </div>

    <!-- Command Line Version Doc -->
    <div style='display:inline' id='man_cmd'>
      <h3>Input</h3>
      <div class="indent">
        <h4><span class='popt'>[-db &lt;motif file&gt;]*</span></h4>
        <div class="indent">
          <p>(Optional but recommended) The names of one or more files containing 
	  <a href="meme-format.html">MEME formatted motifs</a>.
          Outputs from MEME, STREME and DREME are supported, as well as Minimal MEME
          Format. You can convert many other motif formats to MEME Motif format
          using <a href="overview.html#motif_conversion_utilities">conversion scripts</a>
          available with the MEME Suite.
          These motif file(s) will used by Tomtom and CentriMo.</p>
        </div>
        <h4><span class='popt'>&lt;primary sequence file&gt;</span></h4>
        <div class="indent">
          <p>The name of a file of sequences in <a href="fasta-format.html">FASTA format</a>.
          Ideally the sequences should be all
          the same length, between 100 and 500 base-pairs long and centrally
          enriched for motifs. The immediate regions around individual ChIP-seq 
          "peaks" from a transcription factor (TF) ChIP-seq experiment are ideal.
          The suggested 100 base-pair minimum size is based on the typical 
          resolution of ChIP-seq peaks but it is useful to have more of the
          surrounding sequence to give CentriMo the power to tell if a motif is
          centrally enriched. We recommend that you "repeat mask" your
          sequences, replacing repeat regions to the "N" character.</p>
        </div>
      </div>
      <h3>Output</h3>
      <div class="indent">
        <p>
          MEME-ChIP writes its output to files in a directory named
          <code>memechip_out</code>, which it creates if necessary. You can change the
          output directory using the <span class='popt'>-o</span> or <span class='popt'>-oc</span> options.
          The directory will contain the following files:
        </p>
        <ul>
          <li><a href=examples/memechip_example_output_files/meme-chip.html><code>meme-chip.html</code></a> - 
            an HTML file that provides the results in an interactive, human-readable format that contains
	    links to the other files produced by the analyses performed by MEME-ChIP
          <li><a href=examples/memechip_example_output_files/summary.tsv><code>summary.tsv</code></a> -
            a TSV (tab-separated values) file that provides 
            a summary of the results in a format suitable for parsing by scripts and viewing with Excel
	  <li><a href=examples/memechip_example_output_files/combined.meme><code>combined.meme</code></a> -
	     a text file that contains all the motifs identified by MEME-ChIP 
	     in <a href=meme-format.html>MEME Motif Format</a>
        </ul>
        <p>
	  In addition, the MEME-ChIP output directory will contain sub-directories
	  with the results of each of the individual analyses it performed. The results
	  in these directories are all linked to from the MEME-ChIP HTML output file.
        </p>
        <p>
	  <b>Note:</b> See this detailed description of the 
	  <a href=meme-chip-output-format.html>MEME-ChIP output formats</a> for more information.
	</p>
      </div>
      <h3>Options</h3>
      <div class="indent">
        <table class="options">
          <thead>
            <tr>
              <th>Option</th><th>Parameter</th><th>Description</th><th>Default Behavior</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th colspan="4">General Options</th>
            </tr>
            <script>
              wropt("all-o", {"dir": "memechip_out"});
              wropt("all-oc", {"dir": "memechip_out"});
            </script>
            <tr>
              <td class="popt">-db</td><td><span class="pdat">file</span></td>
              <td>Use <span class="pdat">file</span> containing DNA
              motifs in <a href="meme-format.html">MEME Motif format</a>.
              This file will used by Tomtom and CentriMo.
              This option may be used multiple times to pass multiple files.</td>
              <td>When no files are provided, Tomtom can't suggest similar motifs
              and CentriMo is limited to the discovered motifs.</td>
            </tr>
            </tr>
              <td class="popt">-neg</td><td><span class="pdat">file</span></td>
              <td>MEME-ChIP will look for motifs enriched in the primary sequences relative 
	      to this control set of sequences in <a href="fasta-format.html">FASTA format</a>.
              These sequences will be input as control sequences to MEME, STREME
              and CentriMo.  MEME will use its <a href="meme.html?man_type=cmd#objfun">"Differential Enrichment"</a> 
	      objective function.  When this option is used, the primary and control sequences should all
              be the same length; otherwise CentriMo <i>E</i>-values will be inaccurate.
              If the primary sequences are ChIP-seq peak regions from a
              transcription factor ChIP-seq experiment, similar regions from
              a knockout cell line or organism are a possible choices for control
              sequences.  The control sequences should be prepared in exactly
              the same way (e.g., repeat-masking) as the primary sequences.
              </td>
              <td>No control sequences are used for MEME or CentriMo.  MEME uses the 
	      <a href="meme.html?man_type=cmd#objfun">"Classic"</a> objective function.
              STREME will create the control set by shuffling the positive sequences 
              while preserving the frequency statistics of the order specified by
	      the <span class="popt">-neg</span> option.
	      </td>
            </tr>
            </tr>
              <td class="popt">-psp-gen</td><td></td>
              <td>Use <code><a href=psp-gen.html>psp-gen</a></code>
              to create a <a href=psp-format.html>position-specific prior</a>
              for use by MEME. MEME will use this prior and the 
	      <a href="meme.html?man_type=cmd#objfun">"Classic"</a> objective function,
	      rather than the "Differential Enrichment" objective function.
	      <b>Note1: </b>Requires the <span class="popt">-neg</span> option.
	      <b>Note2: </b>This was the default prior to MEME Suite version 5.0.</td>
              <td>MEME uses the control sequences directly via its 
	      <a href="meme.html?man_type=cmd#objfun">"Differential Enrichment"</a>
	      objective function, and no position-specific prior is created.</td>
            </tr>
	    <script>
	      wropt("all-stdalph", {"ddash": false, "dna": true});
	      wropt("all-stdalph", {"ddash": false, "rna": true});
	      wropt("all-stdalph", {"ddash": false, "experimental": true, "protein": true});
              wropt("all-x_alph", {"ddash": false});
              wropt("all-dna2rna", {"ddash": false, "memechip": true});
	    </script>
            <tr>
              <td class="popt">-bfile</td><td><span class="pdat">file</span></td>
              <td>Pass the <span class="pdat">file</span> specifying a background
	      model in <a href=bfile-format.html>Markov Background Model Format</a>
              to programs that support a background model
              (<a href=meme.html>MEME</a>, <a href=centrimo.html>CentriMo</a>, 
              <a href=fimo.html>FIMO</a>, <a href=spamo.html>SpaMo</a> and 
              <a href=tomtom.html>Tomtom</a>).  Consult the documentation for
	      those programs for details on how they use the background model.
	      Basically, you can use a background model in order to normalize 
              for biased distribution of letters and groups of letters in your sequences.
              A 0-order model adjusts for single letter biases, a 1-order model adjusts for
              dimer biases (e.g., GC content in DNA sequences), etc.
              </td>
              <td>A 0-order Markov background model is calculated from the input sequences 
	      and then passed by MEME-ChIP to the programs that support background models.</td>
            </tr>
            <tr>
              <td class="popt">-order</td><td><span class="pdat">order</span></td>
              <td>Set the order of the Markov background model that 
		will be used.  If a background model is not specified 
                via <span class="popt">-bfile</span>, MEME-ChIP will create one
                from the control sequences (if given) or from the primary sequences.
                STREME will generate its own Markov background model of this
		order, and, if control sequences are not given, it will
		create them from the primary sequences by shuffling
		preserving the frequency statistics of this order. 
              </td>
              <td>A Markov background model of order 2 is generated and passed
                to the programs that support it.</td>
            </tr>
            <tr>
              <td class="popt">-minw</td><td><span class="pdat">width</span></td>
              <td>The minimum width of motifs to find.</td>
              <td>A minimum width of 6 is used unless the maximum width has been 
                set to be less than 6 in which case the maximum width is used.</td>
            </tr>
            <tr>
              <td class="popt">-maxw</td><td><span class="pdat">width</span></td>
              <td>The maximum width of motifs to find.</td>
              <td>A maximum width of 15 is used unless the minimum width has been
              set to be larger than 15 in which case the minimum width is used.</td>
            </td>
            <tr>
              <td class="popt">-seed</td><td><span class="pdat">seed</span></td>
              <td>The seed for the randomized selection of sequences for MEME.</td>
              <td>A seed value of 1 is used.</td>
            </tr>
            <tr>
              <td class="popt">-ccut</td><td><span class="pdat">size</span></td>
              <td>For input to MEME and STREME, trim sequences to their central
	      region of <span class="pdat">size</span> base-pairs. (The full-length
	      sequences are input to CentriMo and SpaMo.)
              A value of 0 indicates that sequences should not be trimmed before being
	      passed to MEME and STREME.</td>
              <td>A maximum size of 100 is used.</td>
            </tr>
            <tr>
              <td class="popt">-group-thresh</td><td><span class="pdat">gthr</span></td>
              <td>Main threshold for clustering highly similar motifs in MEME-ChIP output.
              All motifs in a group will have a Tomtom <i>E</i>-value less than or equal to
              <span class="pdat">gthr</span> when compared to the seed motif for the group,
              which is the most significant motif in the group.
              <td>A value of 0.05 is used.</td>
            </tr>
            <tr>
              <td class="popt">-group-weak</td><td><span class="pdat">wthr</span></td>
              <td>Secondary threshold for clustering highly similar motifs in MEME-ChIP output.
              If this is specified by the user, groups will be merged into a more significant
              group if all their motifs are weakly similar to the seed motif of the more significant
              group. <span class="pdat">wthr</span> specifies the Tomtom <i>E</i>-value threshold
              for merging groups.
              <td>Set to be equal to twice the value of the main clustering threshold:
              2 * <span class="pdat">gthr</span>.</td>
            </tr>
            <tr>
              <td class="popt">-filter-thresh</td><td><span class="pdat">fthr</span></td>
              <td><i>E</i>-value threshold for including motifs in the output.</td>
              <td>A value of 0.05 is used.</td>
            </tr>
            <tr>
              <td class="popt">-time</td><td><span class="pdat">minutes</span></td>
              <td>The maximum time that MEME-ChIP is allowed to run before terminating
                itself gracefully.</td>
              <td>There is no time limit</td>
            </tr>
            <script>
              wropt("all-desc");
              wropt("all-dfile");
            </script>
            <tr>
              <td class="popt">-old-clustering</td><td>&nbsp;</td>
              <td>Pick seed motifs for clustering based only on significance;
              <td>Discovered motifs are preferentially used as seed motifs for clustering.</td>
            </tr>
            <tr>
              <td class="popt">-noecho</td><td>&nbsp;</td>
              <td>Don't echo the commands run.</td>
              <td>Echo the commands run to standard output.</td>
            </tr>
	    <script>
              wropt("all-help");
              wropt("all-version");
	    </script>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">MEME Specific Options</th>
            </tr>
            <tr>
              <td class="popt">-meme-brief</td>
              <td><span class="pdat">nbrief</span></td>
              <td>If there are more than <span class="pdat">nbrief</span> (primary) sequences,
                the size of MEME's output will be reduced by suppressing
                the inclusion of the sequence names, motif sites and scanned sites in MEME's HTML and XML outputs,
                and by suppressing the tables of sequence lengths, sites and block diagrams in MEME's text output.</td>
              <td>A value of 1000 is used for <span class="pdat">nbrief</span>.</td>
            </tr>
              <td class="popt">-meme-mod</td><td><span class="pdat">oops</span>|<span
                  class="pdat">zoops</span>|<span class="pdat">anr</span></td>
              <td>The number of motif sites that MEME will find per sequence.
              <div style="margin-left:1em;">
                <span class="pdat">oops</span> - One Occurrence Per Sequence,<br>
                <span class="pdat">zoops</span> - Zero or One Occurrence Per Sequence,<br>
                <span class="pdat">anr</span> - Any Number of Repetitions<br>
              </div>
              See <span class="popt">-mod</span> in the
              <a href="meme.html?man_type=cmd#site_distribution">MEME command-line documentation</a>.
              </td>
              <td>MEME defaults to using <span class=pdat>zoops</span> mode.</td>
            </tr>
            <tr>
              <td class="popt">-meme-nmotifs</td><td><span class="pdat">num</span></td>
              <td>The number of motifs that MEME should search for.  
              If <span class="pdat">num</span> is 0, MEME will not be run.</td>
              <td>MEME will find 3 motifs.</td>
            </tr>
            <tr>
              <td class="popt">-meme-minsites</td><td><span class="pdat">sites</span></td>
              <td>The minimum number of sites that MEME needs to find for a motif.</td>
              <td>MEME doesn't require any minimum number of sites for a motif.</td>
            </tr>
            <tr>
              <td class="popt">-meme-maxsites</td><td><span class="pdat">sites</span></td>
              <td>The maximum number of sites that MEME will find for a motif.</td>
              <td>MEME doesn't limit the number of sites it will find for a motif.</td>
            </tr>
            <tr>
              <td class="popt">-meme-p</td><td><span class="pdat">np</span></td>
              <td>Use faster, parallel version of MEME with <span class="pdat">np</span> processors.
                  The parameter <span class="pdat">np</span> may be a number or it
                  may be a quoted string <b>starting</b> with a number and followed by arguments
                  to the particular MPI run command for your installation (e.g., <code>mpirun</code>).</td>
              <td>Use a single processor.</td>
            </tr>
            <tr>
              <td class="popt">-meme-pal</td><td>&nbsp;</td>
              <td>Restrict MEME to searching for palindromes only.</td>
              <td>MEME searches for any motif, not just palindromes.</td>
            </tr>
            <tr>
              <td class="popt">-meme-searchsize</td>
              <td><span class="pdat">searchsize</span></td>
              <td>The maximum portion of the primary sequences (in characters) used by MEME
                  in searching for motifs.  See the documentation on the MEME 
		  <span class="popt">-searchsize</span> option in the
		  <a href="meme.html?man_type=cmd#searchsize">MEME documentation</a> for more details.
              </td>
              <td>MEME performs sampling if the primary sequences contain more than 100,000 characters.
            </tr>
            <tr>
              <td class="popt">-meme-norand</td><td></td>
              <td>If your (primary) sequences are sorted in order of confidence (best to worst)
              then you should select this option.  See the
              <a href="meme.html?doc_type=cmd#norand">MEME documentation</a>
              for the <span class="popt">-norand</span> option for more details.</td>
              <td>MEME randomly selects the (primary) sequences to include in its initial analysis
		if there are more than <span class="pdat">searchsize</span> primary sequences.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">STREME Specific Options</th>
            </tr>
            <tr>
              <td class="popt">-streme-pvt</td><td><span class="pdat"><i>p</i>-value</span></td>
              <td>Stop searching for more motifs when three successive motifs have
                <i>p</i>-values larger than this threshold.</td>
              <td>An <i>p</i>-value threshold of 0.05 is used.</td>
            </tr>
            <tr>
              <td class="popt">-streme-nmotifs</td><td><span class="pdat">count</span></td>
              <td>Stop searching for more motifs when <span class="pdat">count</span> motifs have been
                found.  If <span class="pdat">count</span> is 0, STREME will not be run.</td>
              <td>Search stops when the <span class="popt">-streme-pvt</span> criterion
	      has been satisfied.</td>
            </tr>
            <tr>
              <td class="popt">-streme-totallength</td><td><span class="pdat">len</span></td>
              <td>Tell STREME to randomly choose sequences from the input fils so that
	        the maximum total length of the sequences from each sequence file totals at most 
		<span class="pdat">len</span>. 
		This can prevent STREME from running out of memory.</td>
              <td>STREME uses all the input sequences.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">CentriMo Specific Options</th>
            </tr>
            <tr>
              <td class="popt">-centrimo-local</td><td>&nbsp;</td>
            <td>CentriMo perform <i>local</i> motif enrichment analysis,
                computing enrichment in every possible sequence region.</td>
            <td>CentriMo will perform <i>central</i> motif enrichment analysis,
                computing enrichment in centered regions only.</td>
          <tr>
            <td class="popt">-centrimo-score</td><td><span class="pdat">score</span></td>
            <td>Set the minimum accepted score for a match.</td>
            <td>A minimum score of 5 is used.</td>
          </tr>
          <tr>
            <td class="popt">-centrimo-maxreg</td><td><span class="pdat">region</span></td>
            <td>Set the size of the maximum region size tested.</td>
            <td>CentriMo will test all valid region sizes.</td>
          </tr>
          <tr>
            <td class="popt">-centrimo-ethresh</td><td><span class="pdat"><i>E</i>-value</span></td>
              <td>Set the <i>E</i>-value threshold for reporting enriched central regions.</td>
              <td>An <i>E</i>-value threshold of 10 will be used.</td>
            </tr>
            <tr>
              <td class="popt">-centrimo-noseq</td><td>&nbsp;</td>
              <td>Do not store sequence IDs in the output of CentriMo.</td>
              <td>CentriMo stores a list of the sequence IDs with matches in the
                best region for each motif.</td>
            </tr>
            <tr>
              <td class="popt">-centrimo-flip</td><td>&nbsp;</td>
              <td>Reflect the positions of matches on the reverse strand around the center.</td>
              <td>Matches on the reverse strand are counted where they occur in the sequence.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">SpaMo Specific Options</th>
            </tr>
            <tr>
              <td class="popt">-spamo-skip</td><td>&nbsp;</td>
            <td>Do not run SpaMo.  Can be combined with options 
               <span class="popt">-meme-nmotifs 0</span>,
               <span class="popt">-streme-nmotifs 0</span>, and
               <span class="popt">-fimo-skip</span>
               to use MEME-ChIP to run CentriMo and cluster
	       the significant motifs.</td>
               <td>Run SpaMo using most significant motif from each cluster as primary.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">FIMO Specific Options</th>
            </tr>
            <tr>
              <td class="popt">-fimo-skip</td><td>&nbsp;</td>
            <td>Do not run FIMO.  Can be combined with options
               <span class="popt">-meme-nmotifs 0</span>,
               <span class="popt">-streme-nmotifs 0</span>, and
               <span class="popt">-spamo-skip</span>
               to use MEME-ChIP to run CentriMo and cluster
	       the significant motifs.</td>
               <td>Run FIMO using most significant motif from each cluster to scan
               input sequences.</td>
            </tr>
          </tbody>
        </table>
      </div>

    </div>

    <h3>Citing</h3>
    <div class=indent id="citation"><script>print_citation("citation", "MEME-ChIP");</script></div>

    <script src="js/manual.js"></script>

  </body>
</html>
