<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>MEME - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
    <script src="js/citation.js"></script>
    <style>
      div.example {
        font-family: monospace;
      }
    </style>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/meme_icon.png" alt="MEME Logo">
      <h1>MEME</h1>
      <h2>Multiple EM for Motif Elicitation</h2>
    </div>

    <div style='display:inline' id='man_web_button'>
      <p> <button onclick="toggle_man_display()">View Web Version</button> </p>
    </div>

    <div style='display:none' id='man_cmd_button'>
      <p> <button onclick="toggle_man_display()">View Command-Line Version</button> </p>
    </div>

    <div style='display:inline' id='man_usage'>
      <div class="usage">
        <h3>Usage:</h3>
        <code>meme [options] &lt;primary sequence file&gt; </code>
      </div>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <span class="man blurb" data-id="meme"></span>
    </div>

    <!-- Command Line Version Doc -->
    <div style='display:inline' id='man_cmd'>
      <div class="indent">
        <br>
        See these <a href="#examples">examples</a> of how
        to use MEME.</p>
      </div>

      <h3>Input</h3>
      <div class="indent">
        <span class='popt'><b>&lt;primary sequence file&gt;</b></span>
        <div class="indent">
          <p>The name of a file containing the primary sequences in <a href="fasta-format.html">FASTA format</a>,
          or the word <span class="popt">stdin</span> to indicate that the primary sequences should be read from
          standard input. Note that MEME does not attempt to detect the alphabet
          from the primary sequences so you should specify it with the
          <span class="popt">-dna</span>, <span class="popt">-rna</span>,
          <span class="popt">-protein</span> or <span class="popt">-alph</span>
          options.
          MEME also supports a modification to the FASTA format for
          <a href="fasta-format.html#weights">weighting</a> the primary sequences. 
	  <!--
	  Large datasets will need to be <a href="fasta-subsample.html">sub-sampled</a>
          to allow MEME to run in <a href="#running_time">reasonable times</a>.</p>
	  -->
        </div>
      </div>

      <h3>Output</h3>
      <div class="indent">
        <p>
          MEME writes its output to files in a directory named
          <code>meme_out</code>, which it creates if necessary. You can change the
          output directory using the <span class='popt'>-o</span> or <span class='popt'>-oc</span> options.
          The directory will contain:
        </p>
        <ul>
          <li><a href=examples/meme_example_output_files/meme.html><code>meme.html</code></a> -
	    an HTML file that provides the results in an interactive, human-readable format
          <li><a href=examples/meme_example_output_files/meme.txt><code>meme.txt</code></a> -
	    a plain text file of the results for backwards compatibility with earlier versions of MEME
          <li><a href=examples/meme_example_output_files/meme.xml><code>meme.xml</code></a> -
	    an XML file that provides the results in a format designed for machine processing
          <li><a href=examples/meme_example_output_files/logo1.png><code>logoN.png,.eps</code></a> -
	    PNG and EPS images files containing sequence logos for each of the
	    motifs found by MEME (where <i>N</i> is the motif number)
          <li><a href=examples/meme_example_output_files/logo_rc1.png><code>logo_rcN.png,.eps</code></a> -
	  (<b>complementable alphabets only</b>) 
	    PNG and EPS images files containing reverse-complement sequence logos for each of the
	    motifs found by MEME (where <i>N</i> is the motif number)
        </ul>
        
        <p>MEME gives each motif it discovers a <b>name</b>, which is a 
	  <a href=motif-consensus.html>consensus sequence</a>
	  that <i>approximately</i> describes the motif.  Motif names are intended
	  to be mnemonic only, and are not intended to be used for searching
	  sequences for matches; use the position-specific probability matrix (PSPM) 
	  contained in each of the MEME output file formats for that purpose 
	  via programs such as FIMO and MAST.  
        </p>

      </div>

      <h3>Options</h3>
      <div class="indent">
        <table class="options">
          <thead>
            <tr>
              <th>Option</th>
              <th>Parameter</th>
              <th>Description</th>
              <th>Default Behavior</th>
            </tr>
          </thead>

          <tbody>
            <tr>
              <th colspan="4">Output</th>
            </tr>
            <script>
              wropt("all-o", {"ddash": false, "dir": "meme_out"});
              wropt("all-oc", {"ddash": false, "dir": "meme_out"});
            </script>
            <tr>
              <td class="popt">-text</td><td></td>
              <td>Output in text format only to standard output.</td>
              <td>The program behaves as if <code>-oc&nbsp;meme_out</code> had
              been specified.</td>
            </tr>
            <tr>
              <td class="popt">-brief</td>
	      <td><span class="pdat">nbrief</span></td>
              <td>If there are more than <span class="pdat">nbrief</span> (primary) sequences,
		the size of MEME's output will be reduced by suppressing
		the inclusion of the sequence names, motif sites and scanned sites in MEME's HTML and XML outputs,
	        and by suppressing the tables of sequence lengths, sites and block diagrams in MEME's text output.</td>
              <td>A value of 1000 is used for <span class="pdat">nbrief</span>.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4" id="objfun">Objective Function</th>
            </tr>
            <tr>
              <td class="popt">-objfun</td>
              <td> 
                 <span class="popt">classic</span>|&#8203;
                 <span class="popt">de</span>|&#8203;
                 <span class="popt">se</span>|&#8203;
                 <span class="popt">cd</span>|&#8203;
                 <span class="popt">ce</span>|&#8203;
                 <span class="popt">nc</span>
              </td>
              <td>This option is used to select the objective function that
		  MEME uses to select the width and number of sites in a motif,
		  to calculate the motif's statistical significance,
	          and to choose the best motif in the current iteration.
                <table class="dark">
                  <tr>
                    <th>Value</th><th>Name</th><th>Description</th>
                  </tr>
                  <tr>
                    <td><span class="popt">classic</span></td><td>Classic</td>
                    <td>This is the original MEME objective function.  It scores
			motifs using an approximation to the <i>E</i>-value of the 
			information content of the motif. 
			MEME will run much more quickly on larger datasets
			but the results may differ from previous versions of
			MEME. To more closely emulate previous versions of MEME,
		       	see the <span class="popt">-searchsize</span> option, below.
                    </td>
                  </tr>
                  <tr>
                    <td><span class="popt">de</span></td><td>Differential Enrichment</td>
                    <td>This objective function scores motifs based on the
			enrichment of their sites in the primary sequences
			compared with the control sequences. 
			MEME reserves two subsets of primary sequences---not seen by EM or
			in the search for EM starts---for use in 1) choosing the width and 
			number of sites of a given motif, and 2) computing the statistical 
			significance of the motif, respectively.  
			(See option <span class="popt">-hsfrac</span>, below.)
                        Motif enrichment is computed (by default) using the multiple 
			hypergeometric test (see option <span class="popt">-test</span>, below).
		    </td>
                  </tr>
                  <tr>
                    <td><span class="popt">se</span></td><td>Selective Differential Enrichment</td>
                    <td>This objective function scores motifs based on the
			enrichment of their sites in the primary sequences
			compared with the control sequences.  It is very similar to
			the <span class="popt">de</span> objective function,
			but determines the best motif width by masking
			all but a <b>selected</b> set of highest-information
			motif columns.  (See the discussion of 
			<a href="#motif_refinement"><b>Motif Refinement</b></a>, below.)
                    </td>
                  </tr>
                  <tr>
                    <td><span class="popt">cd</span></td><td>Central Distance</td>
                    <td>This objective function scores motifs based on their
			tendency to occur near the center of the primary
			sequences, which must all be of the same length.
			The width and number of sites of each motif are determined
			using the Bates distribution applied to the mean of the distances
			of the best sites from the centers of their sequences.
                        MEME reserves a subset of primary sequences for use in 
			computing the statistical significance of the motif.
			(See option <span class="popt">-hsfrac</span>, below.)
                        Motif enrichment is computed using the Bates distribution.
			This objective function is designed for use with sequences
			from whole genome approaches such as ChIP-seq and CLIP-seq,
			and the primary sequences should include adequate flanking
			region around the expected motif sites&mdash;e.g., use sequences
			of 500bp for ChIP-seq.  
                        <br><b>Note: </b>The search for starting
			points and their optimization are limited to the 
			central regions of sequences, as determined by the 
			<span class="popt">-cefrac</span> option, below.
                    </td>
                  </tr>
                  <tr>
                    <td><span class="popt">ce</span></td><td>Central Enrichment</td>
                    <td>This objective function scores motifs based on their
			enrichment in the central regions of the primary
			sequences, which must all be of same the length.
			(See option <span class="popt">-cefrac</span>, below.)
			The width and number of sites of each motif are determined
			using the binomial test.  
			MEME reserves a subset of primary sequences for use in 
			computing the statistical significance of the motif.
			(See option <span class="popt">-hsfrac</span>, below.)
                        Motif enrichment is computed using the binomial test.
			This objective function is designed for use with sequences
			from whole genome approaches such as ChIP-seq and CLIP-seq,
			and the primary sequences should include adequate flanking
			region around the expected motif sites&mdash;e.g., use sequences
			of 500bp for ChIP-seq.
                    </td>
                  </tr>
                  <tr>
                    <td><span class="popt">nc</span></td><td>Numerically Correct</td>
                    <td>This objective function is identical to the Classic algorithm except
			it scores motifs based on the estimate of
			the <i>p</i>-value of the LLR according to the NC algorithm
			of Hertz and Stormo, rather than using the estimate based
			on the product of <i>p</i>-values of the LLR of motif columns.  
			<b>Warning: </b>Using this function causes MEME to run very slowly.
                    </td>
                  </tr>
                </table>
              <td>MEME uses the Classic objective function.</td>
            </tr>
            <tr>
              <td class="popt">-test</td>
              <td> 
                 <span class="popt">mhg</span>|&#8203;
                 <span class="popt">mbn</span>|&#8203;
                 <span class="popt">mrs</span>
              </td>
              <td>The statistical test used for computing motif enrichment
		  when the objective function is specified as 
		  Differential Enrichment (<span class="popt">-objfun</span> <span class="popt">de</span>)
		  or Selective Differential Enrichment (<span class="popt">-objfun</span> <span class="popt">se</span>).
		  The three options cause MEME to use the
		  the multiple hypergeometric test (<span class="popt">mhg</span>),
		  the multiple binomial test (<span class="popt">mbn</span>),
		  or the multiple rank-sum test (<span class="popt">mrs</span>), respectively.
              </td>
              <td>MEME uses the multiple hypergeometric test with the <span class="popt">de</span> 
		<span class="popt">se</span> objective functions.
              <td>
            </tr>
            <tr>
              <td class="popt">-use_llr</td>
              <td></td>
              <td>Use the log-likelihood ratio method rather than the original 
		  <i>E</i>-value method for evaluating EM starting points when the objective 
		  function is set to
		  <span class="popt">classic</span>.  (See <a href="#sp_search"><b>"Search for starting points"</b></a>
		  below for more details.)
              </td>
              <td>The <span class="popt">classic</span> objective function uses the original
		  <i>E</i>-value method for evaluating EM starting points.
              <td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Control Sequences, Hold-out Set and Efficiency Control</th>
            </tr>
            <tr>
              <td class="popt">-neg</td>
              <td><span class="pdat">control sequences</span></td>
              <td>The name of a file containing control sequences in <a href="fasta-format.html">FASTA format</a>.
		  The control sequences must be in the same sequence alphabet as the primary sequences.
              </td>
              <td>If the objective function requires control sequences and none are
		  provided, MEME creates a set of control sequences by shuffling a copy
		  of each primary sequence while preserving the frequencies of k-mers 
		  (see next option.)
              <td>
            </tr>
            <tr>
              <td class="popt">-shuf</td>
	      <td><span class="pdat">kmer</span></td>
              <td>Preserve the frequencies of words of this size when shuffling primary sequences
		  to create control sequences.  <span class="pdat">kmer</span> 
		  must be in the range [1,..,6]. 
		  <br><b>Note: </b>Using <span class="pdat">kmer</span> = 1 is not advised if your 
		  sequences have been masked 
		  (e.g., using <a href="http://repeatmasker.org">RepeatMasker</a>) 
		  since the runs of "N"s in the masked sequence will be broken up by shuffling, 
		  resulting in shuffled sequences with very different statistical properties 
		  than the primary sequences.
              <td>MEME preserves the frequencies of words of length 2.
              <td>
            </tr>
            <tr>
              <td class="popt">-hsfrac</td>
              <td><span class="pdat">hsfrac</span></td>
              <td>The <b>minimum</b> fraction of the primary sequences that will be held out for
		  estimating the width, number of sites and significance of motifs
		  for certain objective functions (see above).  A larger fraction of the
		  primary sequences may be held out if the size of the primary
		  sequences exceeds the value specified using the <span class="popt">-searchsize</span> option
		  (see below).
		  <br><b>Note:</b> The letter frequencies in the final motifs reported by MEME are based 
		  on <b>all</b> primary sequences, including those in the hold-out set(s).
              </td>
              <td>MEME reserves 0.5 (50%) of the primary sequences for estimating the width, number
		  of sites and significance of motifs with objective functions that require hold-out set(s).
              </td>
            </tr>
            <tr>
              <td class="popt">-cefrac</td>
              <td><span class="pdat">cefrac</span></td>
              <td>The size of the central region (as a fraction of sequence length) that
		  MEME uses for computing central enrichment of motifs when the 
		  Central Enrichment objective function 
		  (<span class="popt">-objfun</span> <span class="popt">ce</span>)
		  is specified.  
		  <br><b>Note: </b>Candidate starting points and EM are also
		  restricted to this portion of the sequences with both the
		  Central Enrichment
		  (<span class="popt">-objfun</span> <span class="popt">ce</span>)
		  and Central Distance 
		  (<span class="popt">-objfun</span> <span class="popt">cd</span>)
		  objective functions.
              </td>
              <td>MEME uses a central region consisting of the middle 0.25 (25%) of 
		  each sequence with the Central Enrichment objective function.
              <td>
            </tr>
            <tr id="searchsize">
              <td class="popt">-searchsize</td>
              <td><span class="pdat">searchsize</span></td>
              <td>The maximum portion of the primary sequences (in letters) used by MEME 
		  in searching for motifs.
		  To reduce running time, if the total size of the primary sequences exceeds 
		  <span class="pdat">searchsize</span> letters, 
		  then MEME performs the search for motifs on a random
		  sample of the primary sequences of (approximately) 
		  <span class="pdat">searchsize</span> letters. 
                  To ensure that the order of the primary sequences has no effect,
		  MEME orders the sequences alphabetically (by sequence, not sequence ID) 
		  before choosing the sample (unless you specify
		  option <span class="popt">-norand</span>, see below).
		  <br><b>Note 1:</b> This option only affects the search for motifs; the motif letter 
		  frequencies and sites reported by MEME are based on <b>all</b> primary sequences.
		  <br><b>Note 2:</b> MEME's running time will increase roughly as the <b>square</b>
		  of <span class="pdat">searchsize</span>.
                  <br><b>Note 3:</b> Unless you specify <span class="pdat">searchsize</span> as <b>0</b>,
                  no individual sequence may be longer than <span class="pdat">searchsize</span>.
		  <br><b>Note 4:</b> If you specify <span class="pdat">searchsize</span> as <b>0</b>, 
                  MEME will use <b>all</b> of the sequences (no sampling).  This will cause the 
                  <span class="popt">classic</span> objective function to mostly closely
		  emulate previous versions of MEME, and MEME will run very <b>slowly</b> on 
		  large datasets.
              </td>
              <td>MEME performs sampling if the primary sequences contain more than 100,000 letters.
              (This is done to improve MEME's running time.)
	      </td>
            </tr>
            <tr id="norand">
              <td class="popt">-norand</td>
 	      <td></td>
 	      <td>If your (primary) sequences are sorted in order of confidence (best to worst)
              then you should select this option. It causes MEME not to randomize the order of the input sequences
	      if there are more than <span class="pdat">searchsize</span> letters in the (primary) sequences 
              (see the <span class="popt">-searchsize</span> option, above). 
	      <br><b>Note: </b>Using this option may make the results of MEME depend on the order of the (primary) sequences.</td>
 	      <td>The order of the input sequences is randomized before applying
	      the <span class="popt">-searchsize</span> option.  (This is done to ensure that
	      MEME's results, by default, do not depend on the order of the input sequences.)</td>
	    </tr>
            <tr>
              <td class="popt">-csites</td>
              <td><span class="pdat">csites</span></td>
              <td>Maximum number of sites used for computing the <i>E</i>-value 
		  with the <b>Classic</b> objective function
                  (<span class="popt">-objfun</span> <span class="popt">classic</span>).
                  To reduce running time, if the total number of primary sequences exceeds
		  this value with site distribution models <span class="popt">oops</span> and 
		  <span class="popt">zoops</span>,
		  then the search for motifs is performed on a random sample of the primary sequences
		  of (approximately) <span class="pdat">csites</span> sequences. 
		  If the site distribution model is <span class="popt">anr</span>, sequence sampling is not done, 
		  but the maximum number of sites a motif may have is limited to <span class="pdat">csites</span>
		  (unless this is overridden by the <span class="popt">-maxsites</span> 
		  or <span class="popt">-nsites</span> options).
		  <br><b>Note 1:</b> This only affects the search for motifs; the motif letter 
		  frequencies and sites reported by MEME are based on <b>all</b> sequences.
		  <br><b>Note 2:</b> MEME's running time will increase roughly as the <b>cube</b>
		  of <span class="pdat">csites</span> when it is larger than 10,000.
	  	  <br><b>Note 3: </b>This option has no effect if the objective function is not Classic.
              </td>
              <td> With the <b>Classic</b> objective function, MEME performs sequence sampling if 
		  there are more than 1000 primary sequences (site distribution 
		  models <span class="popt">oops</span> and 
		  <span class="popt">zoops</span>), or limits the total number of sites per motif to 
		  1000 (site distribution model <span class="popt">anr</span>).  
		  <br><b>Note: </b>You can suppress sequence sampling
		  by setting <span class="pdat">searchsize</span> to 0 using the 
		  <span class="popt">-searchsize</span> option, above.
              </td>
            </tr>
            <tr>
              <td class="popt">-seed</td>
              <td><span class="pdat">seed</span></td>
	      <td>Random seed for shuffling and sampling sequences (see above).</td>
	      <td>MEME uses a random seed of 0.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Alphabet</th>
            </tr>
            <script>
              wropt("all-alph", {"ddash": false, "meme": true, "ambig-unknown": true});
              wropt("all-stdalph", {"ddash": false, "meme": true, "ambig-unknown": true, "dna": true});
              wropt("all-stdalph", {"ddash": false, "meme": true, "ambig-unknown": true, "rna": true});
              wropt("all-stdalph", {"ddash": false, "meme": true, "ambig-unknown": true, "protein": true});
            </script>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Strands &amp; Palindromes</th>
            </tr>
            <tr>
              <td class="popt">-revcomp</td><td></td>
              <td>Consider both the given strand and the reverse complement
                strand when searching for motifs in a complementable alphabet (ie DNA).</td>
              <td>MEME searches for motifs of complementable alphabets on the given strand only.</td>
            </tr>
            <tr>
              <td class="popt">-pal</td><td></td>
              <td>This causes MEME to only look for palindromes in complementable datasets.
                During EM, MEME will average the letter frequencies in complementary columns of
                the motif (PSPM) together. For instance, if the width of the
                motif is 10, columns 1 and 10, 2 and 9, 3 and 8, etc., are
                averaged together. For DNA the averaging would combines the
                frequency of A in one column with T in the other,
                and the frequency of C in one column with G in the other.</td>
              <td>MEME does not average the letter frequencies in complementary columns of motifs.</td>
            </tr>
          </tbody>


          <tbody>
            <tr id="site_distribution">
              <th colspan="4">Contributing Site Distribution</th>
            </tr>
            <tr>
              <td><span class="popt">-mod</span></td><td><span class="popt">oops</span>|&#8203;<span
                  class="popt">zoops</span>|&#8203;<span class="popt">anr</span></td>
              <td>This option is used to describe the distribution of motif sites.
                <table class="dark">
                  <tr>
                    <th>Value</th><th>Name</th><th>Description</th>
                  </tr>
                  <tr>
                    <td><span class="popt">oops</span></td><td>One Occurrence Per Sequence</td>
                    <td>MEME assumes that each sequence in the dataset contains
                      exactly one occurrence of each motif. This option is the
                      fastest and most sensitive but the motifs returned by MEME
                      may be "blurry" if any of the primary sequences is missing 
                      them.</td>
                  </tr>
                  <tr>
                    <td><span class="popt">zoops</span></td><td>Zero or One Occurrence Per Sequence</td>
                    <td>MEME assumes that each sequence may contain at most one
                      occurrence of each motif. This option is useful when you
                      suspect that some motifs may be missing from some of the
                      primary sequences. In that case, the motifs found will be more
                      accurate than using the first option. This option takes
                      more computer time than the first option (about twice as
                      much) and is slightly less sensitive to weak motifs
                      present in all of the primary sequences.</td>
                  </tr>
                  <tr>
                    <td><span class="popt">anr</span></td><td>Any Number of Repetitions</td>
                    <td>MEME assumes each sequence may contain any number of
                      non-overlapping occurrences of each motif. This option is
                      useful when you suspect that motifs repeat multiple times
                      within a single sequence. In that case, the motifs found
                      will be much more accurate than using one of the other
                      options. This option can also be used to discover repeats
                      within a single sequence. This option takes the much more
                      computer time than the first option (about ten times as
                      much) and is somewhat less sensitive to weak motifs which
                      do not repeat within a single sequence than the other two
                      options. 
 		      <br><b>Note</b>: <span class="popt">tcm</span> is a synonym
		      for <span class="popt">anr</span> for backwards compatibility.
		    </td>
                  </tr>
                </table>
              
              </td>
              <td>MEME assumes the Zero or One Occurrence Per Sequence model.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Number of Motifs</th>
            </tr>
            <tr>
              <td class="popt">-nmotifs</td><td><span class="pdat">nmotifs</span></td>
              <td>MEME will stop searching for motifs after finding
                <span class="pdat">nmotifs</span> motifs.</td>
              <td>MEME will find 1 motif.</td>
            </tr>
            <tr>
              <td class="popt">-evt</td><td><span class="pdat">ev</span></td>
              <td>MEME will stop searching for motifs if the last motif found
                has an <i>E</i>-value &gt; <span class="pdat">ev</span>, or when
		1000 motifs have been found.  If you also specify option 
		<span class="popt">-nmotifs</span> (above),
		MEME will stop searching if <span class="pdat">nmotifs</span> have been found.
	      </td>
              <td>MEME will rely on other limits to decide when to stop
                searching for motifs.</td>
            </tr>
            <tr>
              <td class="popt">-time</td><td><span class="pdat">secs</span></td>
              <td>MEME will stop searching for motifs if it has found at least
                1 motif and it estimates that finding any more will cause the
                total running time to exceed <span class="pdat">secs</span> seconds.</td>
              <td>MEME will rely on other limits to decide when to stop
                searching for motifs.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Number of Motif Occurrences</th>
            </tr>
            <tr>
              <td class="popt">-nsites</td><td><span class="pdat">nsites</span></td>
              <td>When the motif site distribution model allows, MEME will only
                attempt to find motifs with <span class="pdat">nsites</span> sites.
                Specifying this option is equivalent to setting the
                <span class="popt">-minsites</span> and <span class="popt">-maxsites</span>
                options to the same value.
                When the site distribution model is <span class="popt">oops</span> this is ignored and the
                number of sites is set to the number of primary sequences.</td>
              <td>See the <span class="popt">-minsites</span> and
              <span class="popt">-maxsites</span> options for information on the
              default behavior.</td>
            </tr>
            <tr>
              <td class="popt">-minsites</td><td><span class="pdat">minsites</span></td>
              <td>When the motif site distribution model allows, MEME will attempt
                to find motifs with at least <span class="pdat">minsites</span> sites.
                When the site distribution model is <span class="popt">oops</span> this is ignored and the
                number of sites is set to the number of primary sequences.</td>
              <td>The minimum number of sites is set to 2 when it is not
                otherwise defined by the use of the <span class="popt">oops</span> model or the
                <span class="popt">-nsites</span> option.</td>
            </tr>
            <tr>
              <td class="popt">-maxsites</td><td><span class="pdat">maxsites</span></td>
              <td>When the motif site distribution model allows, MEME will attempt
                to find motifs with at most <span class="pdat">maxsites</span> sites.
                When the site distribution model is <span class="popt">oops</span> this is ignored and the
                number of sites is set to the number of primary sequences.</td>
              <td>When the site distribution model is <span class="popt">zoops</span> 
		<span class="pdat">maxsites</span> is set to the number of primary sequences.
                When the site distribution model is <span class="popt">anr</span> 
		<span class="pdat">maxsites</span> is set to 5 times the number of primary sequences.
              </td>
            </tr>
            <tr>
              <td class="popt">-wnsites</td><td><span class="pdat">weight</span></td>
              <td>This controls the strength of the bias towards motifs with
                exactly the expected number of sites as defined by the
                <span class="popt">-nsites</span>, <span class="popt">-minsites</span> and
                <span class="popt">-maxsites</span> options. It is a number in
                the range [0..1). The closer to one it is, the stronger the bias
                towards motifs with exactly the expected number of sites.</td>
              <td>The value of <span class="pdat">weight</span> is set to 0.8.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Motif Width</th>
            </tr>
            <tr>
              <td class="popt">-w</td><td><span class="pdat">w</span></td>
              <td>Search for motifs with a width of <span class="pdat">w</span>.</td>
              <td>Search for motifs with widths between the range set by
              <span class="popt">-minw</span> and <span class="popt">-maxw</span>.</td>
            </tr>
            <tr>
              <td class="popt">-minw</td><td><span class="pdat">minw</span></td>
              <td>Search for motifs with a width &ge; <span class="pdat">minw</span>.
                </td>
              <td>Searches for motifs with a minimum width of 8.</td>
            </tr>
            <tr>
              <td class="popt">-maxw</td><td><span class="pdat">maxw</span></td>
              <td>Search for motifs with a width &le; <span class="pdat">maxw</span>.
                </td>
              <td>Searches for motifs with a maximum width of 50.</td>
            <tr>
              <td class="popt">-allw</td>
	      <td></td>
              <td>Find starting points for EM for each possible motif width
		  from <span class="pdat">minw</span> to <span class="pdat">maxw</span>, 
		  inclusive. 
		  <br><b>Note:</b> Using this option with large
		  motif width ranges will slow MEME substantially.
                </td>
              <td>MEME finds EM starting points only for widths in a geometric
		  progression from <span class="pdat">minw</span> to <span class="pdat">maxw</span>
		  and then refines the width of the motif after running EM.
              </td>
            </tr>
            <tr>
              <td class="popt">-nomatrim</td><td></td>
              <td>Do not <a href="#multiple_alignment_trimming">adjust motif width</a> 
		using multiple alignments.</td>
              <td>With the <b>Classic</b> objective function, 
		  the motif is trimmed to avoid insertions and deletions.
		  <br><b>Note:</b> This is not done with the other objective functions,
		  so this option has no effect with them.</td>
            </tr>
            <tr>
              <td class="popt">-wg</td><td><span class="pdat">wg</span></td>
              <td>The gap opening cost for creating the alignments
                used to 
                <a href="#multiple_alignment_trimming">trim the motif</a>.</td>
              <td>The opening cost for a gap is 11.</td>
            </tr>
            <tr>
              <td class="popt">-ws</td><td><span class="pdat">ws</span></td>
              <td>The gap extension cost for creating the 
                alignments used to 
                <a href="#multiple_alignment_trimming">trim the motif</a>.</td>
              <td>The extension cost for a gap is 1.</td>
            </tr>
            <tr>
              <td class="popt">-noendgaps</td><td></td>
              <td>Do not count end gaps in the alignments used to
                <a href="#multiple_alignment_trimming">trim the motif</a>.</td>
              <td>End gaps are penalized like any other gap.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Background Sequence Model</th>
            </tr>
            <tr>
              <td class="popt">-bfile</td><td><span class="pdat">bfile</span></td>
              <td>The name of <a href="bfile-format.html">Markov background model</a> file.  
		You can use a background model with MEME in order 
	      	to normalize for biased distribution of letters and groups of letters in your sequences.
      		A 0-order model adjusts for single letter biases, a 1-order model adjusts for
      		dimer biases (e.g., GC content in DNA sequences), etc.
                <p style="margin-bottom:0">The background model is used by MEME for four purposes:</p>
                  <ol style="margin-top:0">
                    <li>during EM as the "null model",</li>
                    <li>for calculating the log likelihood ratio of a motif,</li>
                    <li>for calculating the significance (<i>E</i>-value) of a motif, and,</li>
                    <li>for creating the position-specific scoring matrix (log-odds matrix).</li>
                  </ol>
		  If the <span class="popt">-markov_order</span> option is given (see below),
		  MEME only reads the Markov model in <span class="pdat">bfile</span>
		  up to the given order, ignoring higher order entries.
                  <p>Note that MEME uses only the 0-order portion (single letter
                  frequencies) of the background model for purpose 4 for all objective
		  functions other than <span class="popt">classic</span>, and for purposes
		  3 and 4 when the objective function is <span class="popt">classic</span>.<p>
		  <p>Note that if both strands are being considered, MEME will create a background
		  model that averages the frequencies of letters (0-order) or words (higher order)
		  and their reverse complements.</p>
              </td>
              <td>A background model of the order requested (see <span class="popt">-markov_order</span>)
		is estimated from the primary sequences.</td>
            </tr>
            <tr>
              <td class="popt">-markov_order</td>
              <td><span class="pdat">order</span></td>
              <td>The (maximum) order of the Markov model to read from the background file 
		  or to create from the primary sequences.
		  If the <span class="popt">-bfile</span> option is not specified,
		  MEME uses the <code>fasta-get-markov</code> program with a total
		  pseudocount of 0.1 to create a Markov model from the primary sequences.
              </td>
              <td>MEME uses the 0-order Markov model from the background file, or
		  creates a 0-order Markov model using an add-one prior from the primary sequences.
              </td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Position-Specific Priors</th>
            </tr>
            <tr>
              <td class="popt">-psp</td><td><span class="pdat">PSP file</span></td>
              <td>The name of a <a href="psp-format.html">MEME position-specific priors</a> file. 
		This can be used to bias the search for
                motifs in MEME. The <span class="pdat">PSP file</span> supplies
                a position-specific prior distribution on the location of motif
                sites in sequence(s) in the input dataset.</td>
              <td>All motif sites are considered equally likely.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">EM Algorithm</th>
            </tr>
            <tr>
              <td class="popt">-maxiter</td><td><span class="pdat">maxiter</span></td>
              <td>The number of iterations of EM to run from any starting point.
                EM is run for <span class="pdat">maxiter</span> iterations or
                until convergence (see <span class="popt">-distance</span>, below) 
		from each starting point.</td>
              <td>MEME will use a maximum of 50 EM iterations from any starting point.</td>
            </tr>
            <tr>
              <td class="popt">-distance</td><td><span class="pdat">distance</span></td>
              <td>MEME stops iterating EM when the change in the motif frequency
                matrix is less than <span class="pdat">distance</span>. Change is
                defined as the euclidean distance between two successive
                frequency matrices.</td>
              <td>The distance used for measuring convergence is 0.001.</td>
            </tr>
            <tr>
              <td class="popt">-prior</td><td><span class="popt">dirichlet</span>|&#8203;<span 
                  class="popt">dmix</span>|&#8203;<span class="popt">mega</span>|&#8203;<span 
                  class="popt">megap</span>|&#8203;<span class="popt">addone</span></td>
              <td>The type of prior to use.
                <table class="light">
                  <tr>
                    <th>Value</th><th>Name</th><th>Description</th>
                  </tr>
                  <tr>
                    <td><span class="popt">dirichlet</span></td><td>Simple Dirichlet</td>
                    <td>This is the default for DNA sequences.
                      It is a simple dirichlet prior based on the background
                      model as set by the <span class="popt">-bfile</span>
                      option.</td>
                  </tr>
                  <tr>
                    <td><span class="popt">dmix</span></td><td>Dirichlets Mix</td>
                    <td>This is the default for protein sequences when using the
                      <span class="popt">oops</span> site distribution model. It is a mixture of Dirichlets
                      prior. The source of the Dirichlets is specified by the
                      <span class="popt">-plib</span> option.</td>
                  </tr>
                  <tr>
                    <td><span class="popt">mega</span></td><td>Mega-weight Dirichlets Mix</td>
                    <td>This is an extremely low variance version of
                      <span class="popt">dmix</span>. The variance is scaled
                      inversely with the size of the dataset.</td>
                  </tr>
                  <tr>
                    <td><span class="popt">megap</span></td><td>Mega-weight Dirichlets Mix Plus</td>
                    <td>This is the default for protein sequences when using the
                      <span class="popt">zoops</span> or <span class="popt">anr</span> models. This behaves like
                      <span class="popt">mega</span> until the last iteration of
                      EM when it reverts to <span class="popt">dmix</span>
                      behavior.</td>
                  </tr>
                  <tr>
                    <td><span class="popt">addone</span></td><td>Add One</td>
                    <td>Add +1 to each observed count.</td>
                  </tr>
                </table>
              </td>
              <td>
                When the prior is not selected:
                <table class="light" style="width:100%">
                  <tr>
                    <th>Model</th><th>DNA</th><th>Protein</th>
                  </tr>
                  <tr>
                    <td>One Occurrence Per Sequence</td>
                    <td><span class="popt">dirichlet</span></td>
                    <td><span class="popt">dmix</span></td>
                  </tr>
                  <tr>
                    <td>Zero or One Occurrence Per Sequence</td>
                    <td><span class="popt">dirichlet</span></td>
                    <td><span class="popt">megap</span></td>
                  </tr>
                  <tr>
                    <td>Any Number of Repetitions</td>
                    <td><span class="popt">dirichlet</span></td>
                    <td><span class="popt">megap</span></td>
                  </tr>
                </table>
              </td>
            </tr>
            <tr>
              <td class="popt">-b</td><td><span class="pdat">b</span></td>
              <td>The strength of the prior on model parameters: 
                <span class="pdat">b</span> = 0 means use intrinsic strength of
                prior for prior = <span class="popt">dmix</span>.</td>
              <td>
                <table class="light" style="width:100%">
                  <tr>
                    <th>Prior</th><th><span class="pdat">b</span></th>
                  </tr>
                  <tr><td><span class="popt">dirichlet</span></td><td>0.01</td></tr>
                  <tr><td><span class="popt">dmix</span></td><td>0</td></tr>
                </table>
              </td>
            </tr>
            <tr>
              <td class="popt">-plib</td><td><span class="pdat">plib</span></td>
              <td>The name of the file containing the
                <a href="dmix-format.html">Dirichlet mixtures prior library</a>.</td>
              <td>
                The default value of <span class="pdat">plib</span> depends on
                the sequence alphabet.
                <table class="light" style="width:100%">
                  <tr>
                    <th>Alphabet</th><th><span class="pdat">plib</span></th>
                  </tr>
                  <tr><td>DNA</td><td><code>INSTALL_DIR/etc/dna.plib</code></td></tr>
                  <tr><td>Protein</td><td><code>INSTALL_DIR/etc/prior30.plib</code></td></tr>
                </table>
              </td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Selecting Starts for EM</th>
            </tr>
            <tr>
              <td class="popt">-spfuzz</td><td><span class="pdat">fuzz</span></td>
              <td>The fuzziness of sequence to theta mapping. The meaning of this
                parameter depends on the choice of mapping function as set by the
                <span class="popt">-spmap</span> option.</td>
              <td>
                The default value of <span class="pdat">fuzz</span> depends on
                the mapping function (see <span class="popt">-spmap</span>).
                <table class="dark" style="width:100%">
                  <tr>
                    <th>Mapping Function</th><th><span class="pdat">fuzz</span></th>
                  </tr>
                  <tr>
                    <td>Uniform</td><td>2 / | alphabet |<br>e.g., 0.5 for DNA</td>
                  </tr>
                  <tr>
                    <td>Point Accepted Mutation</td><td>120</td>
                  </tr>
                </table>
              </td>
            </tr>
            <tr>
              <td class="popt">-spmap</td><td><span class="popt">uni</span>|<span class="popt">pam</span></td>
              <td>The mapping function to use for estimating theta.
                <table class="dark">
                  <tr>
                    <th>Value</th><th>Name</th><th>Description</th>
                  </tr>
                  <tr>
                    <td><span class="popt">uni</span></td><td>Uniform</td>
                    <td>Add a uniform prior of <span class="pdat">fuzz</span>
                    when converting a substring to an estimate of theta.</td>
                  </tr>
                  <tr>
                    <td><span class="popt">pam</span></td><td>Point Accepted Mutation</td>
                    <td>Use <a href="http://en.wikipedia.org/wiki/Point_accepted_mutation">PAM</a>
                      matrices, with the number of mutation events set by
                      <span class="pdat">fuzz</span>, to estimate theta.</td>
                  </tr>
                </table>
              </td>
              <td>
                The default mapping function depends on the sequence alphabet.
                <table class="dark" style="width:100%">
                  <tr>
                    <th>Alphabet</th><th>Mapping Function</th>
                  </tr>
                  <tr>
                    <td>DNA</td><td>Uniform</td>
                  </tr>
                  <tr>
                    <td>Protein</td><td>Point Accepted Mutation</td>
                  </tr>
                </table>
              </td>
            </tr>
            <tr>
              <td class="popt">-cons</td><td><span class="pdat">consensus</span></td>
              <td>(May be repeated.) Override the sampling of starting points and just use a
                starting point derived from <span class="pdat">consensus</span>.
                This is useful when an actual occurrence of a motif is known and
                can be used as the starting point for finding the motif.  If this
		option is specified <i>N</i> times, starting point sampling will be
		suppressed for the first <i>N</i> motifs.
                <br><b>Note:</b> With DNA and RNA sequences, motifs shorter than 6 can
		be difficult for MEME to find, so you may want to pad the consensus
		sequence with "N"s to width 6, e.g. <span class="popt">-cons NGATAA</span>
                rather than <span class="popt">-cons GATAA</span>.</td>
              <td>Refer to the <span class="popt">-spmap</span> option for the
              default behavior.</td>
            </tr>
          </tbody>

<!--
          <tbody>
            <tr>
              <th colspan="4">Branching Search on EM Starts</th>
            </tr>
            <tr>
              <td class="popt">-heapsize</td><td><span class="pdat">hs</span></td>
              <td>Size of heaps for widths where substring search occurs.
              See the <a href="#branching_search">branching search</a> section
              for more details.</td>
              <td></td>
            </tr>
            <tr>
              <td class="popt">-x_branch</td><td></td>
              <td><span class="experimental">Experimental</span>
                Perform x-branching. This does a <a 
                  href="http://en.wikipedia.org/wiki/Beam_search">BEAM search</a>
                and tests the actual words in the input as well as words at
                hamming distance 1 at each successive level of branching.</td>
              <td>Normal branching search is performed. See the
              <a href="#branching_search">branching search</a> section for more
              details.</td>
            </tr>
            <tr>
              <td class="popt">-w_branch</td><td></td>
              <td><span class="experimental">Experimental</span>
                Perform width branching. This is not optimised.</td>
              <td>Normal branching search is performed. See the 
              <a href="#branching_search">branching search</a> section for more
              details.</td>
            </tr>
            <tr>
              <td class="popt">-bfactor</td><td><span class="pdat">bf</span></td>
              <td>The number of iterations of branching search. See the
              <a href="#branching_search">branching search</a> section for more
              details.</td>
              <td>MEME uses 3 iterations of branching search.</td>
            </tr>
          </tbody>
-->

          <tbody>
            <tr>
              <th colspan="4">Misc</th>
            </tr>
              <td class="popt">-p</td><td>
              <span class="pdat">np</span></td>
              <td>Use faster, parallel version of MEME with <span class="pdat">np</span> processors.
                  The parameter <span class="pdat">np</span> may be a number or it 
                  may be a quoted string <b>starting</b> with a number and followed by arguments
                  to the particular MPI run command for your installation (e.g., <code>mpirun</code>).</td>
              <td>Use a single processor.</td>
            </tr>
            <tr>
              <td class="popt">-maxsize</td><td><span class="pdat">maxsize</span></td>
              <td>Maximum allowed dataset size (in letters). You can set this in scripts to prevent very
		large running times. A <span class="pdat">maxsize</span> value of 0 means "no limit".
                <br><b>Note:</b> Unlike previous versions of MEME, there is no limit on dataset size
		by default, so this option does not normally need to be used.
              </td>
              <td>No limit on dataset size.</td>
            </tr>
            <tr>
              <td class="popt">-nostatus</td><td></td>
              <td>Print no status messages to terminal.</td>
              <td>Print minimal status messages to terminal.</td>
            </tr>
            <tr>
              <td class="popt">-sf</td><td><span class="pdat">sf</span></td>
              <td>Print <span class="pdat">sf</span> as name of sequence file.</td>
              <td>Print actual file name.</td>
            </tr>
            <tr>
              <td class="popt">-V</td><td></td>
              <td>Print extensive status messages to terminal.</td>
              <td>Print minimal status messages to terminal.</td>
            </tr>
	    <script>
	      wropt("all-help", {"short": true});
	      wropt("all-version");
	    </script>
          </tbody>

        </table>
      </div>
      <h3>MEME algorithm overview</h3>
      <div class="indent">
        <p>MEME searches for motifs in a four step process: 
	  <ol>
	    <li><a href="#sp_search">Search for starting points.</a></li>
            <li><a href="#EM">Expectation Maximization (EM).</a></li> 
	    <li><a href="#motif_refinement">Motif Refinement.</a></li> 
	    <li><a href="#motif_erasing">Motif Erasing.</a></li>
          </ol>
	First, MEME searches for words in the primary sequences that are likely to
	be good starting points for EM. Second, MEME runs EM to convergence from each such starting point.
	Third, after EM converges, MEME refines the width and depth (number of sites) of the motif.
	Finally, MEME performs a "soft" erase procedure that essentially erases all the
	sites of the best motif found proportional to the certainty that the site is real.
	MEME repeats the above four steps to find subsequent motifs until a stopping criterion
	is met, as specified in the "Number of Motifs" section above.
        </p>
        <p id="sp_search"><b>Search for starting points.</b>
	The search for starting points is over words that occur in the primary sequences.
	The search tests words in a geometric progression of widths from 
	<span class="pdat">minw</span> to <span class="pdat">maxw</span>, inclusive.
	Each candidate word is tested for its likelihood as a good starting point for EM
	for finding a motif with (approximately) a given number of sites, testing
	separately for each value of <span class="pdat">nsites</span>
        in a geometric progression from
	<span class="pdat">minsites</span> to <span class="pdat">maxsites</span>, inclusive.
        Starting points are scored by using them for one iteration of EM to produce 
	an initial motif, and then scoring the resulting motif using an objective function.  
	When the objective function is <b>Classic</b> or <b>Numerically Correct</b>, 
	the score is an estimate of the <i>E</i>-value 
	of the information content of the initial motif.  
	When the objective function is <b>Central Enrichment</b>,
        the score is the number of sites of the initial motif in the 
	central regions of the sequences.
	When the objective function is <b>Central Distance</b>,
        the score is the mean distance of the sites of the initial motif from
	the centers of the sequences.
	When the objective function is <b>Differential Enrichment</b> and <b>Selective Enrichment</b>,
	starting points are scored by optimizing the <i>p</i>-value of the motif
	objective function over the number of sites.  With all objective functions,
	MEME uses the highest scoring candidate word for each combination
	of width and number of sites tested (or the optimal number of sites for
	<b>Differential Enrichment</b> and <b>Selective Enrichment</b>) as a 
	starting point from which to initialize EM.
        </p>
        <p id="EM"><b>Expectation Maximization.</b>
	The EM algorithm is used to generate motif models that are optimal with respect to the
	primary sequences in a maximum likelihood sense.  EM is prone to getting stuck in local 
	optima, so it is important to start it from several good, initial motif models.
	Thus, MEME converts each candidate starting point to an initial motif model and runs EM
	to convergence.
        </p>
        <p id="motif_refinement"><b>Motif Refinement.</b>
	MEME subjects each motif found by EM to further refinement using the selected objective
	function, selecting the as the final motif the one with the best value of the objective function.  
	This refinement is over both the width and depth (number of sites) of the motif.
	To compute the objective function on a motif, MEME predicts all possible sites using
	the motif, sorts them, and then applies the objective function to the sorted sites.
	With the <b>Classic</b> and <b>Numerically Correct</b> objective functions, 
	MEME tests all possible sub-motifs consisting of consecutive columns of the motif in this way.  
        With the <b>SE</b> objective function, trims the motif to a single width by
        <b>selecting</b> the highest <code>nkeep</code> columns and masking
	the remainder by setting them to the background frequencies, 
	where <code>nkeep = max(6, 6 + floor(width-6/3)</code>.
	MEME uses a greedy search over sub-motifs with high information content.
	First, it sorts the columns of the motif by information content.  
	Then MEME tests the sub-motifs spanned
	by the first 2, 3, ... etc. columns.  Consequently, if the first and last motif 
	columns have the highest information content, only the original motif is tested.  
	With the <b>Classic</b> and <b>Numerically Correct</b> objective
	functions, MEME also further shortens each candidate submotif using a procedure 
	that minimizes the number of gaps in a multiple alignment of the motifs sites.
        </p>
        <p id="motif_erasing"><b>Motif Erasing.</b>
	To enable EM to find additional motifs, MEME "erases" the predicted sites of the last
	motif found.  This is done in a "soft" fashion by adjusting a weight associated each 
	position in the primary sequences, reducing it in proportion to its likelihood 
	as the the start of a site for the motif.  These weights are used when computing 
	the scores of both starting points
	and final motifs, and a low weight causes a sequence position to be less likely to be
	chosen as a starting point and reduces its contribution to the motif model during EM.
        </p>
      </div>

      <h3 id="objective_function">Motif Objective Function</h3>
      <div class="indent">
        <p>MEME uses an objective function on motifs when refining a motif, and
	to select the "best" motif from among all the refined motifs obtained from
	different starting points for EM.
	MEME supports several choices of objective function, which are described
	above under the <span class="popt">-objfun</span> option.  
	Each of the objective functions is based on the significance
	(<i>E</i>-value) of the motif according to a statistical test.
        </p>
        <p id=classic_objfun>
        The <b>Classic</b> objective function is based on the statistical significance of the
        log likelihood ratio (LLR) of the occurrences of the motif. 
	The <i>E</i>-value of the motif is an estimate of the number of motifs (with
        the same width and number of occurrences) that would have equal or
        higher log likelihood ratio if the input sequences had been
        generated randomly according to the (0-order portion of the) background
        model. An approximate algorithm is used to estimate the probability of the
	observed (or better) LLR, and this <i>p</i>-value is then multiplied
	by an estimate of the number of possible motifs to give the <i>E</i>-value.
        </p>
	<p>The <b>Differential Enrichment</b> and <b>Selective Enrichment</b> objective functions are 
	based on the ability of the motif to <b>discriminate</b> 
	predicted sites in the primary sequences from those in the control sequences.  
        The motif is used to predict sites in a held-out subset of the primary sequences,
 	and in the control sequences.  The sites are sorted by their likelihood ratio scores.  
	A statistical test is then applied to the sites in the sorted list above and 
	below each possible score threshold.
	With the <b>Selective Enrichment</b> objective function, the best <i>p</i>-value among the 
	tests is then multiplied by the total number of motifs and thresholds tested to give 
	the <i>E</i>-value of the final motif.  
        With the <b>Discriminative Enrichment</b> objective function,
	the best motif is selected as above, and then a second hold-out set of
	primary sequences is used for computing the final <i>E</i>-value. 
	The statistical validity of the <i>E</i>-value is ensured
	by basing it on held-out primary sequences that are not used in
	finding the motif.
        </p>
        <p>
	The <b>Central Enrichment</b> and <b>Central Distance</b> objective functions 
	are based on the <b>central enrichment</b> of motif
	in the central region of the (equal-length) primary sequences.  
	The motif is used to predict sites in each of the primary sequences,
	and the sites are sorted by their likelihood ratio scores. 
	With the <b>Central Enrichment</b> objective function,
	the binomial test is then applied to the counts of the number of times the
	highest-scoring site occurs in the central vs. the flanking regions of the sequences.
	With the <b>Central Distance</b> objective function, the average distance of the
	best site in each sequence to the center of the sequence is computed, and its significance
	is estimated using the Bates distribution.
        </p>
<!--
        <p>The log likelihood ratio of a motif is
        <code>llr = log (Pr(sites | motif) / Pr(sites | back))</code> and is a
        measure of how different the sites are from the background model.
        <code>Pr(sites | motif)</code> is the probability of the occurrences
        given the a model consisting of the position-specific probability matrix
        (PSPM) of the motif. <code>Pr(sites | back)</code> is the probability
        of the occurrences given the background model. The background model is
        an <a href="bfile-format.html">n-order Markov model</a>. By default,
        it is a 0-order model consisting of the frequencies of the letters in
        the input sequences. A different 0-order Markov model or a higher-order
        Markov model can be specified to MEME using the
        <span class="popt">-bfile</span> option.</p>
        <p>The <i>E</i>-value reported by MEME is actually an approximation of
        the <i>E</i>-value of the log likelihood ratio (an approximation is used
        because it is far more efficient to compute). The approximation is based
        on the fact that the log likelihood ratio of a motif is the sum of the
        log likelihood ratios of each column of the motif. Instead of computing
        the statistical significance of this sum (its <i>p</i>-value), MEME
        computes the <i>p</i>-value of each column and then computes the
        significance of their product. Although not identical to the
        significance of the log likelihood ratio, this easier to compute
        objective function works very similarly in practice.</p>

        <p>The motif significance is reported as the <i>E</i>-value of the
        motif. The statistical significance of a motif is computed based on:</p>
        <ol>
          <li>the log likelihood ratio,</li>
          <li>the width of the motif,</li>
          <li>the number of occurrences,</li>
          <li>the 0-order portion of the background model,</li>
          <li>the size of the input sequences, and</li>
          <li>the type of model (<span class="popt">oops</span>,
          <span class="popt">zoops</span>, or
          <span class="popt">anr</span>, which determines the
          number of possible different motifs of the given width and number of
          occurrences).</li>
        </ol>

        <p>MEME searches for motifs by performing Expectation Maximization (EM)
        on a motif model of a fixed width and using an initial estimate of the
        number of sites. It then sorts the possible sites according to their
        probability according to EM. MEME then calculates the <i>E</i>-values of
        the first n sites in the sorted list for different values of n. This
        procedure (first EM, followed by computing <i>E</i>-values for different
        numbers of sites) is repeated with different widths and different
        initial estimates of the number of sites. MEME outputs the motif with
        the lowest <i>E</i>-value.</p>

-->
      </div>

      <h3 id="multiple_alignment_trimming">Multiple Alignment Trimming (Classic objective function only)</h3>
      <div class="indent">
        <p>Once a candidate motif has been found the multiple alignment method
        performs a separate pairwise alignment of the site with the highest
        probability and each other possible site (the alignment includes
        width/2 positions on either side of the sites).</p>

        <p>The pairwise alignments are then combined and the method determines
        the widest section of the motif with no insertions or deletions. If this
        alignment is shorter than <span class="pdat">minw</span>, it tries to
        find an alignment allowing up to one insertion/deletion per motif column.
        This continues (allowing up to 2, 3 ... insertions/deletions per motif
        column) until an alignment of width at least <span class="pdat">minw</span> is found.</p>

        <p>The switches <span class="popt">-nomatrim</span>, 
        <span class="popt">-wg</span>, <span class="popt">-ws</span> and
        <span class="popt">-noendgaps</span> control trimming of
        motifs using the multiple alignment method. Specifying the
        <span class="popt">-nomatrim</span> option causes MEME to skip trimming
        altogether. The <span class="popt">-wg</span> and
        <span class="popt">-ws</span> options set the costs of including gaps in
        the alignments and the <span class="popt">-noendgaps</span> allows gaps
        at the beginning and end to be treated specially.</p>
        
        <p>After trimming, the number of occurrences is then adjusted to
        maximize the motif <i>E</i>-value, and then the motif width is further
        shortened to optimize the <i>E</i>-value.</p>

      </div>

<!--
      <h3 id="branching_search">Branching Search</h3>
      <div class="indent">
        <p>The search for good EM starting points can be improved by using
        branching search.</p>

        <p>Branching search begins with a fixed-sized heap of best EM starts
        identified during the search of subsequences from the dataset. These
        starts are also called "seeds". The fixed-sized heap of seeds is used as
        the "branch_heap" during the first iteration of branching search.</p>

        <p>For each iteration of branching search, all seeds in the current
        branch_heap are considered. All seeds in the ball within hamming
        distance 1 of a given seed are evaluated and added to a new heap. The
        ball of new seeds is generated by mutating each letter of the initial
        seed to each alternative letter in the alphabet.</p>

        <p>After the ball for every branch_heap seed has been evaluated, the
        seeds in the resulting new heap are added to the heap of best EM starts.
        The new heap is then used as the branch_heap for the next iteration of
        branching search.</p>

      </div>
-->

      <h3 id="running_time">Running time on large inputs</h3>
      <div class="indent">
        <p>MEME's running time depends on the type of model you have chosen:
	<span class="popt">oops</span> &lt; <span class="popt">zoops</span> &lt; <span class="popt">anr</span>.
	Running time also depends on the objective function you have chosen:
	<span class="popt">ce</span>,<span class="popt">cd</span> 
          &lt; <span class="popt">de</span>, <span class="popt">se</span> 
	  &lt; <span class="popt">classic</span>
	  &laquo; <span class="popt">nc</span>, 
	for small datasets; <b>the order reverses (except for <span class="popt">nc</span>)
	for very large datasets</b>.
	Running time also is affected by the options
	<span class="popt">-csites</span>,
	<span class="popt">-searchsize</span>,
	<span class="popt">-hsfrac</span>, or
	<span class="popt">-cefrac</span>.  
        Running time is shorter with smaller alphabets
	(e.g., DNA) compared to with larger alphabets (e.g., Protein).  
	Running time increases with the number of motif widths that MEME considers 
	(see the options
	<span class="popt">-w</span>,
	<span class="popt">-maxw</span>,
	<span class="popt">-minw</span>, and
	<span class="popt">-allw</span>).
	The table below gives more details on the relationship between sequence
	dataset size and MEME's running time with large datasets.
	  <table class="dark" style="width:100%">
	    <tr><th>objective function</th><th>running time per motif (for large datasets)</th></tr>
	    <tr>
              <td><span class="popt">classic</span>, <span class="popt">nc</span></td>
              <td>approximately <i>n</i>log(<i>n</i>) for
	        motifs with up to <span class="pdat">csites</span> sites, 
		where <i>n</i> is the size of the sequence file; 
		thereafter, the running time is essentially constant
              </td>
            </tr>
	    <tr>
              <td><span class="popt">de</span>, <span class="popt">se</span></td>
              <td>approximately <i>n</i>log(<i>n</i>) for
		sequence files with up to <span class="pdat">searchsize</span> divided by <span class="pdat">hsfrac</span> letters, 
		where <i>n</i> is the size of the (primary) sequence file;
		thereafter, the running time is essentially constant
              </td>
            </tr>
	    <tr>
              <td><span class="popt">ce</span>, <span class="popt">cd</span></td>
              <td>approximately <i>n</i>log(<i>n</i>) for
		sequence files with up to <span class="pdat">searchsize</span> divided by <span class="pdat">cefrac</span> letters, 
		where <i>n</i> is the size of the (primary) sequence file;
		thereafter, the running time is essentially constant
              </td>
            </tr>
	  </table>
        </p>
        <p>
	The following table shows actual running times for MEME on DNA datasets containing different
	numbers of 100bp sequences on a 4 GHz Intel Core i7 processor with 16GB of memory. 
	MEME was run using either one or 6 processors (option <span class="popt">-p</span>), 
	different models and objective functions 
	(options <span class="popt">-mod</span> and <span class="popt">-objfun</span>), 
	and the option <span class="popt">-revcomp</span>.
        (The parallel version of MEME scales up to about 128 processors.
        Please see <a href="https://academic.oup.com/bioinformatics/article/12/4/303/183600">
        https://academic.oup.com/bioinformatics/article/12/4/303/183600</a>
        for a discussion of the parallel version of MEME.)
        </p>
	  <table class="dark" style="width:100%">
            <tr>
              <th>processors</th>
	      <th>OOPS model</th>
	      <th>ZOOPS model</th>
	      <th>ANR model</th>
            </tr>
            <tr>
	      <td>1</td>
	      <td><img src="images/exp28b.1.oops.png"></td>
	      <td><img src="images/exp28b.1.zoops.png"></td>
	      <td><img src="images/exp28b.1.anr.png"></td>
            </tr>
            <tr>
	      <td>6</td>
	      <td><img src="images/exp28b.6.oops.png"></td>
	      <td><img src="images/exp28b.6.zoops.png"></td>
	      <td><img src="images/exp28b.6.anr.png"></td>
            </tr>
	  </table>
        </p>
      </div>

      <h3 id="examples">Examples</h3>
      <div class="indent">
        <p>The following examples use data files provided in this release of the
        MEME Suite (in the tests directory).</p>
        <ol>
          <li>
          <p>A simple DNA example:</p>
          <div class="example">meme crp0.s -dna -mod oops -pal</div>
          <p>MEME looks for a single motif in the file <code>crp0.s</code> which
          contains DNA sequences in FASTA format. 
	  The <span class="popt">oops</span> site distribution model is used so MEME
          assumes that every sequence contains exactly one occurrence of the
          motif. The palindrome switch is given so the motif model (PSPM) is
          converted into a palindrome by combining corresponding frequency
          columns.  MEME automatically chooses the best width for the motif in
          this example since no width was specified.</p>
          </li>

          <li>
          <p>Searching for motifs on both DNA strands:</p>
          <div class="example">meme crp0.s -dna -mod oops -revcomp</div>
          <p>This is like the previous example except that the
          <span class="popt">-revcomp</span> switch tells MEME to consider both
          DNA strands, and the <span class="popt">-pal</span> switch is absent
          so the palindrome conversion is omitted. When DNA uses both DNA
          strands, motif occurrences on the two strands may not overlap. That
          is, any position in the sequence given in the input sequences may be
          contained in an occurrence of a motif on the positive strand or the
          negative strand, but not both.</p>
          </li>

          <li>
          <p>A fast DNA example:</p>
          <div class="example">meme crp0.s -dna -mod oops -revcomp -w 20</div>
          <p>This example differs from example 1) in that MEME is told to only 
          consider motifs of width 20. This causes MEME to execute about 10
          times faster. The <code>-w</code> switch can also be used with protein
          datasets if the width of the motifs is known in advance.</p>
          </li>

          <li>
          <p>Using a higher-order background model:</p>
          <div class="example">meme INO_up800.s -dna -mod anr -revcomp -bfile yeast.nc.6.freq</div>
          <p>In this example we use <code>-mod anr</code> and <code>-bfile
            yeast.nc.6.freq</code>. This specifies that</p>
          <ol type="a">
            <li>the motif may have any number of occurrences in each sequence, and,
            <li>the Markov model specified in yeast.nc.6.freq is used as the 
            background model. This file contains a fifth-order Markov model for
            the non-coding regions in the yeast genome.</li>
          </ol>
          <p>Using a higher order background model can often result in more
          sensitive detection of motifs. This is because the background model
          more accurately models non-motif sequence, allowing MEME to
          discriminate against it and find the true motifs.</p>
          </li>

          <li>
          <p>A simple protein example:</p>
          <div class="example">meme lipocalin.s -mod oops -maxw 20 -nmotifs 2</div>
          <p>The <code>-dna</code> switch is absent, so MEME assumes the file
          <code>lipocalin.s</code> contains protein sequences. MEME searches for
          two motifs each of width less than or equal to 20. (Specifying
          <code>-maxw 20</code> makes MEME run faster since it does not have to
          consider motifs longer than 20.) Each motif is assumed to occur in
          each of the primary sequences because the <span class="popt">oops</span> distribution 
	  model is specified.</p>
          </li>

          <li>
          <p>Another simple protein example:</p>
          <div class="example">meme farntrans5.s -mod anr -maxw 40 -maxsites 50</div>
          <p>MEME searches for a motif of width up to 40 with up to 50
          occurrences in the entire set of input sequences. The <span class="popt">anr</span> 
	  site distribution model is
          specified, which allows each motif to have any number of occurrences
          in each sequence. This dataset contains motifs with multiple repeats
          of motifs in each sequence. This example is fairly time consuming due
          to the fact that the time required to initialize the motif probability
          tables is proportional to <span class="pdat">maxw</span> &times;
          <span class="pdat">maxsites</code>.</p>
          </li>

          <li>
          <p>A much faster protein example:</p>
          <div class="example">meme farntrans5.s -mod anr -w 10 -maxsites 30 -nmotifs 3</div>
          <p>This time MEME is constrained to search for three motifs of width
          exactly ten. The effect is to break up the long motif found in the
          previous example. The <span class="popt">-w</span> switch forces
          motifs to be <b>exactly</b> ten letters wide. This example is much faster
          because, since only one width is considered, the time to build the
          motif probability tables is only proportional to
          <span class="pdat">maxsites</span>.</p>
          </li>

          <li>
          <p>Splitting the sites into three:</p>
          <div class="example">meme farntrans5.s -mod anr -maxw 12 -nsites 24 -nmotifs 3</div>
          <p>This forces each motif to have 24 occurrences, exactly, and be up
          to 12 letters wide.</p>
          </li>

          <li>
          <p>A larger protein example with <i>E</i>-value cutoff:</p>
          <div class="example">meme adh.s -mod zoops -nmotifs 20 -evt 0.01</div>
          <p>In this example, MEME looks for up to 20 motifs, but stops when a
          motif is found with <i>E</i>-value greater than 0.01. Motifs with
          large <i>E</i>-values are likely to be statistical artifacts rather
          than biologically significant.</p>
          </li>

        </ol>
      </div>
    </div>

    <h3 id="citing">Citing</h3>
    <div class=indent id="citation"><script>print_citation("citation", "MEME");</script></div>

    <script src="js/manual.js"></script>

  </body>
</html>
