<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>motiph - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2>motiph</h2>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>motiph [options] &lt;alignment file&gt; &lt;tree file&gt;
        &lt;motif file&gt;</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>Scan a DNA multiple sequence alignment for occurrences of given
      motifs, taking into account the phylogenetic tree relating the
      sequences. Any sequences containing nothing but gaps are removed from
      the alignment. 
      <code>motiph</code> supports several evolutionary models that it uses to compute
      the score for aligning the motif to each possible region of the
      multiple alignment.
      Read about the <a href="#algorithm">algorithm</a> for more details.</p>
    </div>

    <h3>Input</h3>
    <div class="indent">
      <h4><span class='popt'>&lt;alignment file&gt;</h4>
      <div class="indent">
        <p>The name of a file containing a DNA multiple alignment in <a 
        href="clustalw-format.html">ClustalW</a> format. Alternatively, if the
        <span class="popt">--list</span> option is used, this file may contain
        a list of alignment files. In this case each of the alignment files
        will be scanned for each of the motifs.</p>
      </div>

      <h4><span class='popt'>&lt;tree file&gt;</h4>
      <div class="indent">
        <p>The name of a file containing a phylogenetic tree in 
        <a href="http://evolution.genetics.washington.edu/phylip/newicktree.html"
        >Phylip Newick</a> format. This tree may contain additional species
        not represented in the alignment.</p>
      </div>

      <h4><span class='popt'>&lt;motif file&gt;</h4>
      <div class="indent">
        <p>The name of a file containing <a href="meme-format.html">MEME formatted motifs</a>.</p>
      </div>
    </div>

    <h3>Output</h3>
    <div class="indent">
      <p>Output is created as a table in HTML (.html) and text (.txt) formats. 
      Each line in the table represents a match of the motif to a "window" in the alignment.
      Matches to the motif and its reverse complement are listed separately.
      The output columns contain the motif name, the alignment name, the start position of the match
      in the alignment, the end position of the match,
      the match score, the match <i>p</i>-value, the match <i>q</i>-value, 
      and the matched portion of the first sequence in the alignment, respectively.
      </p>
      <p>
      Output is also created in 
      <a href="http://gmod.org/wiki/GFF2">GFF</a> format (.gff)
      and in CISML (.xml) format.
      </p>
    </div>
    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behavior</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">--bg</td><td><span class="pdat">mutation rate</span></td>
            <td>The mutation rate for sites in the background model.</td>
            <td>A background mutation rate of 1 is used.</td>
          </tr>
          <tr>
            <td class="popt">--bfile</td><td><span class="popt">background file</span></td>
            <td>Read background frequencies from
            <span class="popt">background file</span>. The file should be in 
            <a href="bfile-format.html">MEME background file format</a>.
            If the argument is the keyword <code>motif-file</code>, then the frequencies will be taken from
            the motif file. If both strands are being scored, the background model 
	    will be modified by averaging the frequencies of letters and their reverse complements.</td>
            <td>Use frequencies embedded in the application from
            an old version of the NCBI non-redundant DNA database.</td>
          </tr>
          <tr>
            <td class="popt">--fg</td><td><span class="pdat">mutation rate</span></td>
            <td>The mutation rate for sites in the foreground model(s).</td>
            <td>A foreground mutation rate of 1 is used.</td>
          </tr>
          <tr>
            <td class="popt">--gap</td><td><span
                class="popt">skip</span>|&#8203;<span
                class="popt">fixed</span>|&#8203;<span
                class="popt">wildcard</span>|&#8203;<span
                class="popt">minimum</span>|&#8203;<span
                class="popt">model</span></td>
            <td>Specifies the gap handling strategy.
              <table class="dark">
                <tr>
                  <th>Value</th><th>Name</th>
                </tr>
                <tr>
                  <td class="popt">skip</td>
                  <td>Skip those sites where any position in the alignment
                    window contains a gap.</td>
                </tr>
                <tr>
                  <td class="popt">fixed</td>
                  <td>Sites containing gaps are assigned a fixed score,
                    specified by <span class="popt">--gap-cost</span>.</td>
                </tr>
                <tr>
                  <td class="popt">wildcard</td>
                  <td>The gap character matches any base, and the score is the
                    product of the corresponding probabilities. </td>
                </tr>
                <tr>
                  <td class="popt">minimum</td>
                  <td>The gap character is assigned the score corresponding to
                    the least likely letter at the given position.</td>
                </tr>
                <tr>
                  <td class="popt">model</td>
                  <td>Use model-specific gap handling. Currently, the only
                    model that supports this is <span class="popt">f81_gap</span>.</td>
                </tr>
              </table>
            </td>
            <td>Behaves as if <span class="popt">--gap skip</span> had been specified.</td>
          </tr>
          <tr>
            <td class="popt">--gap-cost</td><td><span class="pdat">cost</span></td>
            <td>Specifies the costs for gaps when using the
              <span class="popt">fixed</span> gap handling strategy.</td>
            <td>A gap cost of zero is used.</td>
          </tr>
          <tr>
            <td class="popt">--hb</td><td></td>
            <td>Use the Halpern-Bruno modification to the evolutionary model.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">--list</td><td></td>
            <td>Treat the second required input as a list of alignments,
              rather than a single alignment.</td>
            <td>The second required argument is treated as a single alignment.</td>
          </tr>
          <tr>
            <td class="popt">--max-stored-scores</td><td><span class="pdat">storage</span></td>
            <td>Set the maximum number of scores that will be stored.
            Keeping a complete list of scores may exceed available memory.
            Once the number of stored scores reaches the maximum allowed,
            the least significant 50% of scores will be dropped.
            In this case, the list of reported motifs may be incomplete and 
            the q-value calculation will be approximate.</td>
            <td>The maximum number of stored matches is set to 100,000.</td>
          </tr>
          <tr>
            <td class="popt">--model</td><td><span
                class="popt">single</span>|&#8203;<span
                class="popt">average</span>|&#8203;<span
                class="popt">jc</span>|&#8203;<span
                class="popt">k2</span>|&#8203;<span
                class="popt">f81</span>|&#8203;<span
                class="popt">f84</span>|&#8203;<span
                class="popt">hky</span>|&#8203;<span
                class="popt">tn</span></td>
            <td>The evolutionary model to use.
              <table class="dark">
                <tr>
                  <th>Value</th><th>Name</th><th>Description</th>
                </tr>
                <tr>
                  <td class="popt">single</td><td>Single Score</td>
                  <td>score first sequence: compute standard log-odds score of
                    first sequence in the alignment; ignores tree but does NOT
                    remove gaps.</td>
                </tr>
                <tr>
                  <td class="popt">average</td><td>Average Score</td>
                  <td>compute average of standard log-odds score of aligned
                    sites.</td>
                </tr>
                <tr>
                  <td class="popt">jc</td><td>Jukes-Cantor</td>
                  <td>equilibrium base frequencies are all 1/4; the only free
                    parameter is the mutation rate.</td>
                </tr>
                <tr>
                  <td class="popt">k2</td><td>Kimura 2-parameter</td>
                  <td>equilibrium base frequencies are all 1/4; the free
                    parameters are the mutation rate and the
                    transition/transversion rate ratio.</td>
                </tr>
                <tr>
                  <td class="popt">f81</td><td>Felsenstein 1981</td>
                  <td>equilibrium base frequencies are taken from the
                    alignment; the only free parameter is the mutation rate.</td>
                </tr>
                <tr>
                  <td class="popt">f84</td><td>Felsenstein 1984</td>
                  <td>equilibrium base frequencies are taken from the
                    alignment; the free parameters are the mutation rate and
                    the transition/transversion rate ratio. The ratio of
                    purine-purine to pyrimidine-&gt;pyrimidine transitions is
                    assumed to be 1.</td>
                </tr>
                <tr>
                  <td class="popt">hky</td><td>Hasegawa-Kishino-Yano</td>
                  <td>equilibrium base frequencies are taken from the
                    alignment; the free parameters are the mutation rate and
                    the transition/transversion rate ratio. The ratio of
                    purine-purine to pyrimidine-pyrimidine transitions is
                    assumed to be equal to the ratio of purines to
                    pyrimidines.</td>
                </tr>
                <tr>
                  <td class="popt">tn</td><td>Tamura-Nei</td>
                  <td>equilibrium base frequencies are taken from the
                    alignment; the free parameters are the mutation rate, the
                    transition/transversion rate ratio, and the ratio of
                    purine-purine transitions to pyrimidine-pyrimidine 
                    transitions.</td>
                </tr>
              </table>
              A description of the <span class="popt">f81</span> model is
              available in chapter 13 of <i>Statistical Methods in
              Bioinformatics</i> by Ewens and Grant. The other models are
              described in chapters 9 and 13 of <i>Inferring Phylogenies</i>
              by Felsenstein. 
            </td>
            <td>Behaves as if <span class="popt">--model f81</span> was specified.</td>
          </tr>
          <tr>
            <td class="popt">--motif</td><td><span class="pdat">ID</span></td>
            <td>Use only the motif identified by <span class="pdat">ID</span>.
              This option may be repeated.</td>
            <td>All motifs are used.</td>
          </tr>
          <tr>
            <td class="popt">--no-qvalue</td><td></td>
            <td>Do not compute a <i>q</i>-value for each <i>p</i>-value. The
              <i>q</i>-value calculation is that of Benjamini and Hochberg (1995).</td>
            <td>The program computes <i>q</i>-values.</td>
          </tr>
          <tr>
            <td class="popt">--norc</td><td></td>
            <td>Do not score the reverse complement DNA strand.</td>
            <td>Both strands are scored.</td>
          </tr>
          <script>
            wropt("all-o", {"ddash": true, "dir": "motiph_out"});
            wropt("all-oc", {"ddash": true, "dir": "motiph_out"});
          </script>
          <tr>
            <td class="popt">--output-pthresh</td><td><span class="pdat"><i>p</i>-value threshold</span></td>
            <td>The <i>p</i>-value threshold for displaying search results. If
              the <i>p</i>-value of a match is greater than this value, then
              the match will not be printed. If both the
              <span class="popt">--output-pthresh</span> and
              <span class="popt">--output-qthresh</span> options appear on the
              command line, whichever appears later on the command line will
              be applied.</td>
            <td>The <i>p</i>-value threshold is set to 1e<sup>-4</sup>.</td>
          </tr>
          <tr>
            <td class="popt">--output-qthresh</td><td><span class="pdat"><i>q</i>-value threshold</span></td>
            <td>The <i>q</i>-value threshold for displaying search results. If
              the <i>q</i>-value of a match is greater than this value, then
              the match will not be printed.</td>
            <td>No <i>q</i>-value threshold is applied.</td>
          </tr>
          <tr>
            <td class="popt">--pseudocounts</td><td><span class="pdat">pseudocount</span></td>
            <td>A pseudocount to be added to each count in the motif matrix,
              weighted by the background frequencies for the nucleotides
              (Dirichlet prior), before converting the motif to
              probabilities.</td>
            <td>A pseudocount of 0.1 is added.</td>
          </tr>
          <tr>
            <td class="popt">--pur-pyr</td><td><span class="pdat">ratio</span></td>
            <td>The ratio of the purine transition rate to pyrimidine
              transition rate. This parameter is used by the Tamura-nei
              model.</td>
            <td>The ratio is set to 1.0.</td>
          </tr>
          <tr>
            <td class="popt">--seed</td><td><span class="pdat">n</span></td>
            <td>Seed for random number generator.</td>
            <td>The seed is set to the current time.</td>
          </tr>
          <tr>
            <td class="popt">--text</td><td></td>
            <td>Limits output to plain text sent to standard out. This mode
              allows the program to search an arbitrarily large database,
              because results are not stored in memory.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">--transition-transversion</td><td><span class="pdat">ratio</span></td>
            <td>The ratio of the transition rate to the transversion rate.
              This parameter is used by the Kimura 2-parameter, F84, HKY, and
              Tamura-nei models.</td>
            <td>The ratio is set to 0.5.</td>
          </tr>
          <script>wropt("all-verbosity", {"ddash": true});</script>
        </tbody>
      </table>
    </div>

    <h3 id="algorithm">Algorithm</h3>
    <div class="indent">
      <p>
      The algorithm is as follows:
      </p>
      <pre class="indent">
A = a multiple alignment
T = a phylogenetic tree
M = a motif frequency matrix
B = the background frequencies of the four bases in A
U = a background evolutionary model with equilibrium frequencies B

// Build evolutionary models.
for j in positions of M {
  E[j] = an evolutionary model with equilibrium frequencies 
    from the jth position of M
}

// Build matrix of all possible scores
for i in all possible alignment columns {
  for j in positions of M {
    foreground = site_likelihood(E[j], A[:][i], T)
    background = site_likelihood(U, A[:][i] T)
    score_matrix[i][j] = log_2(foreground / background)
  }    
}
// Calculate p-values of all possible scores for motif sized windows
// windows in the alignment.
pvalues = get_pv_lookup(score_matrix, B)

// Calculate score for each motif sized window in the alignment.
for i in columns of A {
  score = 0
  for j in positions of M {
    index = calculate the index of A[:][i + j] in the array
      of all possible alignment columns
    score = score + score_matrix[index][j]
  }    
  print pvalues[score]
}
      </pre>

      <p>The core of the algorithm is a routine (<code>site_likelihood</code>)
      for scoring a particular column of the multiple alignment using a
      given evolutionary model and a given phylogenetic tree.  The alignment
      site provides the observed nucleotides at the base of the tree, and we
      sum over the unobserved nucleotides in the rest of the tree,
      conditioning on the equilibrium distribution from the evolutionary
      model at the root of the tree (Felsenstein Pruning Algorithm).  The
      tree must be a maximum likelihood tree, of the kind generated by DNAml
      from Phylip or by FastDNAml.  Branch lengths in the tree are converted
      to conditional probabilities using the specified evolutionary
      model.</p>

    </div>

    <!-- 
    Note this bit was in the original, I don't see a good way to include it
    because large bits don't make much sense.


    <b>Warning messages:</b> None

    <b>Bugs:</b>

    <ul>

    <li>
    F84 and Tamura-Nei models fail trying to take the log of a negative number.</li>
    <li>
    Allow local realignment, a la Monkey.</li>
    <li>
    The <code>--motif</code> option should allow multiple motifs
    to be selected from the motif file.
    </li>
    <li>
    Print motif consensus as part of feature properties.</li>
    <li>
    Allow reading of a column frequency file; this requires
    writing a stand-alone to create such a file and establishing
    a format such as:
    <pre>
       # species_list_1
       AAA f_AAA
       ...
       TTT f_TTT
       # species_list_2
       AAAAA f_AAAAA
       ...
       TTTTT f_TTTTT
       ...
    </pre>

    -->

  </body>
</html>

