<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Release Notes - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>

    <h2>MEME Suite Release Notes</h2>

    <p>
    <hr>
      <b>MEME version 5.4.1 (August 25, 2021)</b>
      <ul>
        <li><b>Bug Fixes</b></li>
	<ul>
          <li>Fixed bug in MEME HTML output files that caused the "Motif Locations"
	  download buttons to fail when the file was on the server.</li>
        </ul>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.4.0 (August 22, 2021)</b>
      <ul>
        <li><b>New Features and Enhancements</b></li>
          <ul>
	    <li>XSTREME: Combined motif discovery and enrichment analysis.
		Combines the results of ab initio discovery using MEME and STREME
		with motif enrichment analysis using SEA.  Groups similar
		ab initio motifs with each other and with known motifs.
		XSTREME is similar to MEME-CHIP, but is more appropriate when the motifs
		may not tend to be enriched in the centers of the input sequences, and
		works well with protein sequences and nucleotide sequences.
	    </li>	
            <li>SEA: Simple Enrichment Analysis of motifs. Reports known motifs that are
		enriched in your sequences relative to your control sequences or to
		shuffled versions of your sequences.  Replaces and enhances ENR.
	    </li>	
	    <li>STREME: 
              <ul>
		<li>Added options --thresh and --evalue to allow STREME to stop
		  looking for motifs based on their <i>E</i>-values rather than their
		  <i>p</i>-values.  (The --thresh option replaces the --pvt option,
		  which is still available for backward compatability.)
                </i>
                <li>Added motif <i>E</i>-value column in HTML output and change MEME format
		  motif header to E= in streme.txt.
		</li>
                <li>Added "Positional Distribution" plots to HTML output.
		  These plots show how the sites of each motif are distributed
		  within the input sequences.  Alignment of sequences for the
		  plots is controlled by the new --align option.
	        </li>	
		<li>Added output of a file (sequences.tsv) containing the names of each
		    input sequence that contains a site of a discovered motif.
		</li>	
                <li>Added "Matches per Sequence" plots to HTML output.
		  These plots show the distribution of the percentage of sequences
		  with different numbers of matches to each significant motif.
	        </li>	
		<li>The Sites column in the HTML now shows both the number of
		  percentage of positive sequences with a site as well as the number of sequences.
		</li>	
		<li>The meme.txt file of motifs created by STREME now shows
		    the Score (Training Set <i>p</i>-value) rather than the Test Set <i>p</i>-value
		    when there was no Test Set due to the input containing too few sequences.
		    This is indicated in the motif "letter-probability matrix" line, where "P=" is 
		    now replaced by "S=".
		</li>	
              </ul>
            </li>
            <li>MEME: Will now search for up to 1000 motifs if -evt is given
		and you don't specify -nmotifs.  Previously, it would only search
		for (up to) 1 motif.
            </li>
            <li>AME: 
              <ul>
                <li>Added the motif alternate ID to the sequence.tsv output.  The
		  new column in the TSV file is named "motif_ALT_ID".
		<li>Added option --text to just send TSV output to standard output.
              </ul>
            </li>
	    <li>fasta-holdout-set: New utility for splitting primary (positive) and control
		(negative) sequence sets into a main and holdout sets.  Input is
		one or two FASTA sequence files.  Output is two or four FASTA sequence
		files.  If no control sequences are given, they are created by shuffling
		the primary sequences.  By default, each FASTA file (primary and control) 
		is split into two FASTA files containing 90% ("main file") 
		and 10% (holdout file) of the input sequences, respectively.  If either of the
		holdout files (primary or control) would contain fewer than 10 sequences, 
		no holdout files are created and all sequences are put in the main files.
		This utility was create for use by XSTREME.
            </li>
            <li>configure.ac: Added --with-mpi-nprocs to allow MPI jobs (e.g., meme) to
		use multiple processors when run as web services.
            </li>
            <li>Consolidated all citations in doc/js/citation.js to remove redundancy
		and to prevent things from getting out of sync.
            </li>
	    <li>Added check for multiple cores to speed up "make test".  Adds dependency
		on perl module "Sys::Info".
 	    </li>
	    <li>Simplified Tomtom web-server implementation.  Removed the "run-immediate"
		option; all Tomtom jobs now go through the "short queue".
 	    </li>
          </ul>
        <li><b>Bug Fixes</b></li>
          <ul>
	    <li>AME: fixed bug in webserver, was ignoring advanced options "uniform" and "Model in motif".</li>
	    <li>CentriMo: fixed bug with "Download EPS (for publication)" button not working on some browsers.</li>
	    <li>GOMO: fixed links from GO terms to their descriptions at "amigo".</li>
	    <li>MAST: fixed bug in usage message.</li>
            <li>MEME:
              <ul>
                <li>Fixed core dump caused with the "anr" model by very long sequences.
                <li>The -searchsize option now also limits the length of the longest sequence allowed,
                 unless you specify "-searchsize 0".
              </ul>
            </li>
	    <li>STREME: 
              <ul>
                <li>Fixed failure to honor "T=U" in custom RNA alphabets.</li>
                <li>Fixed error when maximum motif width is larger than average sequence length.</li>
                <li>Fixed printing of motifs in MEME format when there were no test positives or negatives
		but there was a hold-out set.
              </ul>
	    <li>fasta-get-markov: was not deleting its temporary files /tmp/bfile.*.</li>
            <li>All web servers: "Clear Input" button was no clearing "Hidden Modifications" warning on Advanced options menu header.</li>
            <li>Scripting Access: Fixed broken links to sample scripts.</li>
          </ul>
        </li>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME Suite version 5.3.3 (February, 2021)</b>
      <ul>
        <li><b>Bug Fixes</b>
	<ul>
	  <li>INSTALL -- fixed bug that prevented building of embedded version of libxslt.</li>
	</ul>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME Suite version 5.3.2 (February, 2021)</b>
      <ul>
        <li><b>Bug Fixes</b>
	<ul>
	  <li>MEME -- fixed problem that prevented downloading block diagram PDFs.</li> 
	  <li>MEME -- RNA and Protein motifs may now be submitted to Tomtom via the Submit/Download menu
		in the MEME HTML output.
	  <li>configure.ac -- changed default SITE URL to "https://meme-suite.org/meme".</li>
	  <li>sample outputs -- updated all sample outputs with new site URL.</li>
	  <li>Installation guide -- Simplified Quick Install instructions.</li>
	  <li>INSTALL -- updated Quick Install instructions.</li>
	  <li>various -- removed hard-coded references to meme-suite.org.
          <li>menu -- fix URL of menu header when the menu is on a server page
	</ul>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.3.1 (January, 2021)</b>
      <ul>
        <li><b>New Features and Enhancements</b>
	  <ul>
	    <li>MEME-CHIP web server: Modified Tomtom to send short jobs to a short job queue, 
              rather executing them directly</li>
	    <li>Changed the time outputs from CPU time to ELAPSED time in MEME, STREME, MAST and TOMTOM so that the "-time"
	      option works even when the job gets less than 100% of CPU time.</li>
	    <li>Added configure switches for setting the maximum memory and run time of jobs
	      submitted via the web.</li>
	    <li>Added --totallength = maximum memory / 1000 to input to STREME when
	      submitted via the web.  Same for STREME called from MEME-ChIP submitted via the web.
	      This will prevent STREME from using too much memory.</li>
	  </ul>
	</li>
        <li><b>Bug Fixes</b>
          <ul>
            <li>MEME web application: 1-order model was missing from background model menu.</li>
	    <li>AME web application: the uniform background was broken,
    	      and 1-order mode menu choices were being displayed when they should not be.</li>
          </ul>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.3.0 (November, 2020)</b>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
	    <li>MEME-CHIP: 
              <ul>
                <li>STREME now replaces DREME in the MEME-CHIP pipeline.
		<li>The maximum allowed motif width is reduced to 30 (from 300).
		<li>The motif width search range now defaults to [6,15] (from [6,30]).
		<li>The motif width options are now -minw and -maxw (were -meme-minw and -meme-maxw),
		and the minimum and maximum widths apply to motifs found by both MEME and STREME.
		<li>The default for the Markov model order option (-order) is raised from 1 to 2.
		<li>The Markov model is now created from the control sequences if given, not the primary sequences.
		<li>The -norc option is replaced by the -dna2rna option, which causes 
		STREME and MEME to output RNA motifs given DNA sequence.  CentriMo and SpaMo will display
		the DNA sequences as RNA; FIMO shows the sequences as DNA even when the motifs are RNA.
		<li>SpaMo and FIMO are now only run on motifs discovered by MEME and STREME (not on motifs
		from the motif database(s), which saves lots of time.  
		The user can perform a separate SpaMo analysis using any motif later, if desired.
              </ul>
            </li>
	    <li>MEME-CHIP web server: 
              <ul>
                <li>The default background model order is now 2nd order (was 1st order).
		<li>The width range option is now under Universal options (was under MEME options).
		<li>The Universal option "Scan both DNA strands" has been removed, and now a checkbox
		appears when you input DNA sequences: "Convert DNA sequences to RNA?".
		<li>You may now specify that MEME find 0 motifs, and STREME find 0 motifs,
		so only CentriMo will run, followed by a clustering of enriched motifs.
              </ul>
            </li>
	    <li>STREME: Added the "Sites" column to the HTML output to make it more like MEME output.
            </li>
          </ul>
        </li>
        <li><b>Bug Fixes</b>
          <ul>
	    <li>STREME: fixed bugs in links to TEXT and XML outputs in the template HTML output.
	    <li>STREME: fixed rare bug when palindrome went past end of seed.
	    <li>STREME: fixed globals in header that were not declared extern.
	    <li>ENR: fixed typo in documentation.
	    <li>MEME-CHIP webserver: fixed bug where the the control sequence disappeared after page refresh.
	    <li>General: Added charset=UTF-8 to etc/job_status.tmpl to stop Firefox console warnings.
	    <li>General: Fixed duplicate variable declarations that caused compile failure with GCC v10.
	    <li>General: Added -fno-common to CFLAGS so that variables not declared
		"extern" in .h files will cause an error with MacOS compilers, as
		the do with GCC v10.
	    <li>General: Fixed bug in SameSite attribute of Google Analytics cookies.
          </ul>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.2.0 (October, 2020)</b>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
	    <li>STREME: The new <a href="streme.html">STREME</a> motif discovery algorithm is designed to replace DREME 
		and to complement MEME.  STREME finds more accurate motifs than DREME, finds
		more subtle motifs, can find much wider motifs (up to 30 wide vs. 8 wide), 
		and can quickly process very large sequence datasets 
		(e.g., 4,000,000bp DNA input in 100 seconds/motif).
		STREME provides <b>accurate estimates</b> of the statistical significance
		of the motifs it discovers.
		Like MEME, STREME works with a wide range of sequence alphabets: DNA, RNA,
		protein and custom (user-specified) alphabets.  Like DREME, STREME can
		work with a single input set of sequences, or you can provide a set
		of control sequences for discriminative motif discovery.
		<br>
		STREME is available both via its webserver and as a command-line program
		if you download the MEME Suite software.
	    <li>ENR: The new <a href="enr.html">ENR</a> (Motif ENRichment Analysis) algorithm applies the same objective 
                function used by the STREME motif discovery algorithm to measure the enrichment 
                of motifs large sequence datasets. It is useful for comparing the discriminative 
		power of motifs found by STREME and other motif discovery tools.
		The input to ENR is one or two sets of sequences. 
                The control sequences should have approximately the same length distribution as the 
                primary sequences. If you do not provide a control set, the program shuffles the 
                primary set to create a control set. The program uses Fisher's Exact Test to 
                determine significance of each motif found in the positive set as compared 
                with its representation in the control set, using a significance threshold 
                that may be set on the command line.
	    <li>fasta-grep: Now -erase now works with proteins (as well as DNA) replacing 
		matching words with 'X's.</li>
	    <li>fasta-shuffle-letters: New option -fix allows shuffling
		while preserving the position of a selected character.
		This is useful to allow 0-order (-kmer 1) shuffling of masked sequences
		while preserving runs of the masked character.</li>
            <li>MEME motif text format: Now allows "P= &lt;motif p-value&gt;" in place of
		"E= &lt;motif p-value&gt;" on the letter-probability matrix line.
          </ul>
        </li>
        <li><b>Documentation:</b>
	  <ul>
	    <li><a href="streme.html">STREME Manual</a>.</li>
	    <li><a href="streme-tutorial.html">STREME Tutorial</a>.</li>
	    <li><a href="enr.html">ENR Manual</a>.</li>
	  </ul>
        </li>
	<li><b>Bug fixes:</b>
	  <ul>
            <li>CentriMo webserver: fixed the Advanced option "Choose the match score threshold; optimize score" button.
	    <li>MAST: fixed -mev so that it works.  Also works when the motif file
		specifies the motif p-value rather than E-value.</li>
	  </ul>
        </li>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.1.1 (February 1, 2020)</b>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
	    <li>The MEME Suite build will now detect whether the version
		of Python being used is version 2 or version 3, and generate
		the appropriate version of tools such as Dreme.</li>
	    <li> Tomtom: Added limit on the query width to prevent
		memory overrun.  TOMTOM_MAX_QUERY_WIDTH is set to 100 by
		default but can be changed by installer via src/user.h.
		Also, added a <code>-time</code> option to limit the CPU
		time used by a Tomtom job.
            </li>
            <li>Tomtom webserver: Set limit (default 3) on number of
		concurrent "run immediately" Tomtom jobs that can run, and limit
		such jobs to 1 minute of CPU time.
		This prevents exceeding the server's memory.
	     </li>
          </ul>
        <li><b>Documentation:</b>
	  <ul>
	    <li>The web application now records basic usage information to 
	      Google Analytics</li>
	    <li>transfac2meme: Fixed the documentation pertaining to TRANSFAC-like format
		and added the -use_name option.  Updated the motif_conversion.html
		document regarding transfac2meme and the TRANSFAC-like motif format.
	    </li>
	  </ul>
        </li>
	<li><b>Bug fixes:</b>
	  <ul>
	    <li>Glam2: Fixed the HTML output so that the "letter-probability" motif columns
		sum correctly to 1.  This bug caused Glam2 motif not to work with other MEME Suite 
		programs such as Tomtom.
            </li>
	    <li>transfac2meme: Improved the usage message.
            </li>
	  </ul>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.1.0 (October 2019)</b>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
	    <li>CisMapper: removed from the MEME-Suite; has been replaced by <b>T-Gene</b>.
            </li>
	    <li>T-Gene: New command line program for improved prediction of the gene targets of transcription factors
                from BED files containing the genomic locations of possible regulatory elements (typically,
		a bed file of ChIP-seq peaks).  T-Gene's only other required input is a gene annotation file,
		and computes a statistical, distance-based score for each potential regulatory link
		between a locus in the BED file and a transcription start site (TSS) of a transcript in the annotation file. 
		T-Gene can also use a Tissue Panel containing histone modification data and gene expression
		data for a set of tissue or cell types.  It then computes a statistical score for each
		link that combines distance with the correlation in histone state at the locus and expression
		of the gene, achieving higher accuracy than methods based only on distance or correlation.
		T-Gene also includes a link to the <b>closest</b> TSS to each locus (putative
		regulatory element) element, and such links have much higher predictive accuracy in 
		empirical studies.  T-Gene computes <i>p</i>-values for the observed expression/histone correlation by
		generating a null model based on shuffling the order of the expression tissues
		relative to the histone tissues.  T-Gene computes the <i>p</i> value for the
		distance between an RE and a TSS assuming a uniform distribution on the distance:<br>
		&nbsp;&nbsp;&nbsp;&nbsp;<tt><i>p</i>-value = (distance + RE width) / (maximum distance + RE width)</tt>.<br>
		T-Gene computes a score that combines Correlation and Distance (<b>CnD <i>p</i>-value</b>)
		by computing the <i>p</i>-value of the product of the correlation and distance <i>p</i>-values.
		Finally, T-Gene computes the Q-value of the CND <i>p</i>-value, which is defined as the
		minimum false discovery rate (FDR) required to consider this link statistically significant.
		All regulatory links predicted by T-Gene are labeled as to whether they are either
		"closest-TSS" or "closest-locus" links.
	    </li>
	    <li>T-Gene Web Server: added, featuring tissue panels for Human and Mouse and Genomes
		for 8 model organisms: Arabidopsis, Worm, Zebrafish, Fly, Human, Mouse, Rat, and Yeast.
            </li>
            <li>The MEME HTML output now allows you to generate an image of the 
                <b>motif location diagram</b> as either 1) a PDF image file suitable for use in publications, 
		or, 2) an SVG (Scalable Vector Graphics) image file suitable for inclusion in HTML documents.
            </li>
            <li>CentriMo now lists the number of reported regions in the control sequences
                containing the best match to a motif, and the number of motif matches in the control scoring 
	        higher than the score threshold.
            </li>
          </ul>
        </li>
        <li><b>Documentation:</b>
	  <ul>
	    <li> T-Gene manual.
	    <li> T-Gene tutorial.
	  </ul>
        </li>
	<li><b>Bug fixes:</b>
	  <ul>
            <li>Fixed integer overflow on large sequences in <code>fasta-shuffle-letters</code></li>
            <li>SpaMo HTML output file: added missing '.eps' extension to Alignment Logo download file.
          </ul>
        </li>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.0.5 (March 2019)</b>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
	    <li>CisMapper: Now part of the MEME Suite.  CisMapper is useful
		for predicting regulatory links between regulatory regions (chromosome
		locations) and genes.  Typical uses include predicting regulatory links
		from transcription factor ChIP-seq peaks, provided as a BED file.
		CisMapper can be run from its new webserver page, or from the
		command line.  
		CisMapper uses gene expression and histone modification
		data from a panel of tissues in a given organisms.
		The CisMapper webserver currently provides Expression & Histone panels
		only for hg19 (human) and mm9 (mouse).
	    </li>
	    <li>MoMo: Upgraded HTML output to use JSON like other MEME Suite 
		programs, and added help bubbles,
		improving its layout and readability.  
	    </li>
	    <li>MoMo: Added PERL-style regular expressions for each 
		discovered motif in the HTML, TSV and text (MEME) outputs.
	    </li>
	    <li>MoMo: Simplified and improved documentation of MoMo's
		output format.
	    </li>
	    <li>MoMo Web Server: Added a gzipped tar file of the results
		because the HTML file is not stand-alone--it requires the 
		log PNG files.</li>
          </ul>
	</li>
        <li><b>Documentation:</b>
	  <ul>
	    <li>Added the CisMapper documentation: the manual page and the tutorial page.
	  </ul>
        </li>
	<li><b>Bug fixes:</b>
	  <ul>
	    <li>Fixed MoMo bug: motif logo files contained square brackets, which didn't work over
		the web; now they are replace by 'b' for '[' and 'd' for ']'.</li>
	    <li>Fixed MoMo bug: MoDL algorithm could result in divide by zero when calculating
		the fold change.</li>
	    <li>Fixed MAST bug: could crash if FASTA sequence descriptor line contained non-ASCII character;
		now replaces them with underscores.</li>
	    <li>Fixed FIMO bug: could crash if FASTA sequence header line contained non-ASCII character;
		now replaces them with underscores.</li>
	  </ul>
    </ul>
    </p>
  
    <p>
    <hr>
      <b>MEME version 5.0.4 (January 2019)</b>
      <ul>
	<li>
	  Bug fixes:
	  <ul>
	    <li>Fixed CentriMo bug: crashed if a motif was as wide or wider than the sequences; now it skips the motif and prints a warning.</li>
            <li>Fixed AMA/GoMO bug: could crash if the motif was very wide.</li>
            <li>Fixed AME bug: could crash if there was only one sequence in the input.</li>
            <li>Fixed SpaMo bug: memory leak would exhaust memory with big input files.</li>
            <li>Fixed FASTA bug: read_one_fasta was failing to leave room for trailing null in read buffer.</li>
            <li>Fixed Tomtom bug: would crash if primary motif was not valid.</li>
            <li>Fixed MCAST bug: would crash if any sequences was shorter than the widest motif.</li>
            <li>Fixed MCAST bug: would crash if any motif was only 1 wide; now motifs are skipped if
		their width is less than 2.</li>
	    <li>Fixed MCAST bug: would sometimes crash when --synth switch given.  MCAST
		results are now slightly different (different <i>p</i>-, <i>E</i>- and <i>q</i>-values, mostly.)</li>
	    <li>Fixed MCAST bug: crash could occur when there was only one or two motifs.</li>
            <li>Fixed MCAST bug: output would not display properly sometimes when the number of motifs 
		was very large.</li>
            <li>Fixed CentriMo bug: would crash if any motif was as long as the sequences; now motifs
		are skipped if they are not at least 2 shorter than the sequences.</li>
	    <li>Fixed missing documentation files fasta-center.html and fasta-dinucleotide-shuffle.html.</li>
          </ul>
        <li>
	  Improvements
	  <ul>
	    <li>AME: Checks that there are a total of at least 2 primary and control sequences.</li>
	    <li>AME webserver: Checks that there are a total of at least 2 primary and control sequences.</li>
	    <li>CentriMo webserver: Added option to use a uniform background model, or the model in the motif file.</li>
	    <li>MCAST webserver: Added option to not hard-mask (convert to 'N') lowercase nucleotides.</li>
	    <li>MEME webserver: Checks that there are at least 2 primary sequences unless you specify
		the 'Any Number of Repetitions' site distribution.</li>
	    <li>SpaMo webserver: Added advanced option to specify the background model.</li>
	    <li>MEME: reduce the amount of memory required, especially with the ANR model.  With ANR,
		no more than 2,000,000 local maxima will ever be stored (1,000,000 each for primary and 
		control sequences).</li>
	    <li>FASTA input: programs will now skip sequences that are too long, rather than dying.
	    <li>test_driver: added --valgrind switch to allow search for memory leaks.
	    <li>Plugged many memory leaks.
          </ul>
        </li>
        <li>
	  Documentation
	  <ul>
	    <li>MEME: Clarify that -nsites and -maxsites override -csites if the model is ANR, 
		and correct the description of the default for -maxsites.</li>
	  </ul>
        </li>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.0.3 (December 2018)</b>
      <ul>
	<li>
	  Bug fixes:
	  <ul>
	    <li>Fixed MAST bug: didn't read E-values of motifs correctly if they were >1e300; checks the log(E-value) of motifs now.
	    <li>Fixed MEME bug: -cons option didn't work when the model type is DE or SE.
	    <li>Fixed MEME bug: -mod zoops sensitivity was decreased for large datasets (greater than -searchsize)
		due to incorrect setting of the initial fraction of sequences with sites.
	    <li>Fixed gendb bug: -alph option caused crashing.
	    <li>Fixed AME bug: output erroneously called E-value threshold "Adjusted p-value threshold".
	    <li>Fixed AME bug: crashed sometimes when sequence names were > 100 characters (bug in shuffle_seq()).
	    <li>Fixed bug: parsing of MEME model file alphabet line was failing if they were missing the trailing ":".
	    <li>Fixed FIMO bug: crash occurred if a motif in a MEME text file had zero alphabet length.
	    <li>Fixed MCAST bug: crash occurred if a motif ID was longer than 50 characters.
	    <li>Fixed MEME bug: crash when sequence name contained a non-ASCII character; now replaces them with
		underscores.
	    <li>Fixed SPAMO bug: crash caused by bug in similar sequence removal algorithm
	  </ul>
	</li>
        <li>
	  Improvements
	  <ul>
	    <li>MEME: Improve results with large alphabets by making the default value of "fuzz" 
		(see MEME option -spfuzz) depend on the alphabet size: fuzz = 2 / size_of_alphabet.
	    <li>MCAST: Improved user message when the --hardmask switch is given and lower case letters
		are changed to wildcards.
	    <li>CentriMo: Improved failure message when the input file contains no sequences of the 
		correct length.
	    <li>MEME-ChiP: Improved documentation to emphasize that motif discovery is only in the
		<b>central regions</b> (100 characters by default) of the input sequences; 
		also removed reference to 50Mb website limit.
	    <li>test_driver: Added --valgrind switch to allow testing for memory errors.
          </ul>
        </li>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.0.2</b>
      <ul>
	<li>
	  Bug fixes
	  <ul>
	    <li>Fixed rna2meme bug when there is no word after the FASTA ID.</li>
	    <li>Fixed bug that caused failure of notification email for AME web app</li>
	    <li>Fixed bug that caused generation of spurious warning message file for AME web app</li>
	    <li>Fixed bug that caused miscounting of background peptides in MoMo with MoDL algorithm.</li>
	    <li>Fixed bug that caused MEME not to honor the limits on nsites.</li>
	    <li>Fixed menu link to MEME SUITE source code.</li>
	    <li>Fixed missing Messages file link in MEME-ChIP output.</li>
	    <li>Fixed missing output format documentation files.</li>
	  </ul>
	</li>
        <li>
	  Improvements
	  <ul>
	    <li>Added -searchsize 0 feature to MEME to make it emulate earlier versions of MEME
		when used with -objfun classic.</li>
	    <li>Added shuffled foreground peptides as default for MoMo background peptides.</li>
	    <li>Greatly improved computation of MoMo E-values.</li>
	    <li>Improved MoMo HTML output.</li>
	    <li>Added new MoMo TSV output file.</li>
            <li>Added new MoMo test.</li>
            <li>Fixed the momo.html file to give correct descriptions of --min-occurrences,
                --remove-unknowns and --eliminate-repeats regarding affect on background peptides.</li>
            <li>Created documentation file 'momo-output-format.html'.</li>
            <li>Fixed the rna2meme documentation file 'rna2meme.html' to corectly describe its purpose.</li>
            <li>Fixed the motif_conversion.html file to give better example of rna2meme.</li>
	    <li>Installation now allows the user to specify the URL for a local website, the
		alternate local website, and the source code repository.  
	        These locations are then updated throughout the installed command-line and website applications.  
                If desired, these can all be set to local URLs so that no links will point 
		outside of the user's installation of the MEME Suite.</li>
	     <li>Updated tool output HTML to specify the location of help on interpreting
		results and the repository fo downloading the source code as separate URLs.</li>
	     <li>Re-formatted FIMO HTML output to resemble the other tools' output.</li>
	  </ul>
	</li>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.0.1</b>
      <ul>
      <li>
      Bug fixes
      <ul>
      <li>Fixed bug in momo_webservice with <code>--remove_unknowns</code> switch</li>
      <li>Fixed AME bug triggered by sequences that are too short for the 
      widest motif</li>
      <li>Fixed NULL pointer bug in MEME triggered by <code>-maxsize</code>
      option</li>
      <li>Fixed bug in momo.test</li>
      </ul>
      </li>
      <li>
      Improvements
      <ul>
      <li>Enhancements to MoMo HTML output.</li>
      </ul>
      </li>
      </ul>
    </p>

    <p>
    <hr>
      <b>MEME version 5.0.0</b>
    </p>
      <ul>
        <li><b>Highlights</b></li>
          <ul>
            <li><b>MEME improvements:</b> 
              <ul>
                <li>Major <b>speed enhancements</b> have been made to MEME when used with <b>large inputs</b>.</li>
	        <li><b>Two new objective functions</b> have been added to MEME for finding differentially enriched
		and centrally enriched motifs, respectively.</li>
                <li>The <b>MEME webserver</b> now allows much larger inputs.  Inputs may now contain up to <b>500,000</b> sequences
		  (up from 1000).  The input form limit of <b>80,000,000</b> characters remains.</li>
              </ul>
            </li>
            <li><b>AME improvements:</b> The AME tool has been completely redesigned, its speed enhanced, and many bugs have 
		been corrected.</li>
	    <li><b>MEME-ChIP improvements:</b>
              <ul>
                <li>MEME-ChIP now <b>uses all input sequences.</b> It no longer subsamples the input sequences if there are more than
		600, which was previously needed due to MEME's time complexity with large inputs. The MEME-ChIP webserver
		now accepts inputs with up to <b>500,000</b> sequences.  The input form limit of <b>80,000,000</b> characters remains.</li>
                <li>The <b>MEME-ChIP output has been improved</b>, and now provides both an output summary in TSV format as well as
		a file containing all the enriched motifs found in MEME format.</li>
               </ul>
	    <li><b>New (TSV) output files:</b> Most tools now also produce tab-separated value format (TSV) output files.
		Links to each of the new output files have been added to the top of the tool's HTML output file.
	        These TSV files simplify and enhance using the output of the tools
		AME, CentriMo, FIMO, GOMo, MCAST, MEME-ChIP, SpaMo and Tomtom.  Each TSV output file contains
		a header line (without the "#" character so R will see the column names as variables).
		The main TSV output files also contain the version number, documentation link and command line.</li>
	    <li><b>More documentation:</b> Added documentation files for the output formats of 
                AME, CentriMo, FIMO, GOMo, MCAST, MEME-ChIP, SpaMo and Tomtom.
	        The MEME documenation now includes plots of running time as a function of input size.</li>
	    <li><b>Improved documentation:</b> 
              <ul>
                <li><b>Ambiguous characters:</b> Added documentation to each HTML man page stating how the tool treats ambiguous characters in sequences.</li>
	        <li><b>Alphabet description:</b> Added the alphabet description to the HTML output for: AME, CentriMo, SpaMo, MEME-ChIP, MCAST, FIMO, Tomtom.</li>
               </ul>
	    <li><b>Backward compatibility testing:</b> Added backward compatability testing of motif reading in motif-in.test (--comprehensive) for versions 
		as old as 4.7.0.</li>
            <li><b>Standard directory structure:</b> The directory structure of the MEME Suite installation now conforms to the
              <a href="http://refspecs.linuxfoundation.org/fhs.shtml">Filesytem Hierarchy Standard</a>.</li>
	    <li><b>Balanced background models:</b> When scanning both strands, all tools now "balance" the background model
		(if the alphabet has complements) before using it to create pseudocounts for the motifs.</li>
	    <li><b>Command-line consistency:</b> All programs now also accept <b>-bfile</b> if they just accepted -bgfile before.</li>
          </ul>
          
        <li><b>New Features and Enhancements</b>
          <ul>
	
	    <li><b>AME</b>
              <ul>
                <li>Increased speed by 38%.</li>
		<li>Added a TSV file summarizing the enriched motifs.</li>
		<li>Added a TSV file listing positive sequences for significant motifs (Fisher only).</li>
		<li>Added links to the TSV files at top of HTML output.</li>
	        <li>Made p-value computations in log space.</li>
		<li>Replaced the option --pvalue-threshold with option --hit-odds-fraction,
			which defines hits as having scores at least X% of the maximum
			possible score for the motif.  Whereas --pvalue-threshold only
			applied to the <i>totalhits</i> scoring method, --hit-odds-fraction applies
			to all scoring methods.</li>
		<li>Removed option --pwm-threshold.  The positive sequence score threshold 
			for the fisher method now uses the --hit-odds-fraction option.
	        <li>fisher method: now performs optimization over sequence score thresholds
		    in --fix-partition or --control modes.</li>
		<li>linreg method: renamed to 'pearson', it now computes p-values and optimizes the p-value of the
			Pearson's correlation coefficient, rather than mean squared error.</li>
		<li>spearmans method: now does partition maximization and computes p-values.</li>
		<li>Report p-value adjusted for number of tests and E-value (adjusted for number of motifs)
			instead of raw p-value and p-value adjusted for tests+motifs.</li>
		<li>Changed --pvalue-report-threshold to --evalue-report-threshold.</li>
		<li>Removed --linreg-normalize and --rs-method options.</li>
		<li>Added Pos, Neg, %TP, %FP and Score columns to Fisher test output.</li>
		<li>Changed "--method mhg" to "--method 3dmhg".</li>
		<li>Removed "--bgformat" option and made selection of background model identical to CentriMo's method.</li>
                <li>Changed the default pseudocount to 0.1 (like CentriMo) and changed the option.</li>
			from "--pseudocount" to "--motif-pseudo" (like CentriMo).</li>
		<li>Added --shuffle-- keyword to --control option, causing shuffled primary sequences
			to be used as the control sequences.</li>
		<li>Added --kmer <k> switch to cause <k>-mers frequencies to be conserved when shuffling.</li>
		<li>Added --seed to allow the random number seed to be set.</li>
	        <li>Added --inc and --exc switches to allow choosing motifs whose names match a pattern.</li>
		<li>Created documentation file 'ame-output-format.html'.</li>
              </ul>
            </li>

	    <li><b>AME Website</b>
              <ul>
		<li>Added "NONE" as choice for control sequences.</li>
		<li>Added all remaining association functions to menu: linear, spearman, 3dmhg, 4dmhg.</li>
		<li>Changed default association function to "fisher".</li>
              </ul>
            </li>

            <li><b>CentriMo</b>
	      <ul>
		<li>Changed centrimo.txt to centrimo.txv, a TSV file summarizing the enriched motifs.</li>
		<li>Added links to the TSV files at top of HTML output.</li>
		<li>Created documentation file 'centrimo-output-format.html'.</li>
		<li>Added "Concentration" field to HTML output, which gives the total probability
		in the central region whose size is controlled by the user-adjustable "smoothing window" size.
	        <li>added --inc and --exc switches to allow choosing motifs whose names match a pattern.</li>
		<li>added --dist switch to perform distance-to-center-of-sequence motif enrichment analysis.</li>
		<li>added --use_pvalues switch to allow thresholding on the adjusted motif p-value rather than the log-odds score.</li>
	        <li>removed --dist switch because it was not useful.</li>
	      </ul>
            </li>

            <li><b>DREME</b>
	      <ul>
		<li>Changed the -v option to -verbosity to match other tools in the MEME Suite.</li>
		<li>Added documentation on the sequence logo files created by DREME.</li>
              </ul>
 
            <li><b>FIMO</b>
	      <ul>
		<li>Changed fimo.txt to fimo.tsv.</li>
		<li>Added links to TSV and GFF3 output in HTML output.</li>
              </ul>
            </li>

            <li><b>FIMO Website</b>
	      <ul>
                <li>Removed non-functional user interface elements for submitting a custom alphabet file from web application.</li>
              </ul>
            </li>

            <li><b>GOMo</b>
	      <ul>
		<li>Changed gomo.txt to gomo.tsv, a TSV file summarizing the enriched motif-GO term pairs.</li>
		<li>Added header line to the TSV output.</li>
		<li>Added links to TSV and XML output in HTML output.</li>
              </ul>
            </li>

            <li><b>MAST Website</b>
	      <ul>
                <li>Removed non-functional user interface elements for submitting a custom alphabet file from web application.</li>
              </ul>
            </li>

	    <li><b>MEME</b>
              <ul>  

		<li>Improved MEME's speed.</li>
		  <ul>
		    <li>MEME's running time is now now O(N) for the OOPS model, 
			    and O(N log N) for the ZOOPS model,
			    where N is the total size of the input sequences.
		    </li>
		    <li> Added -maxsize option, which sets the maximum number of characters
			    from the input sequences that will be used during EM, the second
			    step of MEME. (All data is used in the final, discretization step.)
			    This reduces the time complexity of the second step of MEME
			    to a constant not dependent on the size of the input sequences.
		    </li>
		    <li> Added -csites option, which sets the maximum number of sites allowed
			    for motifs when using the "Classic" MEME objective function.
			    By default, the limit is 1000 sites.  Increasing this can
			    increase running time substantially due to the cost of computing
			    the value of the "Classic" objective function 
			    for motifs with many sites (time = O(#sites^3)).
		    </li>
		    <li> Added documentation on the sequence logo files created by MEME.
                    </li>
		  </ul>
		</li>

		<li>Enhanced MEME with new objective function for finding <u>differentially enriched</u> motifs.</li>
		  <ul>
		    <li>This objective function causes MEME to search for motifs that are enriched
			    in the primary sequences relative to the control sequences.
		    </li> 
		    <li>New objective function is selected using the -objfun option: -objfun de
		    </li> 
		    <li>The user may provide a set of control sequences via the -neg option;
			    otherwise MEME will create a control set of sequences by 
			    shuffling the primary sequences.
		    </li> 
		    <li>Added -shuf option, which controls the size of words whose frequencies
			    are preserved when shuffling sequences to create a control set.
		    <li>Motif significance is computed using the multiple hypergeometric test,
			    which is substantially faster than the "Classic" MEME objective function,
			    and yields less conservative E-values.
		    </li> 
		  </ul>
                </li>

		<li>Enhanced MEME with new objective function for finding <u>centrally enriched</u> motifs.</li>
		  <ul>
		    <li>This objective function causes MEME to search for motifs that are enriched
			    in the central regions of the sequences, which must all be of thes ame length.
		    </li> 
		    <li>New objective is selected using the -objfun option: -objfun ce 
		    </li> 
		    <li>Motif significance is computed using the binomial test,
			    which is substantially faster than the "Classic" MEME objective function,
			    and yields less conservative E-values.
		    </li> 
		    <li>This objective function is designed for use with sequences from whole 
			    genome approaches such as ChIP-seq and CLIP-seq, and the sequences should 
			    include adequate flanking region around the expected motif sites&mdash;e.g., 
			    use sequences of 500bp for ChIP-seq. 
		    </li> 
		  </ul>
                </li>

		<li>Improved MEME's use of background models.</li>
		  <ul> 
		    <li> Added the -markov_order option, which specifies the order of the
			    background model to generate from the input sequences (default),
			    or the maximum order model to extract from the file given
			    using the -bfile option.
                    </li>
		  </ul>
                </li>

		<li>Improved MEME's search for starting points (Classic objective function).</li>
		  <ul> 
		    <li> Added the -use_llr switch, which causes the log-likelihood ratio (rather
			    than the Classic objective function) to be used in evaluating possible
			    starting points for EM.  This allows for higher order background
			    models to affect the search for starting points.
		    </li>
		  </ul> 
                </li>

		<li>Improved MEME's search for motifs (non-Classic objective function).</li>
		  <ul> 
		    <li> MEME now detects if EM "wanders" away from the starting point
			    and uses the (discretized) starting point as the motif in that case.
		    </li>
                  </ul> 
                </li>

		<li>Other MEME enhancements</li>
                  <ul>
		    <li>Removed limitation on the number of motifs MEME can report.</li>
		    <li>Changed: MEME run from -cons will only try the given width with that start.</li>
                    <li>Fixed: computation of background frequencies in MEME when there are ambiguous characters.</li>
                    <li>Fixed: changed the default maximum value of -maxnsites to 600 when using the ANR model, matching the
                       documentation.</li>
                  </ul>
                </li>

              </ul>

	      <li><b>MEME Website</b>
		<ul>
		  <li> Increased allowed input size: Max sequences now 1,000,000 (from 1000).</li>
		  <li> Increased allowed input size: Max letters now 10,000,000 (from 60,000).</li>
		  <li> Added ability to create up to 4th order models from the input sequences.</li>
		  <li> Added "Differential Enrichment mode" as alternative to "Discriminative mode" (which
			uses a position specific prior).</li>
		</ul>
	      </li>
            </ul>

            <li><b>MEME-ChIP</b>
	      <ul>
		<li>Removed subsampling from MEME-ChIP--it now uses all the input sequences.</li>
		<li>Created documentation file 'meme-chip-output-format.html'.</li>
		<li>Added '-keepprimary' to command line.</li>
		<li>Added new test for custom alphabet mode.</li>
	      </ul>
	    <li><b>MEME-ChIP Website</b>
	      <ul>
		  <li> Added "Differential Enrichment mode" as alternative to "Discriminative mode" (which
			  uses a position specific prior).</li>
		  <li> Added ability to create up to 4th order models from the input sequences.</li>
		</ul>
            </li>

            <li><b>MoMo</b>
	      <ul>
		<li>Added documentation on the sequence logo files created by MoMo.</li>
		<li>Make "no hashing" the default for finding the location of peptides
			when a protein database is provided.
              </ul>
            </li>

            <li><b>SpaMo</b>
	      <ul>
		<li>Added feature to output all of the "contributing sequence IDs" as a text or BED file with one click of a button at the top of the output form.  Sequence IDs are introduced by a line containing the file name that would be used if the sequence IDs for a single primary-secondary spacing were output.
		<li>Fixed bug in -xalph that caused the wrong background for secondary motifs.</li>
		<li>Fixed bug in HTML display of number of significant motifs when &gt;1 secondary DBs.</li>
		<li>Fixed bug causing number of erased matches to be wrong with -keepprimary.</li>
		<li>Clarified use of -inc and -exc; they only affect secondary motifs.</li>
		<li>Changed TXT file to TSV file summarizing the enriched motifs.</li>
		<li>Added links to the TSV files at top of HTML output.</li>
		<li>Created documentation file 'spamo-output-format.html'.</li>
	      </ul>
            </li>

	    <li><b>Tomtom</b>
	      <ul>
		<li>Changed text output file name to "tomtom.tsv".</li>
		<li>Added links to TSV and XML output in HTML output.</li>
		<li>Added HTML to smoke tests.</li>
	      </ul>
            </li>


          </ul>
        </li>

      </ul>

     <hr>
      <b>MEME version 4.12.0 -- June 27, 2017</b>
      <ul>
        <li>
            <b>Enhancements</b>
            <ul>
              <li>
              Added MoMo application. 
              MoMo identifies sequence motifs associated with different types of protein post-translational modifications (PTMs).
              The program takes as input a collection of PTMs identified using protein mass spectrometry.
              For each distinct type of PTM, the momo command uses one of three algorithms to constructs a motifs 
              representing amino acid preferences flanking the modification site.
              </li>
            </ul>
        <li>
            <b>Bug fixes</b>
            <ul>
              <li>
              Fixed default values for centrimo options <code>--score</code> and
              <code>--motif-pseudo</code>.
              </li>
            </ul
        </li>
     </ul>
     <hr>
      <b>MEME version 4.11.4 patch 1 -- May 18, 2017</b>
      <ul>
        <li>
            <b>Bug fixes</b>
            <ul>
              <li>
              Fixed bug in Tomtom that disabled the -png and -eps options.
              </li>
              <li>
              Fixed bug in fasta-hamming-enrich with non-complementable alphabets.
              </li>
            </ul
        </li>
        <li>
            <b>Enhancements</b>
            <ul>
              <li>
              Improved documentation on background file inputs
              for AME, MEME, MEME-ChIP, MAST, CentriMo, and Tomtom.
              </li>
              <li>Added the reverse complement as the alternate ID in iupac2meme.</li>
              <li>Added documentation for meme2images script.</li>
            </ul
        </li>
      </ul>
    <hr>

      <b>MEME version 4.11.4 -- March 31 2017</b>
      <ul>
        <li>
          <b>Enhancements</b>
          <ul>
	    <li><b>AMA</b> GFF2 output now has only 9 fields as in the standard, and
		the AMA documentation has been corrected to reflect this.
	    <li><b>MEME</b> motif names are now their <b>consensus sequence</b>, 
		rather than a number, and the alternate ID is "MEME-i", where
		"i" = 1, 2, 3,... is the old motif name. MEME tests now check the integrity of
		all output formats: text, HTML and XML.
            <li><b>DREME</b> motifs now include an alternate ID "DREME-i", where
		"i" = 1, 2, 3... etc. is the index of the motif.
	    <li><b>CentriMo</b> output now contains a column for the motif consensus sequence,
		and the display of this column can be turned on or off.
	    <li><b>MEME-ChIP</b> now produces a <b>summary file</b> 
		in "tab-separated values" format suitable for use by programs such as Excel 
		and for easy parsing by scripts.
		The motif IDs in this file are synchronized with the IDs in the <b>motif file</b> 
		named "combined.meme" file, which contains all of the significant motifs 
		found by MEME-ChIP in MEME Motif Format.  
		New <b>links</b> to these two files are provided at the top of the 
		of MEME-ChIP's HTML output.
		The HTML output of command-line MEME-ChIP is now named "meme-chip.html", as it was
		for the web-server version.  The --index-name option has been removed.
	    <li><b>FIMO</b> output now includes a column for the alternate motif name.  Added command line
		to XML output.  Different FIMO tests check input of all three MEME formats: text, HTML and XML.
		Added the sequence name and strand to the "Group Name" field in GFF3 output so
		that GFF3 files can work with the UCSC Genome Browser, which requires groups members
		to all be on the same sequence and strand.
	    <li><b>Motiph</b> Added command line to XML output.
	    <li><b>MAST</b> text output (including with the -hit_list option) now includes columns
	 	for the motif name and alternate name in addition to the motif index.
	    <li><b>AME</b> --bfile option now overrides --bformat so it is only necessary
		to specify --bfile to get the desired result.
	    <li><b>AME</b> output now includes a column for the alternate motif name;
		in addition, the text output also includes the consensus of the motif.
            <li><b>SpaMo</b> Fixed alignment of inferred secondary logo when the secondary logo
		was trimmed on the left.  Added the alternate ID wherever the motif ID is displayed.
		Replaced spaces with underscores in the Inferred Secondary and Alignment motif names
		when they are downloaded as MEME motifs or EPS images.  Documented the problem
		with Macs not showing the scroll bar under the Alignment logos.
	    <li><b>GOMo</b> Fixed bug causing webserver page to crash if no GOMo databases are installed.
          </ul>
      </ul>

     <hr>
      <b>MEME version 4.11.3 patch 1 -- March 11, 2017</b>
      <ul>
        <li>
            <b>Bug fixes</b>
            <ul>
              <li>
	            Fixed bug causing FIMO not to generate GFF output for motifs using
              RNA or custom alphabets.
              </li>
              <li>
              Fixed bug that caused MEME-ChIP to halt with an error when TomTom finds no matches.
              </li>
              <li>
              Fixed bug that caused MEME Suite build and install to fail when no system copy 
              of the XSLT library was installed.
              </li>
              <li>Fixed Java 8 incompatibility in web site installation.</li>
            </ul
        </li>
      </ul>
      <hr>
      <b>MEME version 4.11.3 -- February 21, 2017</b>
      <ul>
        <li>
	  <b>Enhancements</b>
	  <ul>
            <li>Added AME, DREME, and FIMO tutorials.</li>
            <li>Improved consistency and documentation of -alph, -xalph and --bfile/-bgfile options across the MEME Suite.</li>
            </ul>
        </li>
        <li>
            <b>Bug fixes</b>
            <ul>
              <li>
	      Fixed bug causing MEME-ChIP to fail to show "Known or Similar Motifs" for MEME and DREME
              </li>
            </ul
        </li>
      </ul>
      <p>
    <hr>
      <b>MEME version 4.11.2 patch 2 -- October 24, 2016</b>
      <ul>
        <li>
            <b>Enhancements</b>
            <ul>
            <li>Added support for Python 3.</li>
            </ul>
        </li>
        <li>
            <b>Bug fixes</b>
            <ul>
              <li>
	      Fixed bug causing MEME-ChIP to fail to show "Known or Similar Motifs" for MEME and DREME
	      de novo motifs.
              </li>
              <li>
              Fixed bug in handling of RNA-like custom alphabets.
              </li>
              <li>
              Fixed bug in MAST -comp option.
              </li>
            </ul
        </li>
      </ul>
      <p>
    <hr>
      <b>MEME version 4.11.2 patch 1 -- June 16, 2016</b>
    </p>
      <ul>
        <li>
            <b>Bug fixes</b>
            <ul>
              <li>
              Fixed bug in MCAST 4.11.2 that caused it to prematurely truncate
              reading the sequence file.
              </li>
              <li>
              Modified MEME to fall back to a simple Dirichlet prior when
              using DNA or a custom alphabet with a prior that requires
              a prior library, but no prior library is specified.
              </li>
            </ul
        </li>
      </ul>
      <p>
      <hr>
        <b>MEME version 4.11.2 -- May 5 2016</b>
      </p>
      <ul>
        <li><b>Enhancements</b>
          <ul>
            <li>
              Reduced running time of DREME by 60%.
            </li>
            <li>
              Improved running time of FIMO.
            </li>
            <li>
              Added Ensembl Fungi, Ensembl Metazoa, Emsembl Protists, and
              Ensembl Plants to on-line sequence databases.
            </li>
            <li>
              Improved links to on-line annotations in MAST results.
            </li>
            <li>
              Added <code>fasta-re-match</code> script for scanning FASTA files
              with regular expressions.
            </li>
            <li>
              Updated installation guide.
            </li>
          </ul>
       </li>
      </ul>
      <hr>
      <p>
        <b>MEME version 4.11.1 -- Jan 15 2016</b>
      </p>
      <ul>
        <li><b>Bug fixes</b>
          <ul>
            <li>
            Script tag 'src' attributes for web applications now include
            a 'version' parameter to prevent outdated scripts from being cached.
            </li>
            <li>
              The C standard <code>gnu89</code> is specified for the build.
              This resolves problems building the MEME Suite with GCC 5.1.
              This was a common problem for users running the 
              latest version of Ubuntu.
            </li>
            <li>
              Fixed incorrect formatting of E-value threshold in MEME XML output.
              The incorrect formatting caused failures in <code>meme_xml_to_html</code>.
            </li>
            <li>
              Fixed error in TomTom web application when the RNA alphabet was
              selected.
            </li>
            <li>
              Fixed error in Javascript alphabet parser.
              The parser was not allowing blank lines,
              or lines consisting only of a comment.
            </li>
            <li>
              Fixed error in fasta-center and psp-gen scripts that prevented use of RNA
              alphabet.
            </li>
            <li>
              Fixed errors in specification of options, and parsing errors in 
              elm2meme script.
            </li>
            <li>
              Removed broken link to LAM MPI.
              LAM MPI has been superseded by Open MPI,
              and the host web site no longer exists.
            </li>
          </ul>
        </li>
      </ul>
      <hr>
      <p>
        <b>MEME version 4.11.0 -- Nov 23 2015</b>
      </p>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
            <li>Custom alphabet support
              <ul>
                <li>A majority of the programs in the MEME Suite now support
                custom alphabets including MEME, DREME, MEME-ChIP, CentriMo,
                SpaMo, AME, FIMO, MAST and Tomtom.</li>
                <li>RNA has been added to the list of built-in alphabets.</li>
              </ul>
            </li>
            <li>Priors for sequence databases
              <ul>
                <li>The FIMO and MCAST web services now have support for position
                specific priors for some selected databases.</li>
              </ul>
            </li>
            <li>SpaMo
               <ul>
                 <li>Sensitivity is improved by combining the counts for secondary motif sites in each
                        of the four possible orientations in 9 ways corresponding to patterns
                        expected when the primary and/or secondary motif is palindromic.</li>
                 <li>Sensitivity is improved by using all secondary motif sites scoring above the score threshold 
                        in the analysis rather than just the best-scoring site.  The old SpaMo approach
                        can still be used by specifying the new "-usebestsec" switch.</li>
                 <li>False positives are reduced because, by default, all occurrences
                        of the primary motif are "erased" (other than the central one).
                        This reduces the tendency of motifs similar to the primary
                        motif to appear enriched.  The -keepprimary switch can be used
                        to override this behavior so that spacings between multiple
                        occurrences of the primary motif can be observed.</li>
                 <li>The sequence identifiers for the sequences with the significant spacing
                        are now displayed to allow further downstream analysis (e.g., GO or pathway analysis).</li>
                 <li>The actual secondary sites are combined into an "inferred motif" and displayed
                        as a sequence logo in order to
                        make clearer what the putative sites look like.  This allows cases where
                        the significant secondary motif in the database is similar, but not the same,
                        as the motif actually present in the sequences.</li>
                 <li>The inferred motif is displayed as a frequency matrix to allow further
                        downstream analysis (e.g., using Tomtom to search other motif databases
                        to identify motifs more similar to the inferred motif.</li>
                 <li>Alignments of the sequences with the significant spacing are displayed as 
                        sequence logos.  Among other purposes, this can allow cases where the
                        sequence similarity extends beyond or between the primary and secondary sites,
                        suggesting either 1) the regions are repetitive or 2) the actual secondary
                        or primary motif sites are distinct from the respective primary and secondary
                        motifs used in the analysis.</li>
                 <li>The elimination of overly similar sequences has been sped up using a 
                        heuristic optimization.  A bug was also eliminated that caused the
                        comparison to involve only the given strands of DNA sequences rather
                        than both strands.</li>
                 <li>By default, all sequences containing ambiguous characters in the margins
                        around the primary motif are now removed. This is necessary for the 
                        new heuristic similarity search.  It also helps insure that the 
                        uniformity assumption of random motif occurrences since sites with
                        ambiguous characters are not scored.  This check is turned off when
                        the similarity search is disabled using the "-shared" switch.</li>
                 <li>A tab-delimited text output file "spamo.txt" is now created for
                        easy parsing by other programs.</li>
                 <li>The new -text switch causes SpaMo to create <b>only</b> the text
                        file "spamo.txt", skipping the creation of the HTML output file.</li>
                 <li>Changed handling of duplicate sequence IDs from 'die' to 'skip'.</li>
               </ul>
            </li>
            <li>fasta-shuffle-letters
               <ul>
                 <li>Added capacity to perform efficient, higher-order
                 shuffling of FASTA sequences.</li>
               </ul>
            </li>
          </ul>
        </li>
        <li><b>Bug fixes</b>
          <ul>
            <li>taipale2meme
               <ul>
                 <li>Changed naming of multiple motifs for
                 same TF to start with _1 (rather than no _).</li>
                 <li>Fixed bug that was giving motifs duplicate
                 names when _2 was reached.</li>
               </ul>
            </li>
          </ul>
        </li>
      </ul>
      <hr>
      <p>
        <b>MEME version 4.10.2 -- Aug 11, 2015</b>
      </p>
      <ul>
        <li><b>New Features and Enhancements</b>
        <ul>
          <li>General
            <ul>
              <li>
              Added the utility script <code>fasta-grep</code>
              for finding matches to a Perl regular expression
              in a FASTA file.
              </li>
              <li>Replaced use of system pseudo-random number generator
              with an implementation of Mersenne Twister, providing
              more consistent results across platforms</li>
            </ul>
          </li>
          <li>File conversion utilities
            <ul>
              <li>Added <code>-seed_start</code>, <code>-seed_end</code>
              and <code>-other_count</code> options to rna2meme</li>
            </ul>
          </li>
          <li>FIMO
            <ul>
              <li>Eliminated dependency on XSLT for output</li>
            </ul>
          </li>
          <li>MCAST
          <ul>
            <li>Added support for Position Specific Priors (PSP)</li>
            <li>Improved modeling of p-value distribution: GC content added as a
            parameter in EM estimation of exponential distribution.</li>
            <li>Eliminated dependency on XSLT for output.</li>
          </ul>
          <li>Web pages
            <ul>
              <li>Added a job submission rate limiter to the web server</li>
              <li>Added additional check to all job submission forms to ensure
              the combined form input size is smaller than 80MB
              </li>
              <li>Added links for viewing the load on the alternative servers</li>
              <li>Improvements to the display of sequence and motif databases</li>
            </ul>
          </li>
          <li>sequence-update database utility
          <ul>
            <li>Numerous enhancements and fixes</li>
          </ul>
          </li>
        </ul></li>
        <li><b>Bug fixes</b>
        <ul>
          <li>Documentation
            <ul>
              <li>Numerous documentation fixes and improvements</li>
            </ul>
          </li>
          <li>General
          <ul>
            <li>Added check for sequences shorter than the motif
            used to scan them</li>
            <li>
              Fixed the parsing of MEME motif files to handle cases when the
              E-values are written out in different formats
            </li>
            <li>Fixed the writing of MEME motif files so the 
            E-values are written out using the same formatting in (hopefully) all circumstances
            </li>
          </ul> 
          </li>
          <li>Web pages
            <ul>
              <li>Fixed the logic used to display server news and notices
               to take into account differences between the clock of the server
              and the local computer's clock. </li>
              <li>Fixed a bug where the webserver was not logging 
              the job submission times </li>
              <li>Fixed a bug in the website FASTA file validator
              which was incorrectly rejecting sequence weights for being out of range</li>
              <li>Added more obvious links to the motif database 
              overview page from the job submission pages</li>
              <li>Changed download page to prevent caching of the inner frame 
              in the Safari web browser</li>
              <li>Fixed handling of non-ASCII input on all web-forms</li>
              <li>Fixed a server error caused by the absence of cookies.</li>
              <li>Made disallowed alphabets for sequence inputs on web forms 
              more obvious by using a red cross-out over the names
              </li>
              <li>Fixed inclusion of example files for GLAM2 and minimal
               MEME motifs for the website install
              </li>
              <li>
                Improved the error message the MEME reports 
                when given a single sequence without using "any number of repeats" (ANR) mode
              </li>
              <li>Fixed incorrect URL in RSAT downloader configuration
              </li>
            </ul>
          </li>
          <li>AME
            <ul>
              <li>Fixed bug causing segfault</li>
              <li>
                  Fixed bug causing the maximum motif width to be incorrect
                  resulting in a dependency on motif order
              </li>
              <li>Fixed failure to generate log files</li>
            </ul>
          </li>
          <li>GLAM2
            <ul>
              <li>
                Fixed generation of sequence logos for GLAM2 motifs with pure deletion columns
              </li>
            </ul>
          </li>
          <li>MAST
            <ul>
              <li>MAST now ignores WEIGHTS section in FASTA files</li>
            </ul>
          </li>
          <li>Tomtom
            <ul>
              <li>
                Fixed submission to Tomtom from GLAM2 DNA output
              </li>
            </ul>
          </li>
        </ul>
        </li>
      </ul>
      <hr>
      <p>
        <b>MEME version 4.10.1 -- Mar 19, 2015</b>
      </p>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
            <li>Improved Web Pages
              <ul>
                <li>Updated the schematic on the home page and added more "hover" blurbs
                explaining what tools and databases do.</li>
                <li>Created separate man pages for the Web and Command Line versions
                of all commands.</li>
                <li>Use same blurb for blurb, web man page and command line man page for each tool.</li>
                <li>Added links to Manual on each tool input form (top right).</li>
                <li>Email address are not required for job submissions.</li>
              </ul>
            </li>
            <li>Improved test scripts
              <ul>
                <li>Updated the testing scripts to work with Automake's
                custom test driver specification.</li>
                <li>Split the tests into a file per program to make maintenance
                easier.</li>
              </ul>
            </li>
            <li>CentriMo
              <ul>
                <li>Added "Region Position" sort to allowing sorting from left to right.</li>
                <li>Renamed "Region Center" sort to "Region Distance".</li>
              </ul>
            </li>
            <li>FIMO
              <ul>
                <li>Improved the XML to HTML stylesheet.</li>
              </ul>
            </li>
            <li>MCAST
              <ul>
                <li>Improved the XML to HTML stylesheet.</li>
              </ul>
            </li>
            <li>SpaMo
               <ul>
                 <li>The elimination of overly similar sequences has been sped up using a
                  heuristic optimization.</li>
                 <li>By default, all sequences containing ambiguous characters in the margins
                  around the primary motif are now removed. This is necessary for the
                  new heuristic similarity search.  It also helps insure that the
                  uniformity assumption of random motif occurrences since sites with
                  ambiguous characters are not scored.  This check is turned off when
                  the similarity search is disabled using the "-shared" switch.
                  </li>
               </ul>
            </li>
            <li>getsize
               <ul>
                 <li>Added capacity to perform efficient, higher-order
                 shuffling of FASTA sequences.</li>
               </ul>
            </li>
            <li>update-sequence-db
              <ul>
                <li>Added extra configuration file options for updater URLs so
                in theory local mirrors can be used.</li>
              </ul>
            </li>
            <li>transfac2meme
            <li>beeml2meme
              <ul>
                <li>Added ability to use the "NA" line as
                the ID and make the consensus letter in 
                matrix lines optional in order to support
                TRANSFAC-like formats (i.e., Swiss Regulon).
                </li>
              </ul>
              <ul>
                <li>Added ability to process multiple files.</li>
                <li>Added ability to draw alternate ID and URL from a snapshot
                of a specific Uniprobe webpage.</li>
              </ul>
            </li>
            <li>iupac2meme
              <ul>
                <li>Added ability to invert the meaning of a symbol group with '^'.
                For example with a DNA motif [A] would match A but [^A] would
                match C, G, or T.</li>
              </ul>
            </li>
            <li>taipale2meme
               <ul>
                 <li>Changed naming of multiple motifs for
                 same TF to start with _1 (rather than no _).</li>
                 <li>Fixed bug that was giving motifs duplicate
                 names when _2 was reached.</li>
               </ul>
            </li>
            <li>README
              <ul>
                <li>Reformatted the README as Markdown styled text.</li>
              </ul>
            </li>
          </ul>
        </li>
        <li><b>Bug fixes</b>
          <ul>
            <li>Webpages
              <ul>
                <li>Fixed various issues with viewing the website and outputs
                in Internet Explorer.</li>
              </ul>
            </li>
            <li>Sequence uploads
              <ul>
                <li>RNA sequences are now converted to DNA again as was the
                behavior in release 4.9.1. This is a temporary fix to avoid
                problems in MEME and AME where 'U's were not supported. Support
                for custom alphabets is intended for a future release.</li>
              </ul>
            </li>
            <li>CentriMo
              <ul>
                <li>Ensure rounding up on window sites for the text output. 
                The text output was using bankers rounding which occasionally
                caused different results to the rounding displayed in the HTML
                output.</li>
              </ul>
            </li>
            <li>FIMO
              <ul>
                <li>Fixed PSP reader to honor --parse-genomic-coord option.</li>
                <li>Removed unused --max-seq-length option.</li>
                <li>Changed FIMO to not attempt to rewind the file when reading
                from standard input. FIMO will now report an error when you
                attempt to scan more than one motif when reading the sequences
                from standard input because handing multiple motifs requires
                rewinding the file with FIMO's current architecture.</li>
              </ul>
            </li>
            <li>MEME
              <ul>
                <li>Fixed bug when PSP files contain zeros; 1e-16 is now added to those entries.</li>
              </ul>
            </li>
            <li>MCAST
              <ul>
                <li>Fixed bugs in purging of low score matches when max stored matched reached.</li>
                <li>Fixed GC ranging and match purging code.</li>
              </ul>
            </li>
            <li>SpaMo
              <ul>
                <li>A bug was eliminated that caused the sequence similarity
                comparison to involve only the given strands of DNA sequences
                rather than both strands.</li>
              </ul>
            </li>
          </ul>
        </li>
      </ul>
      <hr>
      <p>
        <b>MEME version 4.10.0 -- May 6, 2014</b>
      </p>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
            <li>Website
              <ul>
                <li>The side-bar menu was completely rewritten to work in all the
                documentation even when the MEME Suite has not been installed.</li>
                <li>Many confusing, extraneous and hard to update summary pages
                that the old side-bar menu required have been removed.</li>
                <li>All the command-line documentation has been standardized to
                a common pattern which will hopefully make it easier to understand.</li>
                <li>The job submission logic has been rewritten using Java
                Servlets instead of Perl CGI scripts. This should enable better
                integration with Opal as the Perl interface to Opal had not been
                updated with the more recent Opal versions.</li>
                <li>The job submission pages use new techniques to validate
                files before you click the submit button.</li>
                <li>All the pages that accept MEME formatted motifs can accept
                hand typed motifs using IUPAC codes, lists of sites or
                frequency/count matrices.</li>

              </ul>
            </li>
            <li>MEME
               <ul>
                 <li>Removed the shell script that used to be needed to execute serial ('meme.bin') or
                 parallel ('meme_p') versions of MEME.  MEME is now a single executable ('meme') like
                 the other C-based MEME Suite programs.</li>
                 <li>New HTML output with a more compact overview, delayed
                 loading to handle display of larger sequence counts and a
                 combined motif site/scan diagram.</li>
               </ul>
            </li>
            <li>CentriMo
              <ul>
                <li>New feature in output to plot site probability vs. distance from center of sequence.</li>
              </ul>
            </li>
            <li>MEME-ChIP
               <ul>
                 <li>Print value of <code>-filter-thresh</code> in header of MOTIFS section of output.</li>
               </ul>
            </li>
            <li>AME
               <ul>
                 <li>Added web interface to the AME "Analysis of Motif Enrichment" tool.
                 AME reports (known) motifs that are enriched in your input nucleotide sequences
                 compared with either (a) shuffled versions of your sequences or (b)
                 control sequences that you supply.</li>
                 <li>Correct Fisher's Exact test to be one-tailed rather than two-tailed so
                 that under-represented motifs don't get small p-values.
                 <li>Disable "better" ranksum test mode (it's broken).
               </ul>
            </li>
            <li>rna2meme
               <ul>
                 <li>Allow reading from standard input if input filename is '-'.</li>
               </ul>
            </li>
            <li>Sequence Databases
               <ul>
                 <li>Added more sophisticated tool for server operators to
                 download sequence databases.</li>
                 <li>Added UCSC Mammal Genome sequence databases.</li>
                 <li>Sequence databases are now versioned where possible and timestamped where not.
                 You can select the desired version for search via MAST, FIMO, MCAST or GLAM2SCAN.</li>
               </ul>
            </li>
          </ul>
      </ul>
      <hr>
      <p>
        <b>MEME version 4.9.1 -- August 23, 2013</b>
      </p>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
            <li>AME
               <ul>
                 <li>Make "overwrite directory ame_out" default behavior like other MEME Suite programs.</li>
                 <li>Change test on --pvalue-report-threshold from "&lt;" to "&le;".</li>
                 <li>Add help popups for Split, p-value and Corrected p-value.</li>
               </ul>
            </li>
            <li>CentriMo
              <ul>
                <li>Added the ability to scan with the reverse complement
                of each motif separately (--sep).  This is useful when
                the sequences have strand information (e.g., TSSs) and the
                motif has a preferred orientation.</li>
                <li>Added the ability to limit the minimum size of regions
                whose significance is computed (--minreg).  This allows
                the user to restrict regions to a width range so that the
                significance levels of all motifs can be compared for
                regions of the same size.
                </li>
                <li>Added the ability to zoom-in on a section of the output graph.</li>
              </ul>
            </li>
            <li>DREME
               <ul>
                 <li>Add "Time" to end of text output file.</li>
               </ul>
            </li>
            <li>MCAST
               <ul>
                 <li>Move "display more" arrow in HTML output to left (same as MAST output).</li>
               </ul>
            </li>
            <li>MEME-ChIP
              <ul>
                <li>Documented the -group-thresh and -group-weak switches; changed
                  default -group-weak threshold to be only 2 times (not 10 times)
                  the -group-thresh to avoid excessive merging of motif clusters.
                </li>
                <li>The number of motifs to be found by MEME or DREME may
                now be set to zero on the command line 
                to prevent either of the programs from being
                run.  CentriMo and the motif clustering algorithm will still
                be run, allowing MEME-ChIP to be used as a front-end to CentriMo.
                </li>
                <li>Add -centrimo-flip option so that matches on reverse strand
                will be reflected around center of sequence.  This permits
                (non-palindromic) motifs with asymmetric binding to be detected.</li>
                <li>Added -meme-maxsize option to expose MEME's maxsize option.</li>
              </ul>
            </li>
            <li>SpaMo
              <ul>
                <li>The (slow) sequence redundancy step can now be skipped
                by setting the --shared option to zero.</li>
                <li>Added extra documentation recommending the use of a
                repeat-masker before running SpaMo.</li>
              </ul>
            </li>
            <li>Other
              <ul>
                <li>The utility fasta-get-markov is now implemented in c
                and is much faster than before. It is backwards compatible
                with the Perl implementation but now has some new features
                like the ability to specify input and output files, 
                a progress display and summary of the sequence
                counts/sizes like from getsize.</li>
                <li>Implemented a better hashing function for use by the 
                hash table datastructure which should improve performance
                and also fix an overflow bug that was present in the
                old function.</li>
                <li>Script fasta-fetch now handles sources with genomic coordinates that list negative strands.</li>
                <li>Disabled building of obsolete programs and made an effort to remove unused "fossil" code.</li>
                <li>Enforced separation of library and program code so Python and Perl libraries
                should no longer be installed into the bin directory.</li>
                <li>Added a link to the MEME Suite's Google Scholar page from the citing page.</li>
                <li>Improved resiliency of get_db_csv to poorly behaved FTP servers.</li>
              </ul>
            </li>
          </ul> 
        </li> 
        <li><b>Bug fixes</b>
        <ul>
          <li>AME
            <ul>
              <li>Fixed usage message to include defaults.</li>
              <li>Fix bug in computation of corrected p-value (LOGEV).</li>
              <li>Fixed a bug where short sequences combined being scanned
              with long motifs could cause the program to attempting
              scanning before the sequence start and into unallocated
              memory.</li>
            </ul>
          </li>
          <li>CentriMo
            <ul>
              <li>Fixed bug causing log-odds scores of "N" columns to be non-zero in CentriMo.</li>
              <li>Increased accuracy of log-odds scores in CentriMo by upping PSSM_RANGE.</li>
              <li>Fixed off by 0.5 error in output graph where the graph was shifted right by half a unit. 
              This error wasn't noticeable until we added the zooming feature.</li>
              <li>Now only creates links to motifs that have URLs.</li>
            </ul>
          </li>
          <li>DREME
            <ul>
              <li>Fixed a bug where DREME would crash when it couldn't find Ghostscript.</li>
            </ul>
          </li>
          <li>FIMO
            <ul>
              <li>Fixed bug where FASTA sequence IDs were truncated at ":" if they looked like BED format.</li>
              <li>Fixed bug with all "N" motifs causing crash.</li>
              <li>Fixed bug causing log-odds scores of "N" columns to be non-zero in FIMO.</li>
              <li>Fixed bug causing free of unallocated memory.</li>
              <li>Fixed bug where the release date would be truncated in output.</li>
              <li>Increased accuracy of log-odds scores in FIMO by upping PSSM_RANGE.</li>
            </ul>
          </li>
          <li>MAST
            <ul>
              <li>Fixed the output's display of the background model when the sequence composition was used.</li>
              <li>Corrected a link to an example motif file on the job submission form.</li>
            </ul>
          </li>
          <li>MCAST
            <ul>
              <li>Fix MCAST bug: wasn't printing motif match threshold.</li>
            </ul>
          </li>
          <li>MEME
            <ul>
              <li>Fixed memory link and double free when <code>-cons</code> option was used.</li>
            </ul>
          </li>
          <li>MEME-ChIP
            <ul>
              <li>Fixed a bug where a motif could be discovered by MEME but excluded by MEME-ChIP due to a large E-value
              and then be rediscovered by CentriMo by being centrally enriched. This sequence of events would cause the
              output to break as the Javascript made assumptions that were incorrect.</li>
              <li>Fixed a bug where on some Perl installations the timeout would not be caught and handled correctly.</li>
            </ul>
          </li>
          <li>SpaMo
            <ul>
              <li>Corrected a link to an example motif file on the job submission form.</li>
            </ul>
          </li>
          <li>Tomtom
            <ul>
              <li>Fixed bug in Tomtom's pearson_scores(): replaced integer division with floating point.</li>
            </ul>
          </li>
          <li>Other 
            <ul>
              <li>When sequences are skipped due to a duplicate name the warning message is not generic anymore.</li>
              <li>Replaced broken link to old user forum with link to new Q&amp;A site.</li>
              <li>Removed non-functional -P switch from ceqlogo. Use "-d ''" to supress fineprint. </li>
              <li>Added documentation of "--uniform--" keyword for "--bgfile" switch.</li>
              <li>Updated source for Eukaryotic Promoter Database (listed in etc/db_general.csv).</li>
              <li>Fixed a bug where the release date was being truncated by some older programs in the MEME Suite.</li>
              <li>Fixed usage message of create-priors.</li>
            </ul>
          </li>
        </ul>
        </li>
      </ul>
      
      <hr>
      <p>
        <b>MEME version 4.9.0 patch 4 -- January 18, 2013</b>
      </p>
      <ul>
        <li><b>New Features and Enhancements</b>
          <ul>
            <li>CentriMo
              <ul>

                <li>Added computation of <i>p-</i>value ratio whenever
                negative sequences are given. The <i>p-</i>value ratio
                seems to be more useful than the discriminative
                <i>p-</i>value computed in the <code>--disc</code> mode,
                and is much faster to compute. Motifs are now sorted by
                <i>p-</i>value ratio by default when negative sequences
                are given and <code>--disc</code> is not.</li>

                <li>Added feature to supress plotting of negative sequence
                  enrichment curves so a single run can be used for
                  comparative and absolute enrichment analysis.

                <li>Sorting on Region Center now sorts by increasing
                  absolute value (centrality) of the location of the region.</li>

                <li>The "region" sorting feature menu is now locked to the
                same value as the "motifs" feature menu, unless a checkbox
                is checked.  This helps prevent confusion.</li>


                <li>Improved documentation popups.</li>

                <li>Fixed bug in the Fisher's Exact test (affected AME,
                too).</li>

                <li>Added new standalone utility: fisher_exact</li>

              </ul>
            </li>
            <li>MEME-ChIP
              <ul>
                <li>Added -centrimo-local switch to allow CentriMo to perform
                local motif enrichment as part of MEME-ChIP.</li>
              </ul>
            <li>FIMO
              <ul>
                <li>Added <code>--parse-genomic-coords</code> option.
                This option turns on parsing of UCSC style genomic
                coordinates from sequence headers.</li>
              </ul>
            </li>
            <li>taipale2meme
              <ul>
                 <li>Add options [-nc <c>]+ and [-oc <c>] to accommodate
                 format of Jolma2013 Cell paper.</li>
              </ul>
            </li>
          </ul>
        </li>
        <li>
          <b>Bug fixes</b>
          <ul>
            <li>Fixed buffer overflow bug in shift_sequence().</li>

            <li>Fixed bug in DREME: -norc option resulted in incorrect PSSM.</li>
          </ul>
        </li>
      </ul>
    <hr>
    <p>
      <b>MEME version 4.9.0 -- October 2, 2012</b>
    </p>
    <ul>
      <li><b>New Features and Enhancements</b>
        <ul>
          <li>CentriMo
            <ul>
              <li>Added local mode for detecting uncentered enriched
              regions.</li>
              <li>Added discriminative mode which takes two datasets
              and uses the Fisher's exact test to detect discriminatively
              enriched regions.</li>
              <li>Added listing of sequence IDs in output for best
              enriched regions.</li>
              <li>Added score threshold optimization mode.</li>
              <li>Updated website form to give access to local mode and
              discriminative mode.</li>
            </ul>
          </li>
          <li>FIMO
            <ul>
              <li>Simplified command line options controlling output threshold.</li>
              <li>FIMO no longer emits a wiggle format file.
              The wiggle file contained the same information as the GFF file,
              but was more verbose.
              </li>
              <li>Output of '--text' option now matches format of 'fimo.txt' file.</li>
              <li>Text file now provides explicit strand information.</li>
              <li>Text file now lists start and stop coordinates in increasing order relative to the positive strand
              for all matches.</li>
              <li>HTML output is now limited to the 1000 most significant matches.</li>
            </ul>
          </li>
          <li>MEME-ChIP
            <ul>
              <li>Complete rewrite of the MEME-ChIP program to run MEME,
              DREME, Tomtom and CentriMo.</li>
              <li>New output which collates all the motifs found by the
              called programs with links to their outputs.</li>
              <li>New website form allows choice of motif database and
              adds more options for DREME and CentriMo.</li>
            </ul>
          </li>
          <li>Tomtom
            <ul>
              <li>Creation of EPS and PNG Logo files is now optional to 
              save disk space.
              Two new switches (-png and -eps) have been added to the 
              command line version of Tomtom.</li>
            </ul>
          </li>
          <li>meme-rename
            <ul>
              <li>Added new script meme-rename for mass renaming of
              output html files so they can inhabit the same directory.</li>
            </ul>
          </li>
          <li>psp-gen
            <ul>
              <li>Added handling of ambiguous IUPAC alphabet 
              symbols for DNA bases.</li>
            </ul>
          </li>
          <li>New Motif Databases
            <ul>
              <li>Malaria transcription factor motifs from PBM experiments
                  (Campbell et al., 2010 PLos Pathogens)</li>
              <li>New motifs derived from the original UniPROBE PBM data using 
                  a different algorithm (Zhao and Stormo,
                  2011 Nature Biotechnology)</li>
            </ul>
          </li>
        </ul>
      </li>
      <li>
        <b>Bug fixes</b>
        <ul>
          <li>Fixed FIMO bug--nucleotide motifs were labeled as polypeptide motifs in GFF output</li>
          <li>Fixed FIMO bug--both motifs and RC motifs were printing in list of best possible matches</li>
          <li>Fixed CentriMo bug--"--motif" option was not functional.
          <li>Fixed bug in gendb--was failing with higher-order models.
          <li>Fixed bug in gendb--was failing with higher-order models.
          <li>Updated the MEME installation guide</li>
          <li>Removed unused documents</li>
        </ul>
      </li>
    </ul>

    <hr>
    <p>
      <b>MEME version 4.8.1 -- February 6, 2012</b>
    </p>
    <ul>
      <li><b>New Features and Enhancements</b>
        <ul>
          <li><code>CentriMo</code>
            <ul>
              <li>Added method to create custom Encapsulated Postscript 
              files of the motif probability graph.</li>
            </ul>
          </li>
          <li><code>MEME-ChIP</code>
            <ul>
              <li>Added SpaMo to the list of programs that MEME-ChIP 
              runs.</li>
            </ul>
          </li>
        </ul>
      </li>
      <li><b>Bug fixes</b>
        <ul>
          <li>Fixed a typo which was breaking the "Upload your own 
          database" feature on Tomtom.</li>
        </ul>
      </li>
    </ul>
    <hr>
    <p>
      <b>MEME version 4.8.0 -- January 16, 2012</b>
    </p>
    <ul>
      <li><b>New Features and Enhancements</b>
        <ul>
          <li><code>CentriMo</code>
            <ul>
              <li> Added HTML output with lots of interactive features
              allowing selecting, sorting and filtering of which motifs to 
              display and plot the site-probability curves for.</li>
              <li>Added <code>CentriMo</code> web submission form.</li>
              <li><code>CentriMo</code> now runs as part of <code>MEME-ChIP</code>.</li>
              <li>Can now use motif matches to the given strand only if the
              <code>-norc</code> switch is given.  Useful with RNA sequences.</li>
              <li>Removed <code>centrimo-plots</code> program as the new
              HTML output replaces and improves on its purpose.</li>
            </ul>
          </li> 
          <li><code>DREME</code>
            <ul>
              <li>Can now look for motifs on the given strand only if the
              <code>-norc</code> switch is given.  This can be useful with
              RNA sequences.</li>
              <li>Using given strand only now available as an option on the web
              submission form.</li>
            </ul>
          </li>
          <li><code>MCAST</code>
            <ul>
              <li>Added support for scanning genome scale sequences.</li>
              <li>Added q-values for match scores.</li>
              <li>Revised HTML output to include block diagrams of matches.</li>
            </ul>
          </li>
          <li><code>SpaMo</code>
            <ul>
              <li>SpaMo now normalizes the p-values reported for particular
                motif spacings (bins) only for the number of spacings (bins)
                tested.  Previously, SpaMo reported p-values normalized for
                both the number of spacings and the number of motifs tested.</li>
              <li>SpaMo now prints E-values for secondary motifs, which is the
                lowest p-value of any spacing (bin) times the number of secondary
                motifs.</li>
              <li>Added -evalue switch to specify largest E-value allowed for
                secondary motifs.</li>
            </ul>
          </li>
          <li>Other
            <ul>
              <li>Improved memory efficiency for reading FASTA sequences in sequence.py.</li>
            </ul>
          </li>
        </ul>
      </li>
      <li><b>Bug fixes</b>
        <ul>
          <li>Fixed segmentation fault in FIMO when option
            <code>--bgfile motif-file</code> is used.</li>
          <li>Fixed png generation code for custom motif logos.</li>
        </ul>
      </li>
    </ul>
    <hr>
    <p>
      <b>MEME version 4.7.0 -- September 22, 2011</b>
    </p>
    <ul>
      <li><b>New Features and Enhancements</b>
      <ul>
        <li>DREME can be accessed as a webservice directly which
        allows use of the discriminative motif discovery.</li>
        <li>DREME now produces XML and HTML outputs as well as text. 
        These new outputs are accepted where MEME motifs are 
        accepted.</li>
        <li>New command-line program 
        <a href="centrimo.html">CentriMo</a> and the associated 
        <a href="centrimo-plots.html">CentriMo Plots</a> for 
        determining which motifs are centrally enriched in ChIP-seq 
        output.</li>
        <li>Completely rewrote motif loading code. The main user
        visible difference is that the minimal MEME format is a 
        lot more forgiving of spacing and MAST can now accept the
        minimal MEME format without any PSSM section.</li>
        <li>Output from all tools now lists the preferred citation.</li>
        <li>Updated and re-formatted list of publications.</li>
      </ul>
      </li>
      <li><b>Bug fixes</b>
      <ul>
        <li>Occasionally MEME would produce HTML files which could not 
        be used as a motif input to other programs. The new motif
        parser should be able to read these files and hence fix this
        problem.</li>
        <li>SpaMo background is now calculated from the sequences as
        stated in the paper.</li>
        <li>The website menu used to use the full URL which 
        unfortunately resulted in people bookmarking specific versions
        of the MEME suite rather than the URL for the latest version.
        The website menu has been corrected to use only relative 
        URLs.</li>
        <li>Links should now be created to the error messages when
        a webservice fails. This used to work but some setting changed
        the way redirects are buffered.</li>
        <li>Added compiler/linker directives to set stack size for 
        Cygwin to 9MB. Cygwin defaults to a 2MB stack, which is too 
        small.</li>
        <li>MAST was not handling the <code>-remcorr</code> switch.
        When a motif was removed because it was too correlated, the
        indices of the other motifs could be incorrect.  Fixed.</li>
      </ul>
      </li>
      <li><b>Known issues</b>
      <ul>
        <li>DREME does not yet support single strand scanning. It is 
        planned but wasn't ready for this release.</li>
        <li>Uploading extremely large sequence sets to MEME-ChIP can 
        cause a HTTP timeout and/or failure of the CGI script.</li>
      </ul>
      </li>
    </ul>
    <hr>
    <p>
      <b>MEME version 4.6.1 -- March 21, 2011</b>
    </p>
    <ul>
      <li><b>New Features and Enhancements</b>
        <ul>
          <li>
            <code>MEME-ChIP</code> command line tool --<br />
            MEME-ChIP was released in 4.6.0 as a web service but it did not
            have a command line equivalent. The webservice has been rewritten
            as a command line tool.
          </li>
          <li>
            <code>Tomtom</code> -- <br />
            The default Tomtom scoring algorithm now considers all columns in the
            query motif, not just the aligned columns. This reduces the
            number of spurious alignments including uninformative columns.
            The older scoring system is available via the
            <code>-incomplete-scores</code> option.
          </li>
          </ul>
      </li>
      <li><b>Bug fixes</b>
        <ul>
          <li>
          <code>MEME-ChIP</code> --<br />
          <ul>
            <li>
            MEME-ChIP has been rewritten to avoid infinite loops that
            could occur when the webservice was passed parameters including
            colons.
            </li>
            <li>
            The webservice now updates its output with the 
            completion of each sub-program so if it is terminated by Opal
            for running too long the already completed output is still available.
            </li>
            <li>
            Error reporting has been improved with the result that 
            MEME-ChIP should not be able to fail mysteriously as it has 
            on occasion been doing.
            </li>
          </ul>
          </li>
          <li>
            FIMO
            <ul>
            <li>
            The <code>--norc</code> option was being ignored.
            This has been fixed.
            </li>
            </ul>
          </li>
          <li>
            Motiph
            <ul>
            <li>
            The <code>--seed</code> option has been added to set the
            seed for the random number generator.
            </li>
            </ul>
          </li>
        </ul>
      </li>
      <li><b>Known issues</b>
        <ul>
          <li>MEME-ChIP when given very large sequence sets will not complete before the Opal webservice time limit.</li>
          <li>Uploading extremely large sequence sets to MEME-ChIP can cause a HTTP timeout and/or failure of the CGI script.</li>
        </ul>
      </li>
    </ul>
    <hr>
    <p>
      <b>MEME version 4.6.0 -- December 24, 2010</b>
    </p>
    <ul>
      <li><b>New Features and Enhancements</b>
        <ul>
          <li>
            <code>MEME-ChIP</code> web service --<br />
            <ul>
              <li><b>Motif Analysis of Large DNA Datasets</b><br />
              The MEME-ChIP service allows submission of a FASTA file of DNA sequences
              of unlimited size and with minimal parameter settings (only the traditional
              MEME web form settings) and runs it through a variety of motif finding and
              analysis tools.
              </li>
              <li>
              <code>dreme</code> --<br />
              added <code>dreme</code> and integrated it into the MEME-ChIP into web service.
              </li>
              <li>
              <code>AME</code> --<br />
              added <code>ame</code> and integrated it into the MEME-ChIP into web service. This
              is a cut-down version of the <code>ame</code> previously described, only supporting one mode
              of motif enrichment analysis.
              </li>
              <li>
              <code>fasta-dinucleotide-shuffle</code> --<br />
              added <code>scripts/fasta-dinucleotide-shuffle</code> and integrated it into the MEME-ChIP into web service.
              Thanks to Peter Clote for allowing use of his dinucleotide shuffle code.
              </li>
              <li>
              <code>fasta-center</code> --<br />
              added <code>fasta-center</code> and integrated it into the MEME-ChIP into web service.
              </li>
            </ul>
          </li>
          <li>
            <code>SpaMo</code> beta release --<br />
            The web interface and documentation is currently rather limited but in the interests of 
            getting this out for review this is a preview edition of the <b>Spaced Motif analysis tool</b>.
          </li>
          <li>
            <code>FIMO</code> -- <br />
              The <code>FIMO</code> HTML output now explicitly labels the strand,
              rather than indicating the reverse strand by having <cdoe>start</cdoe>
              &gt; than <code>stop</code>.
              The HTML output also now includes the matched sequence.
              The GFF output is now GFF 3.
              </li>
          </li>
          </ul>
      </li>
    </ul>
    <p>
      <b>MEME version 4.5.0 -- September 6, 2010</b>
    </p>
    <ul>
      <li><b>New Features and Enhancements</b>
      <ul>
        <li>
        <code>Tomtom</code> --<br />
        <ul>
          <li><b>Search multiple databases</b><br />
          Tomtom can now search more than one motif database. 
          To keep the ability to have web links for each motif
          they were moved into the MEME motif format itself.
          See the <a href="tomtom.html">documentation</a>.
          </li>
          <li><b>Improvements in submission form</b><br />
          Tomtom's submission form has many new enhancements.
          Motifs can now be specified in 3 formats: IUPAC motifs,
          frequency matrices, and MEME format files.
          Go <a href="../cgi-bin/tomtom.cgi">try it</a>.
          </li>
          <li><b>Improvements in output format</b><br />
          Tomtom now outputs an XML file which it converts into
          the HTML output. The HTML output can dynamically
          draw the motif logos when the images aren't available
          on web browsers supporting HTML5. Additional improvements
          include a summary table at the top giving direct links to the
          top 20 matches for each motif. See the 
          <a href="examples/tomtom_example_output_files/tomtom.html">sample</a>.
          </li>
        </ul>
        </li>
        <li><code>FIMO</code> --<br />
        The output for FIMO now includes a 
        <a href="https://cgwb.nci.nih.gov/goldenPath/help/wiggle.html">wiggle track format file</a>.
        </li>
        <li>MEME file format --<br />
        The MEME file format now includes an optional URL for each motif. This is used
        to create the links in the Tomtom output. All the XXXX2meme scripts have been
        retrofitted to take this URL parameter.
        </li>
        <li>
        <code>ama-qvalues</code> --<br />
        New script. See the <a href="ama-qvalues.html">documentation</a>.
        </li>
        <li>
        <code>fasta-subsample</code> --<br />
        New script for extracting a random sampling of the sequences
        in a FASTA file.  This is especially useful for ChIP-seq
        peak datasets to be input to MEME.  Using this script, a
        FASTA file containing a subset of the ChIP-seq peak sequences
        (or any other FASTA file) can be created.  The total
        number of sequences should be less than 1000 (preferably less
        than 500), and the total sequence length should be less
        than a few hundred thousand. MEME typically takes about 20 minutes
        per motif with files of 100,000 characters (DNA, both strands, ZOOPS model), 
        and scales quadratically in the total sequence length (so a file of
        200,000 characters will take four times as long.)
        This new script can also output the remaining sequences (in a separate
        file) for use in cross-validation. See the <a href="fasta-subsample.html">documentation</a>.
        </li>
        <li>
        <code>meme2meme</code> --<br />
        New script. See the <a href="meme2meme.html">documentation</a>.
        </li>
        <li>
        <code>rna2meme</code> --<br />
        New script. See the <a href="rna2meme.html">documentation</a>.
        </li>
      </ul>
      </li>
    </ul>
    <hr>
    <p>
      <b>MEME version 4.4.0 -- April 23, 2010</b>
    </p>
    <ul>
      <li><b> New Features and Enhancements </b>
      <ul>
        <li>
        <code>MCAST</code> --<br />
          <a href="mcast.html?man_type=web">MCAST</a> 
          now been integrated into the MEME Suite web applications.
          MCAST is a tool for searching for 
          statistically significant clusters of motifs in DNA sequences.
          MCAST was previously supported on a separate web site. 
        </li>
        <li>
        <code>MEME</code><br />
          <ul>
          <li><b> Discriminative motif discovery </b> <br />
          The MEME web server now supports discriminative
          motif discovery. The user may now provide a set 
          of "negative sequences" in addition to the normal
          "positive sequences", and MEME will look for 
          motifs that are overrepresented in the positive sequences
          relative to the negative sequences.
          </li>
          <li><b> Improvements in output format </b> <br />
          MEME's html output (<a href="examples/meme_example_output_files/meme.html">sample</a>) has been reorganized to 
          make important information more accessible as well as to make it aesthetically pleasing. Some of the improvements 
          include:<br />
            <ul>
              <li>
                A preview of the motifs found, including their reverse complement for DNA, in miniature at the top of the page.
              </li>
              <li>
                Ability to view full sized reverse-complemented DNA logos.
              </li>
              <li>
                Highlighting of all output for a sequence with a single click.
              </li>
              <li>
                New block diagrams that graphically show the strength (and DNA strand) of motifs, 
                and that scale with the size of the browser window.
              </li>
              <li>
                Support for the new web standard HTML5. This allows the motif logos to be viewed without the images
                on supporting web browsers (Firefox, Chrome, Safari). Unfortunately Opera has inadequate text support for canvas 
                (the new HTML5 element required for this feat) and Internet Explorer currently doesn't support canvas at all. 
                It is anticipated that Internet Explorer 9 will support this feature.
              </li>  
            </ul>
          </li>
          </ul>
        </li>
        <li>
        <code>MAST</code> --<br />
          <ul>
            <li><b> Improvements in output format </b> <br />
            </li>
            MAST's html output (<a href="examples/mast_example_output_files/mast.html">sample</a>) has been reorganized to 
            make important information more accessible as well as to make it aesthetically pleasing. Some of the improvements 
            include:<br />
            <ul>
              <li>
                Everything is now in one place - no longer do you have to search to find the associated annotated sequence for that
                high scoring sequence, as with one click the hidden information is displayed directly below. Additionally if DNA 
                strands are scored separately then the sequence only appears in the list once based on the better scoring strand.
              </li>
              <li>
                New block diagrams - incorporating all the improvements in the new MEME block diagrams such as graphical display of
                motif strength, scaling with the browser window and the DNA strand. As an added bonus separately scored strands 
                are visible on the same diagram.
              </li>
              <li>
                Interactive annotated sequence - a new range selector allows you to view the sequence around the motif hits of your 
                choice, directly relatable to the block diagram. For separate DNA strand scoring only one strand is visible at a time.
              </li>
              <li>
                Now in colour - the annotated sequence for motif hits now have letters colour coded using the same colour scheme 
                as MEME motifs.
              </li>
            </ul>
            <li> MAST now outputs xml in a similar fashion to MEME. This has resulted in differences to how you would invoke MAST
            from the <a href="mast.html">command line</a>. Additionally it no longer has the script wrapper.
            </li>
            <li>MAST now uses an xml style sheet to create its html output.
            </li>
          </ul>
        </li>
        <li>
        <code>mast2txt</code> --<br />
          Created program <a href="mast2txt.html">mast2txt</a> for outputting the old text format for MAST. Note that there are some
          problems with backwards compatibility in separate scoring mode. The use of the text format is no longer recommended.
        </li>
        <li>
          <code>psp-gen</code> --<br />
          added <code>psp-gen</code> and integrated PSP generation into web service.
        </li>
        <li>
          <code>MHMM</code> --<br />
          Fixed bug in <code>--order</code> option.
          The option was handling motifs with strand identifiers incorrectly.
        </li>
        <li><code>Tomtom</code> -- <br />
          Added <code>-no-ssc</code> switch to fix problem with
          motifs with few sites having empty logos; this switch is now
          the default for the Tomtom website.</li>
        <li><code>GOMO</code> --<br />
          Improved gomo html format.
          Added highlighting of the most specific GO terms in 
          the html output. This is activated by the switch --dag
          which expects a GO DAG file.</li>
        <li><code>gomo_highlight</code> --<br />
          Added a program to post process gomo xml files to
          identify the go terms which are implied by other
          more specific go terms to allow the specific terms
          to be highlighted in the html output. This is used 
          by gomo when the --dag switch is specified.
        </li>
        <li>
          <code>obo2dag</code> --<br />
          Added a tool for generating a GO DAG file from a
          Gene Ontology OBO file.
        </li>
        <li>
          <code>GO DAG format</code> --<br />
          Documented the new GO DAG file format and improved GOMO 
          documentation as well as the command line overview.
        </li>
        <li>
          <code>GLAM2</code> -- <br />
          The output now prints a regular expression in addition to the logo
          for each motif found.
        </li>
        <li>
        <code>MEME motif format</code> --<br />
          Fixed errors in motif format description 
          <code>doc/meme-format.html</code> and example files.
        </li>
        <li>
        <code>meme-io</code> --<br />
          Updated meme-io to accept the new MEME html format.
        </li>
      </ul>
    </ul>

    <hr />
    <p>
      <b>MEME version 4.3.0 -- September 26, 2009</b>
    </p>
    <ul>
      <li>GOMO -- 
        <br />
        Added comparative genomics to GOMO.  GOMO can now use
        multiple, comparative genomes in making its predictions.
        This substantially increases the sensitivity of predictions of roles
        for DNA binding motifs.  The GOMO web-service now
        uses GC-compensated AMA <i>p</i>-values.
      <li>AMA -- 
        <br />
        <i>P</i>-values are no longer printed by default.  A switch
        has been added to cause them to be printed.
<br />
        AMA can now compute GC-content compensated <i>p</i>-values.
        The GC-content of an individual sequence is used in estimating
        the <i>p</i>-value of its AMA score.
    </ul>

    <hr />

    <p>
      <b>MEME version 4.2.0 -- June 22, 2009</b>
    </p>
    <ul>
      <li> MEME -- 
        <br />
        MEME now can use position-specific priors (PSP) to improve
        motif discovery.  The PSP is specified using the 
        <code>-psp pspfile</code> switch to MEME.
      <li>GOMO -- 
        <br />
        Added the ability to choose either maximum or average
        binding affinity as the scoring function.  
        <br />
        Fixed a bug--was using
        maximum affinity by default, contrary to the documentation.
      <li>AMA -- 
        <br />
        AMA now computes <i>p</i>-value of each sequence based on
        a 0-order background model for the dataset.
        <br />
        Improved speed and memory efficiency; was re-reading sequences
        and storing them in memory.  
        <br />
        Fixed bug--was using max-odds by default
        and the --scoring switch was being ignored.
      <li>MAST -- website now supports protein databases even when
        DNA database contains long sequences.
      <li>Installation -- 
        <br />
        Now all programs in the MEME suite are built.
        The --enable-both configure switch is now obsolete.  
        <br />
        Simplified the build process by removing support for separately building
        the tools formerly included in Meta-MEME.
        <br />
        Update on-line documentation.
        <br />
        Simplified "quick install" instructions are provided in <code>INSTALL</code>
        and in <code>doc/meme-install.html</code>.
    </ul>

    <hr />

    <p>
      <b>MEME version 4.1.1</b>
    </p>
    <ul>
    <li>Improved documentation.</li>
    <li>Increased default number of iterations for GLAM2.</li>
    <li>Improved error handling for sequence data.</li>
    <li>Added links to server activity summary.</li>
    <li>Fixed bug in install process when <code>--enable-both</code> not used.</li>
    </ul>

    <hr />

    <p>
      <b>MEME version 4.1.0</b>
    </p>

    <ul>

      <li>New tool: GOMO -- Finding Genome Ontology terms associated with
        DNA binding motifs.
      <li>Added FIMO and GOMO buttons to MEME output to make it
        easy to use discovered motifs to search sequence databases
        and to identify the role of DNA binding motifs.
      <li>Fixed placement of sidebar menu in Internet Explorer for output of CGI scripts.</li>
      <li>Fixed support for OpenMPI in MEME.</li>
      <li>Improved the efficiency of FIMO.</li>
      <li>
      Made the <code>-text</code> option to FIMO produce results on-the-fly,
      without sorting or q-values.  This mode allows very large databases to
      be searched, without storing intermediate results.</li>

      <li>
      Modified the command line syntax for <code>gendb</code>, for
      consistency with the rest of the suite.</li>

    </ul>

    <hr>
    <p>
      <b>MEME version 4.0.0 -- June 13, 2008</b>
    </p>
    <ul>
      <li><b> New Features and Enhancements </b>
        <ul>
          <li> MEME and Meta-MEME are now integrated as "The MEME Suite".</li>
          <li> New web-based tools:
            <ul> 
              <li> GLAM2 and GLAM2SCAN--discovery and scanning of <b>gapped</b> motifs</li>
              <li> Tomtom--comparing motifs</li>
              <li> FIMO--basic motif scanning (complements MAST)</li>
            </ul>
          </li>
          <li> Enhancements of existing tools:
            <ul>
              <li> Sequence LOGOS now enhance MEME output.</li>
              <li> Compare new motifs to databases of motifs
              via Tomtom by clicking an easy-to-use button on MEME and GLAM2
              output.</li>
              <li> Improved local search (branching search) in MEME 
              (command line version only).</li>
              <li> Hundreds of new genomic and upstream sequence databases
              are now supported for scanning by MAST, FIMO or GLAM2SCAN.</li>
              <li> Tomtom supports four motif databases: JASPAR, Transfac, SCPD
              and MacIsaac Yeast.</li>
            </ul>
          <li> The MEME Suite tools now have web servers powered by OPAL,
          and return their results via the web, rather than by email.</li>
          <li> A single <a href="../search.html">web server</a> portal simplifies
          different types of motif scanning tasks</li>
        </ul>
      </li>
    </ul>
    <hr />
    <p>
    <b>MEME version 3.5.7 -- December 14, 2007</b>
    </p>
    <ul>
      <li><b> New Features and Enhancements </b>
        <ul>
          <li>Parallel MEME is compiled when no scheduler is found but MPICH is detected.</li>
          <li>Opal integration with meme.</li>
          <li>Added --enable-webservice=&lt;path_to_build-opal.xml&gt; configure flag.  
          Used to automate deployment of MEME inside an Opal installation as a web service.</li>
          <li>Added --enable-web=&lt;opal_url&gt; configure flag.  
          Used to enable building of the MEME web portal.  
          A URL for an Opal-exposed MEME web service must be entered.</li>
        </ul>
      </li>
    </ul>
    <hr/>
    <p>
    <b>MEME version 3.5.5 -- May 31, 2007</b>
    </p>
    <ul>
      <li><b> New Features and Enhancements </b>
        <ul>
          <li>MAST -hit_list can now handle very long sequences
      and prints ALL hits regardless of the sequence E-value cutoff.
      It splits them into "swaths".  The -hit_list switch now
      overrides the -seqp switch, and hits are NOT sorted
      by the E-value of the sequence.</li>
        </ul></li>
    </ul>

    <hr/>
    <p>
    <b>MEME version 3.5.4 -- September 21, 2006 </b>
    </p>
    <ul>
      <li><b> New Features and Enhancements </b>
        <ul>
          <li>Changed the MAST -hit_list behavior to print ONLY a list
          of the non-overlapping significant motif matches in each
          sequence in a machine-readable format.  
          (More complete documentation is given in the MAST "usage"
          message.  Type "mast" on the UNIX command line to see the usage message.)
          An example of the (complete) output of MAST using the -hit_list switch is:
          <pre>
      # All non-overlapping hits in all sequences.
      # sequence_name motif hit_start hit_end score hit_p-value
      ce1cg -2 8 22  1459.90 1.67e-06
      ara +2 2 16  1661.18 5.04e-08
      bglr1 +2 1 15  1274.97 1.42e-05
      cya -2 19 33  1101.37 6.64e-05
      gale +2 5 19  1076.21 8.11e-05
      ilv -2 6 20  1098.85 6.78e-05
      malk +2 37 51  1085.02 7.56e-05
      ompa +2 5 19  1583.18 2.43e-07
      # mast tests/meme.crp0.oops tests/crp0.s -stdout -hit_list -m 2
          </pre></li>
          <li>New "regular expression" section added to each MEME motif.</li>
        </ul></li>
      <li><b>Bug fixes </b>
        <ul>
          <li>Fixed printing of command name in MAST results.</li>
          <li>Fixed the handling of the "View FASTA", "View RAW" and "View motif summary"
          buttons on the MEME form.</li>
          <li>Improved documentation of the "Motif Summary" section of MEME output.</li>
          <li>Fixed bug that prevented having multiple MEME and MAST servers running on a
          single host by different users.  Each server is distinguished by the socket number
          it listens on.</li>
        </ul>
      </li>
     <li><b>Known Issues </b>
       <ul>
         <li>The "MetaMEME" button on the top of MEME results page is still not functional.
         Users are advised to go to the MetaMEME web site and submit manually at this time.
         </li>
         <li>MAST: Uploaded file names containing quotation marks may cause the search to
         fail.</li>
         <li>MAST: Uploaded sequences in DOS/Windows format may cause the search to fail.</li>
       </ul>
     </li>
    </ul>

    <hr/>

    <p>
    <b>MEME version 3.5.3 -- April 20, 2006</b>
    </p>
    <ul>
      <li><b> New Features and Enhancements </b>
        <ul>
          <li>The capability to search three JASPAR databases of DNA motifs
              for matches to a MEME-discovered motif has been added to the MEME
              output form.  Pressing the 'COMPARE PSPM' button on the results
              form of a MEME search of DNA sequences will submit the selected
              motif PSPM to the JASPAR 'Compare' server.  This will allow you
              to determine if your motif is similar to any known motifs contained
              in one of the three JASPAR motif databases.</li>
        </ul>
      </li>

      <li><b>Bug fixes </b>
        <ul>
          <li>The "MAST" button on the top of MEME results page now automatically loads the
          motifs from the MEME results for MAST search. A user only needs to choose the database
          (uploaded or server side) to search against.</li>
        </ul>
      </li>

    </ul>

    <hr/>

    <p>
      <b>MEME version 3.5.2 -- March 1, 2006</b>
    </p>
    <ul>
      <li><b> New Features and Enhancements </b>
        <ul>
          <li>Newly designed web interface using javascript and css styling.</li>
          <li>Add support for Macos X server startup.</li>
          <li>Add meme-install.html with detailed explanation of installation
            procedures.</li>
        </ul>
      </li>

      <li><b>Bug fixes </b> 
        <ul>
          <li>In "configure" remove check dependency on parallel condition when 
            creating server executables.</li>
          <li>Path to images/ in cgi scripts.</li>
          <li>Description syntax in linux startup script meme.linux.</li>
          <li>Provide fix for a cygwin's new line representation when searching strings.</li>
          <li>Bug in e_step (oops.c)</li>
          <li>Bug caused by typo "MEME_BINbin"</li>
          <li>Remove use of min() in meme.pl</li>
          <li>Remove obsolete argument sample_prob from subseq7()</li>
          <li>Remove OS-dependent declarations of accept(), bind(), connect(), 
          socket(), listen(), htons(). They are defined in system header files.</li>
        </ul>
      </li>

    </ul>

    <hr/>
    <p>
      <b>MEME version 3.5.1 -- December 15, 2005 </b>
    </p>
    <ul>
      <li>
      <b>Bug fixes </b> 
        <ul>
          <li>Revert binary and scripts names to what they used to be in version prior to 3.5.0.
              The binary executables have an extension ".bin", and the shell scripts have
              their extension ".csh" removed. This allows to keep compatibility with
              earlier versions.</li>
          <li>Change startup scripts to start servers as a user specified during configuration.
              If none were specified, then the uid of the user who runs configure will be
              used.</li>
          <li>Always install files in etc/. </li>
          <li>Always install Globals.pm in lib/perl/.</li>
          <li>Move meme-explanation.html from web/html/ to etc/. This allows to run scripts that do
              conversion to html when web site is not installed.</li>
        </ul>
      </li>
      <li>
        <b>Enhancements</b> 
        <ul>
          <li> Added support for Macos X (serial version only, without server)</li>
          <li> Users can now compare DNA motifs to the motifs in the JASPAR database of transcription
          factor binding site motifs.  A button is provided in the PSPM section of each DNA motif
          on the MEME output form for this purpose.  Clicking on the "Compare PSPM x to known motifs
          in JASPAR database" sends the motif to the JASPAR website, which returns any similar motifs
          that are found.</li>
          <li>Add meme_config.csh file that is used automatically by csh scripts to set
              needed environment variables.</li>
          <li>Move job.out file (created when SGE scheduler is used) to LOGS/</li>
          <li>Add '--with-serial' option to 'configure.ac' for compiling only serial
          version</li>
          <li>Disable installation of web by default. To install use --with-web configure
          option.</li>
          <li>Lower prerequisite version for autoconf to 2.53, and add missing functions in m4/.
              Add check for autoconf version in bootstrap, and use m4 functions when
              needed.</li>
          <li>Change option '--with-mpi' to '--with-mpidir'</li>
        </ul>
      </li>

    </ul>

    <hr/>
    <p>
    <b>MEME version 3.5.0 -- September 15, 2005 </b>
    </p>
    <p>This is the <b>12th year of MEME development</b>. The current anniversary
    maintenance release 
       for MEME contains a number of improvements to the installation and configuration of MEME/MAST 
       servers and clients. Thanks to Nadya Williams of SDSC/NBCR for this work.
    </p>
    <ul>
      <li>
        <b>Refactoring of installation and configuration</b> 
        <ul>
          <li>Restructure source directory tree to ease configuration.</li>
          <li>Highly simplified and optimized installation and configuration for MEME and MAST servers 
              and clients using GNU autoconf/automake tools. </li>
          <li>Multiple configuration options available via arguments to "configure".</li>
          <li>Default installation of MEME, MAST server and clients, as well as the web
          site.</li>
          <li>Separate installation of server, client or web site if desired.</li>
          <li>Create a small test suite for 'make test'.</li>
          <li>Implement procedures for standard 'configure', 'make', 'make test', and 'make
          install'.</li>
          <li>Create scripts to run MEME/MAST servers as services, and automatically restarted them 
              when physical host server reboots. Currently, Linux and SunOS are supported.</li>
          <li>Simplified start of the servers with 'start-mast' and 'start-meme' scripts.</li>
          <li>Encapsulate site-specific variables in Globals.pm for perl scripts 
              and in meme_config for sh scripts.</li>
          <li>Add csh and sh configuration scripts to set working environment.</li>
          <li>Availability as a Rocks Roll for integration with 
              <a href="http://www.rocksclusters.org">Rocks based cluster system</a>.</li>
        </ul>
      </li>
      <li>
        <b>Enhancements</b> 
        <ul>
          <li>Sent MEME/MAST search results via an email as attachments, rather than inlined in the 
              message body. This allows the MEME output to be saved properly for subsequent 
              use with MetaMEME. It also reduces the chance of the output being modified by email 
              filters.</li>
          <li>Check email addresses for entry accuracy using javascript. Thanks to Chris 
              Misleh of NBCR for this feature.</li>
          <li>Check email addresses for valid domain names.</li>
          <li>Create new 'runtests' script to run all test suites. Simplify parsing of the
          output.</li>
          <li>Update FAQ(s) for MEME and MAST.</li>
          <li>Separate web-related files by type (html, cgi or image) and create a dir structure 
              to support it. This simplifies web maintenance.</li>
          <li>Add examples of database and motif sequences files.</li> 
        </ul>
      </li>
      <li>
        <b>Known Issues</b> 
        <ul>

          <li>The new meme and mast are binary executables, not shell scripts as they used to be. 
              When run with the appropriate arguments, they produce text (non-html) output. To get 
              the "old" behavior of meme and mast with the html output, use meme.csh and mast.csh 
              (without -text option). The following table gives new vs. old usage comparison. 
    <pre>
                   old                                      new
          meme myfile -dna -text &gt; meme.res       meme myfile -dna &gt; meme.res         # output will be a text file
          meme myfile -dna &gt; meme.res             meme.csh myfile -dna &gt; meme.res     # output will be an html file
          mast meme.res                           mast.csh meme.res                   # output will be an html file
    </pre>
          </li>
        </ul>
      </li>

      <li>
        <b>Support Team Members</b>
        Wilfred W. Li, Ph.D., Nadya Williams, M.S., Chris Misleh, and Timothy Bailey, Ph.D.
      </li>
    </ul>

    <hr/>
    <p>
    <b>MEME version 3.0.14 -- July, 2005 </b>
    </p>
    <ul>
      <li>
        <b>Bug Fixes</b> 
        <ul>
          <li>
            an array out of bound error when MEME is run in parallel mode using certain DNA 
             sequences with the "look for palindromes only" option enabled. Thanks to Tim Kaiser of 
             SDSC/NBCR for the bug fix.
           </li>
        </ul>
      </li>
      <li>
        <b>User Support</b> 
        <ul>
          <li>all user questions are directed to our support team meme-suite@uw.edu.</li>
        </ul>
      </li>
    </ul>

    <hr/>
    <p>
    <b>MEME version 3.0 -- December, 2000 </b>
    </p>

    <ul>
      <li> 
        <b>MEME enhancements</b>
        <ul>
          <li>
            <b>Hypertext output:</b> MEME now reports its output in hypertext (HTML)
            format with appropriate internal hot links among the results and the
            self-contained documentation on how to interpret them.  The
            alignments are color-coded by nucleic acid or amino acid category.
          </li>
          <li>
            <b>Direct MAST search:</b>  MEME motifs can now be used to search 
            sequence databases using MAST simply by clicking on a button on the
            MEME HTML output document in an HTML-capable email reader or browser.
          </li>
          <li>
            <b>Direct BLOCKS search:</b>  MEME motifs can now be submitted to the
            BLOCKS multiple alignment processor by clicking on a button on
            the MEME HTML output document in an HTML-capable email reader or 
            browser.  This allows MEME motifs to be converted to LOGOS or trees,
            and to be used to search other databases of motifs.
          </li>
          <li>
            <b>New objective function:</b>  MEME searches for motifs that optimize
            the statistical significance of the log likelihood ratio of the
            occurrences of the motif.
          </li>
          <li>
            <b>E-values:</b>  MEME computes and reports the statistical significance
            of motifs as the E-value of the log likelihood ratio.  This provides
            an objective measure of how likely the motif is to be biologically
            significant.
          </li>
          <li>
            <b>E-value stopping criterion:</b>  MEME will now stop when a motif
            whose E-value is above a user-given threshold is found.  This guarantees
            that only motifs with a given E-value or better will be present in the
            output.
          </li>
          <li>
            <b>Handling reverse complement DNA strands:</b>  MEME now handles
            reverse complement DNA strands correctly with all model types:
            OOPS, ZOOPS and TCM.
          </li>
          <li>
            <b>Handling of ambiguous characters:</b>  MEME now handles ambiguous
            DNA and protein letters by converting them to the character "X". 
            MEME treats the "X" character as "unknown", and correctly computes
            the probabilities of motif occurrences containing it.
          </li>
          <li>
            <b>Higher-order background models:</b> MEME now allows the user to
            specify a Markov model of arbitrary order via a file of tuple 
            frequencies.  This appears to improve the ability to discriminate 
            between biologically interesting DNA motifs and motifs that are 
            artifacts of the higher order statistics of DNA sequence.
          </li>
          <li>
            <b>Multiple-alignment-based motif trimming:</b>  
            MEME defines a motif as
            a set of subsequences that can be correctly aligned without gaps.
            MEME can now trim
            the edges of motifs based on a local multiple alignment with gaps.  MEME
            first determines the occurrences of the motif, then aligns each
            occurrence to the highest-scoring occurrence.  These are combined
            into a multiple alignment, and MEME looks for a set of columns with
            no gaps.  If a set of gapless columns at least &lt;minw&gt; wide is not
            found, MEME searches for a set with at most 1, 2, ... etc gaps until
            a set is found.
          </li>
          <li>
            <b>Prior distribution on the number of occurrences:</b>  MEME now
            places a prior on the number of occurrences that controls how
            strong the bias towards motifs with a given number of sites is.
            Using a prior improves the performance of MEME with DNA, where the
            megaprior heuristic (used by default with protein sequences),
            is not applicable.
          </li>
        </ul>
      </li>

      <li>
        <b>MAST enhancements</b>
        <ul>
          <li>
            <b>Combining DNA strands:</b>  MAST now combines the score of a 
            site in a DNA sequence with the score of the corresponding site
            on the reverse complement strand.  (The final score for the site
            is the maximum of the two scores.)
          </li>
          <li>
            <b>Scoring DNA strands separately:</b>  MAST still allows each strand
            of a DNA sequence to be treated as a separate sequence at the
            user's request.</li>
          <li>
            <b>Ignoring reverse complement DNA:</b>  MAST also allows the user to
            score only the given DNA strand, not scoring the reverse complement
            strand at all.
          </li>
          <li>
            <b>Background model:</b>  MAST now allows the user to specify the
            background residue frequencies used for computing E-values of scores.
          </li>
          <li>
            <b>Composition-adjusted statistics:</b>  MAST can now use a different
            random model for each target sequence, based on the letter composition
            of that sequence.  This can greatly reduce erroneous matches due
            to biased sequence composition.
          </li>
          <li>
            <b>New databases:</b>
            The MAST website includes several new databases including "upstream"
            sequences for E. coli, B. subtilis and S. cerevesiae, and many complete
            genomes from GenBank.
          </li>
        </ul>
      </li>

      <li>
        <b>MEME and MAST enhancements</b>
        <ul>
          <li>
            <b>MEME and MAST can be installed under the following operating systems:</b>
            <ul>
              <li> Mac OS X</li>
              <li> Linux (various manufacturers)</li>
              <li> SunSparc workstations (SunOS and Solaris operating systems)</li>
              <li> Decalpha workstations (OSF/1 operating system)</li>
              <li> Silicon Graphics workstations (IRIX version 5.3 operating system)</li>
              <li> Intel Paragon parallel computers</li>
              <li> Cray T3D parallel computers</li>
              <li> IBM SP (AIX operating system)</li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>

    <hr/> 
    <p>
    <b>MEME version 2.2 -- February 25, 1998 </b>
    </p>
    <ul>
      <li>
        <b>MEME enhancements</b>
        <ul>
          <li>Sequence weights can be given in the input sequence file.</li>
          <li>
            The sites composing each motif are output in BLOCKS or FASTA format.
            BLOCKS format motifs can be converted to LOGOS, PSSMS and phylogeny trees
            using the <a href="http://blocks.fhcrc.org/blocks/process_blocks.html">
            Blocks Multiple Alignment Processor</a>.
          </li>
          <li>
            A "negative" dataset may be specified causing the motifs to be optimized 
            to discriminate between the training set family and the negative family.
            (This option is only available when you install MEME on your own computer.)
          </li>
          <li>
            An advanced version of the MEME data submission form allows 
            <ul>
              <li> discovering palindromic DNA motifs,</li>
              <li> discovering motifs occurring on both DNA strands, and</li>
              <li> discovering more than ten motifs.</li>
            </ul>
          </li>
        </ul>
      </li>
      <li>
        <b>MAST enhancements</b>
        <ul>
          <li>
            DNA sequences (translated in six reading frames) can be searched
            using protein motifs.
          </li>
          <li>
            <a href="examples/mast_example_output_files/mast.html">Hypertext (HTML) output</a>
            including links to the ENTREZ database and
            improved motif diagrams are now output.
          </li>
          <li>Additional searchable databases added to the MAST website.</li>
          <li>
            The annotation section of output is smaller now since only
            regions in sequences where the motifs are present are printed.
          </li>
          <li>
            The ambiguous characters "*" and "-" are now permitted in motifs
            and sequence databases.  They are treated as a single, unknown
            character in databases, and replaced by a weighted average of
            scores in motifs.
          </li>
        </ul>
      </li>
    </ul>

    <hr/>
    <p>
      <b>MEME version 2.1 -- March 25, 1997 </b>
    </p>

    <ul>
      <li>
        <b>Enhancements</b>
        <ul>
          <li>MAST output now includes measurements of similarities between
              all pairs of motifs in the query and a warning if any motifs
              are too similar.  (Too similar motifs in a query can cause
              some p-values and e-values to be underestimated.  These motifs
              can be removed from the query and MAST re-run to avoid the problem.)</li>
          <li>Cray T3E parallel computer now supported for MEME and MAST.</li>
        </ul>
      </li>
    </ul>

    <hr/> 
    <p>
    <b>MEME version 2.0 -- October 17, 1996 </b>
    </p>
    <ul>
      <li>
        <b>Enhancements</b>
        <ul>
          <li> 
            The output format of MEME has been 
            improved to aid readability and user-friendliness.
          </li>
          <li> 
            MEME and MAST  have been compiled and tested on:
            <ul>
              <li> SunSparc workstations (SunOS and Solaris operating systems)</li>
              <li> Decalpha workstations (OSF/1 operating system)</li>
              <li> Silicon Graphics workstations (IRIX version 5.3 operating system)</li>
              <li> Intel Paragon parallel computers</li>
              <li> Cray T3D parallel computers </li>
            </ul>
          </li>

          <li>
            MAST has been completely rewritten to provide more accurate 
            p-values and to run significantly faster.
          </li>
          <li>A MAST server has been added to the MEME system website.</li>
          <li>
            The MAST server lets you use the motifs found by MEME in your sequences
            to search a wide variety of sequence databases including:
            <ul>
              <li> nr (non-redundant protein and nucleotide databases)</li>
              <li> month (new or revised sequences in the last 30 days)</li>
              <li> 
                swissprot (the last major release of the SWISS-PROT protein sequence 
                database)
              </li>
              <li> genpept (protein translations from the GenBank feature table)</li>
              <li> dbest (expressed sequence tag database)</li>
              <li> dbsts (sequence tag site database)</li>
            </ul>
          </li>

          <li> Several improvements have been made to the MEME website: 
            <ul>
              <li> 
                You can now give the name of a file containing your sequences 
                instead of having to cut-and-paste the sequences themselves.
              </li>
              <li> 
                You may ask MEME to favor wide motifs instead of the (default) narrow
                motifs;  this is especially useful with small sequence sets (fewer than
                10 sequences.)
              </li>
              <li> The on-line documentation of MEME has been expanded and improved.</li>
              <li> 
                MEME now determines the type of your sequences
                (PROTEIN or DNA) automatically.
              </li>
              <li>
                The MEME server now runs MAST on your sequences to make it easy for you 
                to see the ordering and spacing of the motifs MEME discovered.
              </li>
            </ul>
          </li>
          <li>
            NOTE and PROBER have been removed from the MEME system since their 
            functionality is now superseded by MAST.
          </li>
        </ul>
      </li>
    </ul>

    <hr/>
    <p>
      <b>MEME version 1.4 -- February 29, 1996 </b>
    </p>
    <ul>
      <li>
        Megaprior heuristic added to MEME.
        <ul>
          <li>
            MEME now uses the megaprior heuristic with TCM models and the modified
            megaprior heuristic with ZOOPS models (by default).  This greatly improves
            the sensitivity and selectivity of motifs using these types of models.
          </li>
        </ul>
      </li>
      <li>
        P-value computation added to MAST.
        <ul> 
          <li>
            MAST now computes p-values for sequences when searching large databases.
            This provides better discrimination between true and false positives than
            using z-scores.  MAST is still only available via ftp, not via the web server.
          </li>
        </ul>
      </li>
      <li> Various bug fixes to MEME. </li>
    </ul>

    <hr/>
    <p>
    <b>MEME version 1.3 -- December 18, 1995 </b>
    </p>
    <ul>
      <li>MEME web server introduced!</li>
      <li> Various bug fixes to MEME.</li>
    </ul>

    <hr/>
    <p>
    <b>MEME version 1.2 -- November 27, 1995 </b>
    </p>
    <ul>
      <li> MAST homology searcher introduced!
        <ul>
          <li>
             MAST allows accurate homology searches of databases using motifs generated
             by MEME.  The scores from all motifs characterizing a family are combined,
             normalized for sequence length, and sorted by z-score.  The output of MAST
             is similar to that of BLAST, but, used in conjunction with MEME, MAST
             allows searching for sequences related to an entire family, not just a
             single sequence in the family.  This provides better sensitivity and
             selectivity than single-sequence homology searches.
             MAST is currently only available in the ftp-able code, not via the web server.
          </li>
        </ul>
      </li>
      <li> Various bug fixes to MEME.</li>
    </ul>

    <hr/> 

    <p>
      <b>MEME version 1.1 -- June 30, 1995 </b>
    </p>
    <ul>
      <li> Initial release of MEME, NOTE and PROBER.</li>
    </ul>
    <hr />
    <h2>Meta-MEME Release Notes</h2>
    <hr />
    <b>Release 3.3</b>
    <ul>
      <li>
        The program <code>motif-scan</code> has been renamed to
        <a href="fimo.html"><code>fimo</code></a>
      </li>
    </ul>
    <hr />
    <b>Release 3.x, Release 3.3</b>, March 24, 2007</b>
    <ul>
      <li>
        Added new tools:
        <dl>
          <dt>
            <a href="fimo.html">motif-scan</a>
          </dt>
          <dd>
            Scans a database of sequences for the presence of motifs by 
            calculating the p-value of the match to the motif at each position 
            in the sequences.
          </dd>
          <dt>
            <a href="shadow.html">shadow</a>
          </dt>
          <dd>
            Calculates the log-odds of a multiple alignment for a given phylogentic
            tree and evolutionary model at each position in the alignment.
          </dd>
          <dt>
            <a href="motiph.html">motiph</a>
          </dt>
          <dd>
            Scans a multiple alignment for the presence of motifs by
            by calculating the p-values of motif-width windows in the alignment
            for a given phylogentic tree and evolutionary model.
          </dd>
          <dt>
            <a href="tomtom.html">tom-tom</a>
          </dt>
          <dd>
            Searches target motifs for elements similar to any in a set of query motifs.
          </dd>
        </dl>
      </li>
      <li>
        Added <code>--global</code> option to <code>mhmms</code>. 
        This option causes <code>mhmms</code> to score sequences using the best 
        match between the model and the entire sequence.
        The default is to score using the best local match within a sequence.
      </li>
      <li>
        Added <code>--maxhits</code> option to <code>mhmms</code>. 
        This option sets an absolute limit to the number of hits 
        returned by <code>mhmms</code>.
        The default is to return all hits consistent with the
        E-value and p-value thresholds.
      </li>
      <li>
        Fixed error in <code>mhmms</code> that resulted in a sequence being considered 
        a hit even if all positions were matched to a spacer state.
      </li>
      <li>
        Fixed error causing segmentation a fault in <code>mhmms</code> when the 
        <code>--motif-scoring</code> or <code>--pthresh</code> options were used.
      </li>
      <li>
        Fixed error causing segmentation a fault in <code>mhmm</code> when the 
        <code>--order</code> option was used.
      </li>
    </ul>
    <hr />
    <b>Release 3.21, August 3, 2005</b>
    <ul>
      <li>Modified meme-io.c to be able to parse the output from the latest versions of MEME.</li>
      <li>
        Modified the mcast script: removed the -allow-weak-motifs swich, and added the --keep-unused switch
        to the call to mhmm from mcast.
      </li>
    </ul>
    <hr />
    <b>Release 3.2,  October 7. 2004</b>
    <ul>
      <li>Reorganized source directory structure and implemented build
      system using autoconf and automake.</li>
      <li>Updated program documentation.</li>
    </ul>
    <hr />
    <b>Release 3.1, February 4, 2004</b>
    <ul>
      <li>Remove the auxiliary program <code>score-n-store</code>.</li>
      <li>Change command-line processing to follow POSIX.2 standard.</li>
    </ul>
    <hr />
    <b>Release 3.0.1, June 2, 2003</b>
    <ul>
      <li>Add documentation for MCAST.</li>
    </ul>
    <hr />
    <b>Release 3.0, May 15, 2003</b>
    <ul>
      <li>Remove <code>mhmmt</code> (EM training program) and
      <code>mhmma</code> (multiple alignment program).</li>
      <li>Add <code>mhmmscan</code> and the <code>mcast</code> wrapper
      program for searching DNA databases for regulatory modules.</li>
      <li>Allow star topology in <code>mhmm</code>.</li>
    </ul>
    <hr />
  </body>
</html>

